---
title: "RC(M)"
author: "Stijn"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document:
    includes:
      in_header: packagesRCM.sty
---

```{r check-install-load-packages, warning=FALSE, message=FALSE, echo=FALSE, purl=TRUE}
WD = "/home/stijn/PhD/Biplots"
knitr::opts_chunk$set(cache = TRUE, autodep = TRUE,  warning=FALSE, message=FALSE, echo=FALSE, eval=TRUE, tidy = TRUE, fig.width=9, fig.height=6, root.dir=WD, purl=TRUE, fig.show = "hold", cache.lazy = FALSE)
# The required package list:
reqpkg <- c("phyloseq","MASS", "parallel","nleqslv", "edgeR", "VGAM","HMP", "ggplot2", "pscl", "vegan", "zCompositions", "DirFactor","alabama", "reshape2", "tensor", "locfit", "SimSeq","fdrtool", "cluster", "ape", "SpiecEasi", "Matrix")
# Load all required packages and show version
for(i in reqpkg)
{
#   print(i) 
#   print(packageVersion(i))
  library(i, quietly=TRUE, verbose=FALSE, warn.conflicts=FALSE, character.only=TRUE)
} 
if(detectCores()==4) setwd(WD)
par(pty="s", mar = c(8,2,4,2), cex.main = 0.9) #Make sure the biplots are square!
palStore = palette()
levelsMeth = c("CA","DCA", "Hellinger","BC","BCrel","BCrelNMDS", "JSD","UniFrac","wUniFrac", "CoDa","RCM", "Control")
labelsMeth = c("CA","DCA","Hellinger","Bray-Curtis-Abs","Bray-Curtis","Bray-Curtis NMDS", "JSD","UniFrac","weighted UniFrac","CoDa","RCM", "Control")
groupsMeth = factor(c("Independence","Independence","Independence","Distance","Distance","Distance","Distance","Distance","Distance", "CoDa", "RCM"), levels = c("Independence","Distance","CoDa","RCM"), ordered = TRUE)
names(groupsMeth) = grep(levelsMeth, value  = TRUE, pattern ="Control", invert = TRUE)
funFiles = dir("R")
funFilesPub = dir("pubFun")
for (i in funFiles) {source(file.path("R",i))}
for (i in funFilesPub) {source(file.path("pubFun",i))}
```

\setcounter{tocdepth}{2}
\tableofcontents

# Introduction

The aim of this notebook is to wrap the codes to fit the RC(M) model with the negative binomial, zero-inflated poisson and zero-inflated negative binomial models. Also an overview of the theory is given here.

The underlying motivation is to develop a visualization technique for the microbiome that does not suffer from the defects of currently applied techniques such as PCA, PCoA with Bray-Curtis distance etc. These methods are in particular sensitive to differences in library size.

# Theory

## Unconstrained analysis

### Correspondence analysis

Suppose we have a $nxp$ count data matrix $\mathbf{X}$ with $n$ samples and $p$ taxa, with $i$ and $j$ row respectively column indices. The count matrix has been trimmed to exclude all-zero rows and columns.

#### Independence model

Under independence between rows and columns we model the counts in a contingency table as 

$$E(X_{ij}) = a_i b_j$$

whereby usually $a_i=x_{i.}=\sum_{j=1}^px_{ij}$ (the library sizes) and $b_j=\frac{x_{.j}}{x_{..}}=\frac{\sum_{i=1}^nx_{ij}}{\sum_{j=1}^p\sum_{i=1}^nx_{ij}}$ (the average relative abundances).

#### Reconsitution formula of Correspondence Analysis (CA)

A more extended model is

$$E(X_{ij}) = a_i  b_j + c_i d_j \sum_{k=1}^K \omega_k v_{ki} w_{jk}$$.

For $a_i = c_i = x_{i.}$ and $b_j = d_j = \frac{x_{.j}}{x_{..}}$ this is regular correspondence analysis. The second term in the equation represents deviation from independence. This is called the *reconstitution formula* since it decomposes the observed average count into its expectation under independence and a residual. The residual is then further decomposed into $k$ pieces. The expected count can then be written as

$$E(X_{ij}) = \frac{x_{i.}x_{.j}}{x_{..}} \big(1 + \sum_{k=1}^K \omega_k v_{ki} w_{jk}\big)$$.

Regular corrsepondence analysis is done through singular value decomposition(SVD) of the matrix of weighted (Pearson) residuals

$$R^{-1/2}(X-E)J^{-1/2}=U\Sigma V$$

with $\Sigma$ a diagonal matrix with the singular values of $X$ on the diagnonal and $R$ and $J$ diagonal matrices with row and column sums.

In matrix notation the reconstitution formula becomes

$$X = E_{independence} + R^{1/2}U\Sigma VJ^{1/2}$$

with $\mathbf{1'}R^{1/2}U=\mathbf{0}$ and $\mathbf{1'}J^{1/2}V=\mathbf{0}$ (weighted means equal zero) and $U'R^{1/2}U=\mathbf{1}$ and $V'K^{1/2}V=\mathbf{1}$ (weighted variances equal one) (see VanDerHeijden et al., 1985).

For plotting usually $\omega_1 \mathbf{v_{1}}$ is plotted versus $\omega_2 \mathbf{v_{2}}$ to correctly represent the distances between the samples. $\mathbf{w_{1}}$ and $\mathbf{w_{2}}$ can then be added to construct a biplot. The orthogonal projection of $(\omega_1 v_{ki}, \omega_2 v_{ki})$ on $(w_{j1},w_{j2})$ then equals $\frac{(\omega_1 v_{ki}, \omega_2 v_{ki})^t(w_{j1},w_{j2})}{||(w_{j1},w_{j2})||}$. This value is proportional to the departure from the independence model of taxon $j$ in sample $i$ in dimension $k$ ($(\omega_1 v_{ki}, \omega_2 v_{ki})^t(w_{j1},w_{j2}) = \omega_1 v_{ki} w_{j1} + \omega_2 v_{ki} w_{j2}$), with the norm of the species vector as proportionality constant.

Note that the use of correspondence analysis implies unimodal response curves (see later). Hence it is not appropriate to represent species as arrows. Representing them as points is more appropriate.

Correspondence analysis has been criticised for putting undue emphasis on rare species (Legendre and Gallagher, 2001).

#### Skewness

Because of the skewness of the distribution, the residuals are also very skewed. If we have an expectation of around 1, it is very likely to observe a zero, but much less likely to observe a 2. Still both deviations are weighted equally. I think a residual-based approach is ill suited to asymmetric, skewed data. Also the library size correction will never be appropriate, so that part of the biplot will reflect differences in library sizes. We considered decomposing

$$R^{-1}(X-E)J^{-1}$$

in the past to account for the fact that the data are not Poisson ($Var(X) = E(X)$) but rather negative binomial ($Var(X) = E(X) + \phi*E(X)^2$) and thus the variance for large means proportional to the squared expectation. We still use this decomposition to obtain starting values for our RC(M) model. Even more general we considered

$$R^{-a}(X-E)J^{-b}$$

and estimate $a$ and $b$ by maximum likelihood but we did not really see how to go on from here.

#### Detrended correspondence analysis

In regular correspondence analysis it is often observed that the second axis points form an arch with respect to the first axis. This effect can be removed by detrending. The procedure is then called Detrended correspondence analysis (DCA) (Hill and Gauch, 1980).

#### Link to Gaussian ordination and niche theory

Species can grow and reproduce best within certain ranges of parameter values that describe their envrionment. The multidimensional space spanned by environmental variables in which a certain species can grow is called its __niche__. The species prospers in the center of its niche, but its abundance decreases as the characteristics of the environment depart from this optimal conditions. In extreme cases the species simply cannot survive or reproduce in certain envrironments and is simply absent. Hence this niche is usually decribed by a unimodal function, often a Gaussian function (Gauc _et al._, 1974).

The expected counts are then modelled as:

$$log(E(X_{ij})) = \sum_{k=1}^K \big(a_{jk} - \frac{(h_{ik} - q_{jk})^2}{2m_{jk}^2}\big)$$

whereby 

 - $h_{ik}$ = a latent variable representing an environmental score
 - $a_{jk}$ is the maximal logged expected abundance achieved when the latent variable $h_{ik}$ equals the optimum $q_{jk}$
 - $m_{jk}$ is the tolerance, it determines how fast the expected value decreases as $\mathbf{c}_{i.}$ moves away from its optimal values $\boldsymbol{\alpha}_k^{-t}h_{ik}$

Under the assumptions that:

  1) All tolerances $m_{jk}$ are equal for all species ($m_k$)
  2) All maxima $a_{jk}$ are equal for all species
  3) The modi $q_{jk}$ are homogeneously distributed over a region larger than $m_k$
  4) Counts are Poisson distributed
  
  CA yields the ML solution of this Gaussian model see ter Braak, 1986).

### Log-linear analysis

Another modelling approach is to use log-linear modelling an thereby introduce the negative binomial as error term. We know from previous goodness of fit testing that this is an appropriate error model for microbiome data.

In log-linear analysis the logged expected count $l_{ij}$ is modelled as

$$log(E(X_{ij})) = l_{ij} = u + u_i + u_j + u_{ij}$$.

For $u_{ij} \neq 0$ this is the saturated model, which provides a perfect fit by setting all expected values equal to the observed ones. If $u_{ij} = 0$ this is the independence model presented above.

### The RC(M)-association model

For the purpose of making biplots, we might consider the following model (proposed by Leo Goodman, 1985. See also VanDerHeijden 1994)

$$log(E(x_{ij})) = l_{ij} = u + u_i + u_j + \psi_1r_{1i}s_{1j} + \psi_2r_{2i}s_{2j}$$

or in matrix form:

$$\mathbf{lE} = \mathbf{u_{sam,MLE}}\mathbf{u_{tax,MLE}}^t + \mathbf{r}_k \boldsymbol{\psi}_k \mathbf{s}_k^t$$

with $\mathbf{lE}$ the n-by-p matrix of logged expectations, $\mathbf{u_{sam,MLE}}$ the vector with length $n$ with the row intercepts estimated by ML, $\mathbf{u_{tax,MLE}}$ the vector of length $p$ with the column intercepts estimated by ML, $\mathbf{r}_k$ the n-by-k matrix of row scores, $\boldsymbol{\psi}_k$ a diagonal matrix with the $k$ $\psi$'s on the diagonal and $\mathbf{s}_k$ the p-by-k matrix of column scores

Constraints are needed to render this model identifiable, since we have three unknown regressors ($r_{ik}, s_{jk} and psi_k$) instead of one. We do restrict the importance parameter $psi_k$ to be positive, and center the (weighted) row and column scores around 0, which is a useful property for the biplot and is also the case for correspondence analysis. In addition the row and column scores are restricted to have (weighted) variance 1. $psi_k$ is the only parameter that can grow in size without restriction. As a result it will automatically figure as a measure of importance of the departure from independence in that direction, since all scores of the dimensions are normalized. This is also the case for correspondence analysis. Thridly, the scores of the different dimensions are orthogonal, so the solutions in different dimensions are linearly independent.

Centering:

$$\sum_{i=1}^nw_ir_{ki} = 0$$

with k=1,2 and

Normalization($k=k'$) and orthogonality ($k \neq k'$)

$$\sum_{i=1}^nw_ir_{ki}r_{k'i} = I(k=k')$$

$$\sum_{j=1}^pz_js_{kj} = 0$$

$$\sum_{j=1}^pz_js_{kj}s_{k'j} = I(k=k')$$
Hereby $w_i$ and $z_j$ are row and column weigths.

#### RC(M)-weights

The weights can be regarded as probability density functions, they represent the likelihood of sampling a certain sample or taxon from the population. On the population level we could say that

$$E(w_iX_{ik}) = 0$$,

i.e. the average row score on the population level is zero. This is a useful restriction to make sure that the biplot is centered around zero. Analogously we want that

$$E(w_iX_{ik}X_{ik'}) = I(k=k')$$

and accordingly for the column scores:
$$E(z_j S_{jk}) = 0$$

$$E(z_j S_{jk} S_{jk'}) = I(k=k')$$. 

But which weights $w_i$ and $z_j$ should we use?

Goodman proposes to use $w_i = x_{i.}$ and $z_j = x_{.j}$ thus uses _weighted_ constraints, to retain the relationship with correspondence analysis (see below). Becker _et al._ (1989) recommend using uniform weights not to let the marginal distribution affect the model fit. 

For the microbiome case, every subject comes from the same population under the null-hypothesis and all subjects are thus eqaully likely to be sampled and have the same importance. The library sizes as a purely technical artefact, unrelated to the biological importance of the subject. Consequently we use uniform row weights $w_i = 1/n$ (or $w_i = 1$, the magnitude is of no importance since the associated $\psi_k$ will grow or shrink accordingly). However, some taxa are more prevalent in the population than others. We want the average column scores on the population level to be centered around zero, have variance one and be orthogonal. That is why we set $z_j = exp(u_j)$; because the more abundant species are in fact more abundant in the population as a whole (as opposed to samples with a large library size), it makes sense to use a marginal weighting scheme for the column scores. The weights $z_j$ will be estimated by maximum likelihood, as explained in the next paragraph.

#### RC(M) margins

The most logical choices for the margins and the independence model have long seemed to use $u=-log(x_{..})$, $u_i = log(x_{i.})$ and $u_j = log(x_{.j})$.

However, the library sizes do not correspond to the maximum likelihood estimate of $u_i$ under the Negative Binomial model. As a result the first dimensional row scores $r_{1i}$ tried to correct for this effect and became related (linearly correlated) to the library sizes, which we want to avoid absolutley. 

To avoid this we set $u=0$ and estimate the $u_i$'s, $u_j$'s and overdispersions iteratively using maximum likelihood in a first step (this converges very quickly). The main disadvantage I see of this approach is that the independence model becomes model dependent. Next we'll have to iterate between fitting the overdispersions, the imporance parameters $\psi$, the $r's$ and the $s's$. Hereby we use marginal weights for the taxa ($z_j = exp(u_j)$) and uniform weights for the rows. 

In practice, the marginal sums differ more from the MLE for the library sizes than for the taxon abundances. We can think of the following mathematical explanation: when calculating library sizes or abundances by row and column sums, each observations receives the same weight. However, when we estimate the margins through ML, we solve the following score equations:

$$\frac{\partial L(\mathbf{X}_i|\mathbf{u}_{i}, \mathbf{u}_j, \boldsymbol{\theta})}{\partial u_{i}} = \sum_{j=1}^p \frac{y_{ij}-\mu_{ij}}{1+\frac{\mu_{ij}}{\theta_j}}$$

$$\frac{\partial L(\mathbf{X}_i|\mathbf{u}_{i}, \mathbf{u}_j, \boldsymbol{\theta})}{\partial u_{j}} = \sum_{i=1}^n \frac{y_{ij}-\mu_{ij}}{1+\frac{\mu_{ij}}{\theta_j}}.$$

Note that since we are estimating offsets the value of the regressor is 1. In this case the difference of $y_{ij}$ with the expected value $\mu_{ij}$ is weighted by a factor $\frac{1}{1+\frac{\mu_{ij}}{\theta_j}} = \frac{\theta_j}{\theta_j+\mu_{ij}}$. When estimating $u_j$, $\theta$ is a constant but when estimating $u_i$ it is different for every observation $y_{ij}$. Hence the weights put on every observation differ much more when estimating the sample offsets than when estimating the taxon offsets. That is the reason why the MLE differs more from the marginal sum for the library size than for the abundances.

Note also that when there is a very large overdispersion for a taxon _j_ ($\theta_j$ small), its observations carry little information and their weights are small in the calculation of the library sizes. However, when there is very little overdispersion ($\theta_j \rightarrow \infty$), the weights of the components of the score function equal 1, as with Poisson regression. These MLEs of the Poisson regression are also equal to the estimators based on the marginal sums. This means that the larger and the more diverse the overdispersion estimates are, the more the MLEs under the negative binomial model will depart from the marginal sums. Finally, we see that departures from the mean $\mu_{ij}$ are weighted down for large values of $\mu_{ij}$, acknowledging the fact that the variance increases faster than linear with the mean in the negative binomial model.

#### Imposing the restrictions

The centering, normalization and orthogonality of the row and column scores are enforced through Lagrange multipliers. This makes the systems of equations harder to solve but assures independence of the dimensions.

#### A note on gradient analysis

Note that this is a case of __indirect gradient__ analysis: the dimensions are estimated without incorporating gradient information (measured covariates) to maximally represent variability in the data. In a next step these obtained gradients can be compared with measured covariates (often graphically through a colour code) to see how well these covariates explain the gradients. 

#### Relationship between CA and log-linear analysis

According to Escoufier, 1985 if $a =\sum_{k=1}^K \omega_k v_{ki} w_{jk}$ is small (i.e. the deviation from independence is small) then $log(1+a) \approx a$ and

$$log(E(x_{ij})) = log(x_{i.}) + log(x_{.j}) - log(x_{..}) + log\big(1 + \sum_{k=1}^K \omega_k v_{ki} w_{jk}\big) \approx log(x_{i.}) + log(x_{.j}) - log(x_{..}) + \sum_{k=1}^K \omega_k v_{ki} w_{jk}$$

i.e. an equivalence between the RC(M)-model and correspondence analysis

Since the same restricitions apply to the scores $v_{ki}$ and $w_{jk}$ as to U and V, we can state that  $\psi_k \approx \omega_k$. The assumption that the departure from independence is small seems unlikely in our case, but it does provide us with some starting values.

#### The RC(M) model and time series

Often we have data from experiments with repeated measurements, e.g. the evolution of the infant microbiome over time. Should we take this into account while fitting the RC(M) model? I do not think so, since it is an exploratroy method and we just want to visualize the variation in the data. It would be interesting to see if the individuals form clusters or not. In the constrained version of this method (see below), we can always add labels for the individuals and  time variable.

### Fitting the RC(M) model

The models are usually fitted by ML. See page 55 of Goodman 1985 for details. Goodman assumed a Poisson distribution, but it might be possible to use a negative binomial model with unique dispersions for each taxon $j$ as well.

See Goodman 1985 (appendix), Becker 1988 for information on estimation. The code below is generic but we take k=2 for simplicity.

Parameters of interest to estimate are $\psi$, $r_1$, $r_2$, $c_1$, $c_2$. In the process we'll also need the nuisance parameters $\mathbf{\theta}$ and the offsets of the independence model $u_i$ and $u_j$. Assume $E(X_{ij}) = \mu_{ij}$ and  $Var(X_{ij})=\mu_{ij} + \frac{\mu_{ij}^2}{\theta}$

#### The negative binomial density function

For the sake of completeness we give the density function of the negative binomial distribution (see Lawless, 1987) in the ($\mu$, $\theta$) parametrization whereby $E(X_{ij}) = \mu_{ij}$ and $Var(X_{ij}) = \mu_{ij} + \frac{\mu_{ij}^2}{\theta_j}$ .

$$f_{NB}(X_{ij}) = \frac{\Gamma(X_{ij}+\theta_j)}{\Gamma(\theta_j)} \big(\frac{\mu_{ij}}{\theta_j+\mu_{ij}}\big)^{X_{ij}} \big(\frac{\theta_j}{\theta_j+\mu_{ij}}\big)^{\theta_j}$$

#### Starting values

1. Obtain a singular value decomposition as $R^{-c}(X-E)C^{-d} = U\Sigma V$. This gives us initial values $[r_{11}^{init}, r_{12}^{init},..., r_{1i}^{init}] = U_{(1.)}$, the first row of $U$, and correspondingly $\Sigma$ and $V$ give initial values for $\mathbf{\psi}$ and $\mathbf{s}$. The values for $c$ and $d$ were determind based on correlation with the final values, and $c=d=1$ performed best.

We still need to ensure that the (weighted) variances equal 1, so we set

$$r_{ki}^{init} = \big(\frac{w_ir_{ki}^{SVD}}{\sum_{i=1}^n{r^{SVD}_{ki}}^2w_i}\big)^{1/2}$$

and

$$s_{ik}^{init} = \big(\frac{z_js_{ik}^{SVD}}{\sum_{i=1}^n{s^{SVD}_{ik}}^2z_j}\big)^{1/2}$$

#### Iteration

2. If the independence model is estimated by MLE, estimate the $u_i$'s, $u_j$'s and $\theta_j$'s iteratively until convergence (set $u=0$). The estimating equeations are independent (Jacobian is diagonal) so this converges very quickly. Otherwise set $u_i = log(x_{i.})$ , $u_j = log(x_{.j})$ and $u = log(x_{..})$.

3. Estimate overdispersions

Theoretically we might estimate the dispersions by ML, but the estimates are very variable, which dramatically affects the rest of the fitting procedure.

 - Maximum likelihood estimation (not used) 
 
$$log(E(x_{ij})) = l_{ij} = offset\big(log(\frac{x_{i.}x_{.j}}{x_{..}}  \big) + \psi_1 (s_{1i}^{init}r_{1j}^{init}) + \psi_2 (s_{2i}^{init}r_{2j}^{init})\big)$$

to obtain estimates for the overdispersions $\theta_j$, whereby the means $\mu_{ij}$ are assumed known

 $$\sum_{i=1}^n \sum^{y_i^*}_{t=0} \big( \frac{t}{1+\frac{t}{\theta_j} } \big) + \theta_j^2 log(1+\frac{\mu_{ij}}{\theta_j})- \mu_{ij}\frac{y_{ij}+\theta_j}{1+\frac{\mu_{ij}}{\theta_j} }$$
 
 For this we can use the theta.ml() function in the _MASS_ package
 
 - edgeR robust
 
To get more stable estimates of the dispersions, RNA-seq packages such as _DESeq_ and _edgeR_ use an empirical Bayes approach to shrink dispersion estimates towards a fitted mean-dispersion trend. Here I use the machinery of edgeR robust (version `r packageVersion("edgeR")`) to estimate the dispersions
 
To reduce the computational cost and because the estimates do not change a lot anyway the estimation of the overdispersions is not repeated in every iteration

4. Estimate the psi parameters
 
 Solve the system of score equations (see Lawless 1987) for $\psi_1$ and $\psi_2$, assuming $\theta_j$'s given
 
 $$\sum_{i=1}^n \sum_{j=1}^p r_{im}s_{mj} \frac{y_{ij}-\mu_{ij}}{1+\frac{\mu_{ij}}{\theta_j}} = 0$$ for all m=1,...,k
 
 with
 
 $$E(X_{ij}) = \mu_{ij} = exp\big(log(\frac{x_{i.}x{.j}}{x_{..}}) + \psi_1^{MLE} s_{1j}^{init}r_{1i}^{init} + \psi_2^{MLE} s_{2j}^{init}r_{2i}^{init} \big)$$.
 
 These are non-linear equations for which we use the _nleqslv_ package, with its default the Broyden method. This is a variation of the Newton-Raphson method (or Fisher scoring in case of ML) whereby the Jacobian is recycled and not recalculated at every iteration. The Jacobian is a square matrix of dimension k*(k-1)/2. The diagonal entry at position $kk'$ is:
 
 $$\frac{\delta^2 L_{NB}(\mu, \theta)}{\delta \psi_k \delta \psi_{k'}} = -\sum_{i=1}^n \sum_{j=1}^p r_{ik}r_{ik'}s_{kj}s_{k'j}\frac{\mu_{ij}(1+\frac{x_{ij}}{\theta_j})}{(1+\frac{\mu_{ij}}{\theta_j})^2}$$
 
5. Estimate the row scores. 

To estimate the $r_{i}$'s we would like to use weighted least squares to minimize

$$\sum_{j=1}^p\Big(log(x_{ij}) - \big(u + u_i + u_j +\sum_{k=1}^M r_{ik} \big( \hat{\psi_ks_{jk}} \big)\Big)^2$$

for every $i$. Unfortunately we cannot calculate $log(x_{ij})$ because of the zeroes.

We'll have to use another way to estimate them, maximum likelihood in our case. I don't like using weighted least squares on the non-transformed counts,

$$\sum_{j=1}^p\Big(x_{ij} - exp\big(u + u_i + u_j +\sum_{k=1}^M r_{ik} \big( \hat{\psi_ks_{jk}} \big)\Big)^2,$$

because of the skewness of the residuals as explained above.

To enforce the constraints on the row scores one option would be to estimate all scores through full maximum likelihood and then modify a few of them to make them satisfy the constraints and repeat until convergence. This runs into numerical problems though, because the modified scores are usually very large initially which leads to overflow when exponenttiated. Instead we use the methods of the Lagrange multipliers to implement the constraints. We thus seek to maximize the Lagrangian function

$$Lag(r_{ik}, \mathbf{\lambda}) = L_{NB}(\mathbf{r}) + \sum_{k=1}^M \big( \lambda_{1k} \sum_{i=1}^n w_i r_{ik} \big) + \sum_{k=1}^M \big( \lambda_{2k} ( \sum_{i=1}^n w_i r_{ik}^2 ) - 1 \big) + \sum_{k \neq k'} \big(\lambda_{3kk'} \sum_{i=1}^n (w_ir_{ik}r_{ik'}) \big)$$

with $L_{NB}(\mathbf{r})$ the log-likelihood function of the negative binomial regression. The derivatives of this function for the following system of equations

$$\frac{\partial Lag(r_{ik}, \mathbf{\lambda})}{\partial r_{ik}} = \sum_{j=1}^p \hat{s}_{jk} \hat{\psi}_k \frac{y_{ij}-\mu_{ij}}{1+\frac{\mu_{ij}}{\theta_j}} + \sum_{k=1}^M w_i \lambda_{1k}  + \sum_{k=1}^M \big(2 \lambda_{2k} w_ir_{ik}\big) +  \sum_{k' \neq k} w_ir_{ki} \lambda_{kk'} = 0$$

and 

$$\frac{\partial Lag(r_{ik}, \mathbf{\lambda})}{\partial \lambda_{1k}} = \sum_{i=1}^n w_i r_{ik} = 0$$

and 

$$\frac{\partial Lag(r_{ik}, \mathbf{\lambda})}{\partial \lambda_{2k}} = (\sum_{i=1}^n w_i r_{ik}^2) - 1 = 0$$

and 

$$\frac{\partial Lag(r_{ik}, \mathbf{\lambda}))}{\partial \lambda_{3kk'}} = (\sum_{i=1}^n w_i r_{ik}r_{ik'}) - 1 = 0$$

. The latter three groups of equations represent the original centering, normalization and orthogonality constraints. 

There are additional equations in the system now, but also as many extra $\lambda$ parameter to optimize. These $\lambda$ parameters are also stored and used as initial values in the next iteration run. There size represent the extent to which the constraints pull away the solution from the optimal, unconstrained solution, but I don't see how we can use that in a biologically or statistically meaningful way.

We now truly have a system of equations to solve, which we do with the _nleqslv_ package. It is straightforward but a bit cumbersome to also provide the analytical Jacobian, which greatly speeds up the iterations (otherwise it has to be approximated numerically). The Jacobian is sparse and symmetric. with following non-zero entries:

 $$\frac{\delta^2 Lag(\mu, \theta)}{ \delta r_{ik} \delta r_{ik'}} = -\sum_{i=1}^n \sum_{j=1}^p \psi_{k}\psi_{k'}s_{kj}s_{k'j}\frac{\mu_{ij}(1+\frac{x_{ij}}{\theta_j})}{(1+\frac{\mu_{ij}}{\theta_j})^2}=0$$
 
  $$\frac{\delta^2 Lag(\mu, \theta)}{\delta r_{ik} \delta \lambda_{1k}} = w_i$$
  
  $$\frac{\delta^2 Lag(\mu, \theta)}{\delta r_{ik} \delta \lambda_{2k}} = 2w_ir_{ik}$$
  
  $$\frac{\delta^2 Lag(\mu, \theta)}{\delta r_{ik} \delta \lambda_{3kk'}} = w_ir_{ik'}$$
  
  All other entries are zero.
  
6. Estimate the column scores

Repeat step 4 but now estimate $s_{jk}$ column scores in the restrictions, e.g. the Lagrangian becomes:

$$Lag(s_{jk}, \mathbf{\lambda}) = L_{NB}(s_{jk}) + \sum_{k=1}^M \big( \lambda_{1k} \sum_{j=1}^p z_j s_{jk} \big) + \sum_{k=1}^M \big( \lambda_{2k}  ( \sum_{j=1}^p z_js_{jk}^2 ) - 1 \big) + \sum_{k \neq k'} \big(\lambda_{3kk'} \sum_{j=1}^p (z_js_{jk}s_{jk'}) - 1\big)$$

#### Termination

7. Repeat steps 3-6 until convergence. Convergence is assumed if between two iterations

 - None of the $\psi$ parameters change less than $0.01\%$ (infinity norm)
 
 AND
 
 - None of the square roots of the L2-norms of the relative changes of row and column scores exceeds 0.1%, or
 
 $$\forall \mathbf{r_k}, \mathbf{s_k} \Big(\sqrt{\sum_{i=1}^n (1-\frac{r^{(iter-1)}_{ik}}{r^{(iter)}_{ik}})^2} < 0.001 \Big) \cup \Big( \sqrt{\sum_{j=1}^p (1-\frac{s^{(iter-1)}_{kj}}{s^{(iter)}_{kj}})^2} < 0.001 \Big)$$
 
 and analogously for all the scores. Other norms, such as L1 or infinity norms can also be used. A norm between 1 and 2 will lead to fast convergence.
 
 Overdispersions are not taken into account in determining the convergence since they're merely nuisance parameters.

#### Biplot

8. Plot $\psi_1r_{1i}$ vs $\psi_2r_{2i}$ as sample points and add $s_{1j}$ vs $s_{2j}$ as arrows to make a biplot. This assures that the orthogonal projection of the vector ($\psi_1r_{1i}$, $\psi_2r_{2i}$) on ($s_{1j}$, $s_{2j}$) equals $cos (\eta) ||(\psi_1r_{1i}, \psi_2r_{2i})|| = \frac{(s_{1j}, s_{2j})^t(\psi_1r_{1i}, \psi_2r_{2i})}{||(s_{1j}, s_{2j})||}$ with $\eta$ the angle between the two vectors. This projection is thus proportional to the departure from independence in the first two dimensions combined. The larger the entries of the species and sample scores (the scaling between these two sets is arbitrary, we usually choose them in the same order of magnitude) and the smaller the angle, the larger the departure of this taxon in this sample. The choice of $\psi_1r_{1i}$ vs $\psi_2r_{2i}$ assures the distances between samples are well represented, since more weight is added to differences in scores in important dimensions.

In the end we'll have estimated  p (abundances) + n (library sizes) + p (dispersions) + kxp (column scores) + kxn (row scores) + k (importance parameters) = (k+2)p + (k+1)xn + k parameters out of np entries. We have imposed 4k + k(k-1) restrictions, so the final model is still very parsimonious for n and p in the hundreds.

##### Model quality

9. Assess the goodness of fit

Since the model is overparametrized, classical ML theory (such as asymptotic behavious of maximum likelihood statistics) does not apply to our solution. Still we can compare the likelihoods of the independence, the RC(2) and the intermediate models to get an idea of the relative importance of the dimensions. The same concept exists for correspondence analysis, where the size of the diagonal elements of $\Sigma$ is proportional to the importance of the corresponding dimension. The log-likelihood of the saturated model is by definition equal to 0.

We can decompose the difference in log-likelihood between the independence and the full RC(K) model as

$$(LL_2 - LL_0) = (LL_1 - LL_0) + (LL_2 - LL_1)$$

with $LL_0$, $LL_1$ and $LL_2$ -2 the log-likelihoods of the independence, RC(1) and RC(2) models respectively. Scaling by $(LL_2 - LL_0)$ will provide interpretable fractions.

### One-by-one (1B1)

Initially all dimension were estimated jointly, but for reasons of speed we moved to a one-by-one approach estimating first the first dimension, then the second given the first, then the third given the first two and so on. By splitting the estimation procedures into smaller parts like this the estimation speeds up considerably, presumably because the systems are easier to solve.

### Confounders (Conditioning)

Sometimes some covariates are known to affect the abundances but one is not interested in their effect. We think first and foremost of technical parameters such as sequencing center or technology. We would like to condition on these parameters and visualize remainign variability.

#### Conditioning in correspondence analysis

Conditioning in correspondence analysis occurs by performing CCA with a confounder matrix Z as covariates, and then perform regular CA on the matrix of residuals.

#### Conditioning in the RCM approach

Thanks to our log-linear approach we can filter out the effect of these parameters by including interaction terms between the taxa and the confounding covariate to be filtered out. This willl occcur after fitting the independence model but before fitting the RC(M) component. We thus fit the following model in case of $c$ confounding variables:

 $$log(E(X_{ij})) = \lambda + \lambda_i + \lambda_j + \textcolor{red}{\sum_{l=1}^c \zeta_{jl}e_{il}} + \sum_{k=1}^M \psi_k r_{ik} s_{jk}$$
 
with $e_{il}$ the value of covariate $l$ in sample $i$ and $\zeta_{jl}$ the interaction parameter between taxon $j$ and confounding variable $l$. Note that we assume that the categorical variables have already been converted to multiple dummy variables.

#### Confounders and zero counts

Filtering out the effect of confounders brings the problems of the zeroes to the fore again. If a taxon has only zero counts in one of the subgroups defined by the confounding variables, the model will fail to fit, since the parameters $\zeta_{jl}$ would need to equal minus infinity. For continuous confounders this is of course not an issue. Maybe there exists a more efficient solution, but for now we will have to trim the taxa for which this occurs from further analysis in order to render the model identifiable. This trimming occurs already before fitting the independence model.

#### Implementation

For the filtering based on the prevalence and abundance, we need a confounder matrix

 1. Without intercept (Overall filtering has happened already)
 2. With __all__ levels of the categorical variables, so set $contrasts=FALSE$ to avoid hidden reference levels of the factors
 
 When actually filtering on the confounders, we can simply use a confounder matrix
 
 1. With intercept
 2. With treatment coding
 
 In this case all that matters is to change the offset formed by the independence model. We will still return the confoundes' parameters though.

### Zero-inflated poisson

We could also augment our model with a zero-inflated Poisson rather than a negative binomial distribution. This opens additional modeling perspectives. We can model the chance on a structural zeroes:

$$logit(P(X_{ij}=0)) = f + f_i + f_j + t_{i1}v_{j1}\chi_1 + t_{i2}v_{j2}\chi_2$$.

The probability of a strucutral zero does not depend on the sample: strucutural zeroes are assumed to be truly absent species and not due to undersampling. All samples come from the same population under $H_0$, and there the structural zeroes are only dependent on the columns. Therefor we set all $f_i=0$. In practice $f$ is not estimated either, instead $f_j$ is estimated without restrictions. The terms $t_{i1}v_{j1}\chi_1 + t_{i2}v_{j2}\chi_2$ indicate in which two independent directions the observed number of zeroes deviates from this base level. I think we need the term $f_i$ because otherwise the first dimension $t_{i1}$ scores will start correcting for it. The independence moel in this case is that the chance on a structural zero only depends on the marginal probabilities. A logit model without an intercept does not make sense, it assumes a baseline probability of structural zero of 0.5.

Fitting the mean part of this models encounters severe numerical problems, this part of the project is not retained and is put in the fridge.

#### Restrictions

The same restrictions apply to the scores $t_i$ and $v_j$ as to $r_i$ and $s_j$

#### Estimation (see Lambert 1992 for details)

The mean and zero probability are modelled independently, absence of a species is independent of its abundance when it is present. Say $Z=0$ when $X$ is from the Poisson state and $Z=1$ when X is a structural zero. Evidently, when $X=0$, Z is unknown.

EM algorithm: iterate between 

 - E: estimate Z, assuming  $\mathbf{r_i}$, $\mathbf{s_j}$ and $\mathbf{\Psi}$ known, through its expectation
 - M: Maximize the log-likelihood given Z
 
Unlike the RC(2)NB model, $f_j$ has to be estimated as well, there is no obvious candidate here. We estimate it marginally and then keep it fixed during the iterations. This way the score visualize departures in zero inflation from this independence of zero-inflation between rows and columns.

Newton-Raphson would be incredibly complicated here, the derivatives of the log-likelihood functions barely fit on a page!

## Constrained analysis

So far we have been doing unconstrained or unsupervised explorative visualization, or _indirect_ gradient analysis. We just visualized the strongest signals in the data, independent of sample covariates. Only then did we use the covariate information to try to correlate the covariates with the latent variables formed by the row and column scores. We usually do this by adding a colour code for the samples in our biplots.

A logical next step is to develop a __constrained__ method in the same framework, i.e. with the same error distribution. We argued that PCoA with Bray-Curtis distance or regular correspondence analysis did not perform well, and we expect the same for canonical correspondence analysis since it relies on the same distributional assumptions. We will visualize those departures from the independence model that can be explained by the sample covariates. Each dimension will thus consist of a linear combination of sample covariates, the weight of each of these covariates in the dimensions reflecting how much they contribute to explaining the variation. For background information see also Zhu _et al._ , 2005, and Zhang and Thas, 2012.

We call our n-by-d covariate matrix $C$, for every sample $i$, $d$ baseline covariates are measured. These covariates can be either discrete or continuous. $c_{iy}$ is the observed value of covariate $y$ in sample $i$, $\mathbf{c_{i.}}$ the whole covariate vector of length $d$ of sample $i$, and $\mathbf{c_{.y}}$ all recorded values of variable $y$.

We could of course just perform our classical, unconstrained methods and then regress the obtained sample scores on the envrionmental variables. However, it is to be expected that this is a less efficient use of the data, which has been shown for the case of correspondence analysis by Prodon and Lebreton (1994).

### Canonical correlation analysis

Canonical correlation analysis seeks to compare multivariate sets of random variables __X__ (n-by-p) and __Y__ (n-by-m) with variables in the columns. This way one discovers which variables correlate best between both data sets. Usually both data sets represent different kinds of measurements, e.g. dietary habits vs. health condition, but all measured on the same individual. It looks for linear transformations __Xa__ and __Yb__ that maximize the correlation $\rho = cor(\mathbf{Xa}, \mathbf{Yb})$. We call these transformations the _canonical covariates_ $\mathbf{Xa_i} = \mathbf{U}_i$ and $\mathbf{Yb_i} = \mathbf{V}_i$ and the corresponding correlation $\rho_i$ the _canonical correlation_. The consecutive canonical covariates are linearly independent, the canonical correlation coefficients decrease over the dimensions of the solution.

__X__ and __Y__ are column centered to obtain __A__ and __B__, and then we look for projecting vectors __x__ and __y__ that maximize correlation between $\mathbf{z}_x = \mathbf{Ax}$ and $\mathbf{z}_y = \mathbf{Ay}$. Note that there is no notion of percentage of variance explained, and that this method is symmetric with respect to datasets __X__ and __Y__.

### Redundancy analysis (RDA)

Redundancy analysis extracts variation between variables of a n-by-p dataset __X__ that can be explained by variables from another n-by-d dataset __Y__. It is a constrained version of principal components analysis (PCA), whereby principal components are restricted to be linear combinations of the set of constraining variables.

### Constrained/canonical correspondence analysis (CCA)

The idea behind canonical or constrained correspondence analysis is to have unimodal response functions as with regular CA, but to restrict the row scores to be linear functions of the environmental variables. Also, it is Gaussian ordination with the scores equally restricted under the same conditions as the unrestricted Gaussian ordination. CCA relates to CA as RDA relates to PCA.

Of course one could perform regular correspondence analysis, and then relate the obtained gradients to known envrionmental variables in a two-step approach. This is again an instance of _indirect_ gradient analysis (see Whittaker 1967).

There are many ways to introduce and explain CCA, with many parallels to other (forgotten) methods, but we will stick to the approach based on the singular value decomposition.

In contrast to redundancy analysis, constrained correspondence analysis (CCA) assumes bell-shaped, unimodal response curves with respect to environmental variables. A __response curve__ describes the expected abundance of a species in function of an environmental score, which is a function fo environmental variables. Redundancy analysis asssumes linear responses to the environmental variables, which is less useful in the ecological context. Classical ecological theory assumes species have an optimal combination of values of the environmental variables, and their expected abundances decrease as the environment departs from these optimal values.

Ter Braak (1987) proposes a method of direct gradient analysis based on CA: canonical correspondence analysis (CCA). It basically assumes that species have a Gaussian response curve in function of an environmental score which is a _linear_ conbination of environmental variables. The linear projection vector $\boldsymbol{\alpha}$ is called the __environmental gradient__. The row scores of the ordination are restricted to be linear combinations of the environmental variables. Prior to the analysis the environmental variables are centered and scaled to avoid arbitrariness in the choice of the units and render the elements of the envrionmental gradient comparable. The estimating equations are:

1) Species scores $\mathbf{s} = \mathbf{J}^{-1}\mathbf{X}^t\mathbf{r}$
2) Sample scores (weighted average taxon scores) $\mathbf{r*} = \mathbf{R}^{-1}\mathbf{Xs}$
3) Environmental gradient $\boldsymbol{\alpha} = (\mathbf{C}^t\mathbf{RC})^{-1}\mathbf{C}^t\mathbf{Rr*}$: weighted regression on environmental variables
4) Sample scores (environmental) $\lambda\mathbf{r} = \mathbf{C}\boldsymbol{\alpha}$

whereby $\lambda$ is a measure of the importance of that dimension. We can write this also as the following estimating equation:

$$(C^TXJ^{-1}X^TC - \boldsymbol{\lambda}C^TRC)\boldsymbol{\alpha} = \mathbf{0}$$

or as 

$$(C^TX(C^TRC)^{-1}X^TC - \boldsymbol{\lambda}J)\mathbf{s} = \mathbf{0}$$

The first equation represents a generalized eigenvalue problem, the second a regular eigenvalue problem since __J__ is a diagonal matrix.

Another approximation to CCA follows more the path we have before with CCA. Hereby the rows of the chi-square transformed matrix $\mathbf{R^{-1/2}(X-E)C^{-1/2}}$ are subjected to weighted least squares regression on the environmental variables, with the sample totals $\mathbf{R}$ as weights. The matrix of environmental variables undergoes weighted centering prior to model fitting, with the library sizes as weights. The row-weight-centered matrix $C'$ is thus calculated as $C' = C(I - \frac{1}{c_{..}d}R\mathds{1}$. Then singular value decomposition is applied to the matrix of fitted values $F = Y (C^T R C)^{-1} C' R R^{-1/2}(X-E)C^{-1/2}$ of this regression, such that $F = U'\Sigma'V'$. Using the fitted values assures that only the systematic part explained by the covariates plays a role. 

A third approach by ter Braak in _"The history of canonical correspondence analysis"_ (2014) uses a direct SVD. Here he also references Goodman's work.

CCA is related to weighted Redundancy analysis (RDA) of the Chi-squared distances, just like regular CA is related to PCoA of Chi-squared distances.

#### CCA triplots

In a triplot of a CCA ordination, the samples and species are plotted as points, and the elements of the environmental gradient as arrows. The species scores are the right singular vectors, the sample scores can be calculated as linear combinations of environmental variables (environmental scores), or as the weighted average of their species scores (weighted by species abundances). In the latter case, the closer the species points lie to to the sample point, the higher the expected abundance. The orthogonal projection of the species point onto the environmental arrow represents how much the expected abundance increases (decreases) with changes in this environmental variable.

#### Conditioning in CCA

Conditioning occurs in the same fashion as with regular CA: the residuals of the CCA on the conditioning matrix Z are the input for the CCA on the environmental variables of interest.

### Redundancy analysis (RDA)

Redundancy analysis (RDA) is a constrained version of PCA, whereby the loadings are constrained to be linear combinations of a set of exploratory variables. This result can be compared to unconstrained PCA to see how much of the total variance is accounted for by the constrained variance. This type of analysis is indicated in case of linear relationships between the response variables (species counts) and explanatory variables. In our case this would e.g. be a environmental gradient in dimension $k$ that is dominated by the concentration of a toxic substance, which only decreases the expected abundance as its concentration increases. Another option is that the gradient is dominated by a variable with a short gradient.

### Constrained RC(M) model

The following approaches to constrained analysis of ecological data exist, or have been considered

#### The classical approach for gradient analysis

This model comes from ecology and is the one used by Zhu _et al._ (2005) and Zhang and Thas (2012 and 2016). This assumes each taxon has an optimal combination of covariates, and its expected abundance decreases as the covariates depart from this optimal combination. The expected abundance follows a Gaussian distribution in function of this departure.

The expected counts are modelled as 

$$log(E(X_{ij})) = \sum_{k=1}^K \big(a_{jk} - \frac{(h_{ik} - q_{jk})^2}{2m_{jk}^2}\big)$$

whereby 

 - $h_{ik} = \boldsymbol{\alpha}_k^t \mathbf{c}_{i.}$ the optimal linear combination of $\mathbf{c}_{i.}$ in dimension $k$, also called the _environmental score_
 - $\boldsymbol{\alpha}_k$ is called the _environmental gradient_
 - $\alpha_{ky}$ is element corresponding to covariate $y$ of the environmental gradient in dimension $k$
 - $a_{jk}$ is the maximal logged expected abundance achieved when the linear combination of covariates $\boldsymbol{\alpha}_k^t \mathbf{c}_{i.}$ equals the optimum $q_{jk}$
 - $m_{jk}$ is the tolerance, it determines how fast the expected value decreases as $\mathbf{c}_{i.}$ moves away from its optimal values $\boldsymbol{\alpha}_k^{-t}h_{ik}$
 
 The scales of $h_{ik}$, $q_{jk}$ and $m_{jk}$ are undefined, so we need an additional restriction to render the model identifiable. Zhu _et al._, 2005 propose setting $\sum_{j=1}^p m_{jk}^2/p=1$. Note that $a_{jk}$, $q_{jk}$ and $m_{jk}$ are species specific parameters whereas $\boldsymbol{\alpha}_k$ is specific for the dimension but common to all species.
 
##### Triplot

- $q_{jk}$ is plotted as points as optimum per species
- The components of $\boldsymbol{\alpha}$ are plotted as arrows for the environmental variables.
- The samples are plotted as their site scores $\boldsymbol{\alpha}_k^t \mathbf{c}_{i.}$
 
##### Estimation

The estimation iterates between estimating $\boldsymbol{\alpha}_k$ and $\boldsymbol{\beta}_{jk} = \big(a_{jk}, q_{jk}, m_{jk})$.

Note that this model is fitted without an intercept

<!---This model was designed with the ecological context in mind, whereby (continuous) environmental variables such as pH, temperature, sunlight intensity, ion concentrations, distance to landscape elements etc. were measured. However for microbiome data, we usually do not have direct measurements of variables in the ecosystem, but rather characterstics of the patient, which are very often categorical variables.

Let's take a look at the nature of the covariate data.

#```{r Nature of the covariates, eval=TRUE, purl=FALSE}
load("/home/stijn/PhD/American Gut/AGphylo.RData")
load("/home/stijn/PhD/Simulations/data/physeqList4Trim.RData")
load("/home/stijn/PhD/Simulations/data/zellerData.RData")
load("/home/stijn/PhD/Biplots/Kostic_data/phyloD.RData")
tableSamVar = function(x){
  table(sapply(sample_data(x), class))
}
phyList = list("American gut"=AGphylo, "Kostic"=phyloD, "HMP" = physeqList4Trim[["Throat"]], "zeller16S" = zellerSphy, "zellerMetagenomics" = zellerMphy)
sapply(phyList, tableSamVar)
#```

Most variables are factors or booleans. Below are the names of the variables shown:

#```{r covariate names, purl=FALSE}
sapply(phyList, sample_variables)
#```

I wonder how realistic the Gaussian response model still is with so many categorical variables and so many species. I guess we would best seek some expert advice of microbiome scientists.

--->

#### RC(M) approach to gradient analysis

#### Constrained RC(M), overparametrized version

If we want to retain analogy with the unconstrained version and model only the departure from independence, we could fit the following model:

$$log(E(X_{ij})) = u + u_i + u_j  + \sum_{k=1}^K \psi_k s_{jk} f_{jk}(\boldsymbol{\alpha}_k^t \mathbf{c}_{i.}) $$

with $f_j$ the species-specific response function, which can be linear, quadratic (Gaussian on count scale), non-parametric, ... . The usual restrictions apply to the column scores $\mathbf{s}$:

$$\mathbf{s^tu_{tax}} = \mathbf{0}$$
$$\mathbf{s_k^tUs_{k'}} = I(k = k')$$

with $u_{tax}$ the vector of $exp(u_j)$'s and  $\mathbf{U}$ a diagonal p-by-p matrix with $u_{tax}$ on the diagonal.

For $\alpha$ (a d-by-k matrix) we enforce the following restrictions:

- Normalization restricts the size, and dimensions are orthogonal (if $k \geq 3$)

$$\boldsymbol{\alpha}^t\boldsymbol{\alpha} = \mathbb{1}_d$$

with $\mathbb{1}_d$ the identity matrix of dimension $d$.

However, this model is overparametrized: $s_{jk}$ and $f_{jk}$ both model the response of a taxon $j$ in the $k$-th dimension.

### Constrained RC(M), identifiable version

A more parsimonious version of this model can be obtained by dropping the column scores.

For any response function with $q$ parameters $\boldsymbol{\beta}_j$ we model the mean as:

$$log(E(X_{ij})) = log(\mu_{ij}) = u_i + u_j + \sum_{k=1}^K \psi_k f_j \big( \boldsymbol{\alpha}_k^t \mathbf{C}_{i}|\boldsymbol{\beta}_j \big) .$$

Basically this represents a marriage of the RC(M) models of Goodman and the gradient models of Zhu and Hastie (2005). For the response function previously a quadratic function (Gaussian on count scale) has been proposed but this may not be realistic. Also the setting is different: we are looking for response function that maximize the __departure from independence__, no longer the abundances themselves! In a first stage we will try fit the linear model with $f(x) = \beta_0 + \beta_1 x$ but later we might estimate $f$ non-parametrically as in Zhu _et al._ (2005). 

Note that the concept of a "row score" is lost here: since every taxon reacts differently to changes in the environment $C_{i.}$. Still this will not yield a perfect fit since $\boldsymbol{\alpha}_k^t$ is only unique to the dimension and the number of parameters $q$ estimated per response function $f_j$ is much smaller than the number of samples $n$.

#### Shape of the response function

Note that our definition of the response function differs from the response function defined by Zhu and Hastie (2005): it no longer models mean abundance but mean departure from independence.

A linear response function may be most appropriate for problems with __short gradients__ i.e. whereby the difference in observed environmental variables is too short to devine more than an increase or decrease in departure from independence. Also in this case it is easy to interpret the effect of each of the environmental variables on the departure from independence. As so often in statistics, the linearity assumption may not be realistic, but renders models that are easy to interpret ("All models are wrong, but some models are useful"). For problems with __long gradients__, i.e. whereby species' departures of independence rise and drop (or drop and rise) within the score of the observed environmental scores, quadratic response functions may be more appropriate. This corresponds e.g. with the scenario whereby a species' abundance does bot depart a lot from independence for extreme values of the environmental score, and but does depart the independence model for an intermediate value of the environmental score, e.g. it thrives in this environment and its abundance is higher than under independence. In essence this is the same as the approach of Zhu and Hastie, only now the baseline is the independence model rather than 0. Every taxon thereby has its own baseline, the response function models departures from this baseline. If the user is unsure and has enough data, he may use non-parametric response funcion. This may improve the sample and covariate ordination, although the result for the species becomes harder to interpret.

##### Relationship to constrained RC(M)

If we fit this method with only one covariate, namely a factor with a unique level for every sample, and all $\beta_{0j}$ (and $\beta_{2j}$) equal to zero , we expect to find the same solution as the unconstrained case. This means we do not constrain the ordination at all in this case and $\beta_{1jk} = s_{jk}$ and $\alpha_{ik} = r_{ik}$.

#### Restricitions

##### Restrictions on the envrionmental gradient

The $\boldsymbol{\alpha}$'s of different dimensions will be constrained as follows:

$$\boldsymbol{\alpha}_k^t\boldsymbol{\alpha}_k = \mathbf{1}$$

$$\boldsymbol{\alpha_k}^t\boldsymbol{\alpha_{k'}} = 0$$

or in shorter notation

$$\boldsymbol{\alpha}^t \boldsymbol{\alpha} = \mathds{1}$$

One restriction I intended to include but which turns out to be a __very bad idea__ is to center the $\alpha$'s:

$$\boldsymbol{\alpha}^t\mathbf{1} = \mathbf{0}$$

This would force the parameter of a single continuous covariate to be 0, the parameters of two continuous variables to be each other's opposite. In short it is not very desirable and is not implemented.

##### Restrictions on parameters of categorical variables in the environmental gradient

For continuous variables the previous discussion suffises, but for categorical variables special attention is required. In a regular, unrestricted setting one would use treatment coding, setting one of the levels of the variable to be the reference level and estimate no parameter for it. However, the above normalization and orthogonality constraints render the result dependent on the choice of reference level, which is of course highly undesirable. Also for quick understanding of the biplot it would be convenient to plot all the values of the categorical variable, and save the reader the trouble of looking up the reference value. Because of these two reasons we will need a coding scheme that includes all parameters.

Say we have $d_{cont}$ continuous covariates and $d_{cat}$ categorical variables with $d = d_{cont} + d_{cat}$, and $m_l$ levels for every categorical variable $l$. Then $\boldsymbol{\alpha}_k$ compromises

$$ q = d_{cont} + \sum_{l=1}^{d_{cat}} (m_l-1) $$
rows, or $q$ parameters to be estimated. Note that there is no intercept needed in this case, intercepts have already be included in the $f()$ function, such that samples with $\alpha_k = \mathbf{0}$ can still have departures from independence.  We can write $\boldsymbol{\alpha}_k$ then as:

$$\boldsymbol{\alpha}_k = (\beta_{cont_1}, ... , \beta_{cont_{d_{cont}}}, \beta_{cat_{11}}, \beta_{cat_{12}}, ..., \beta_{cat_{1(m_1-1)}}, ...,  \beta_{cat_{d_{cat}(m_{d_{cat}}-1)}})_k$$
, i.e. one paramater less than the number of levels. For imposing the restrictions and for plotting we want a representation whereby all covariate levels are represented. We therefore include all parameters, and impose that the parameters sum to 0 within each covariate.

$$\boldsymbol{\alpha}_k' = (\beta_{cont_1}', ... , \beta_{cont_{d_{cont}}}', \beta_{cat_{11}}', \beta_{cat_{12}}', ..., \beta_{cat_{1m_1}}', ...,  \beta_{cat_{d_{cat}m_{d_{cat}}}}')_k$$
Whereby $q' = d_{cont} + \sum_{l=1}^{d_{cat}}(m_l)$, but under the restriction that for all categorical variables $l$

$$\sum_{m=1}^{m_l} \beta_{cat_{lm}k}' = 0$$ 

We again have $g$ degrees of freedom. It is this $\boldsymbol{\alpha}_k'$ that will need to be centered, normalized and orthogonalized _in order to treat each variable level equally_ and _avoid dependence of the outcome on the choice of reference level_. Also this representation will come in handy for plotting, centering the levels of the same covariate around zero. We will call this the __zero-sum__ representation.

##### Restricitions on the parameters of the parametric response functions

The parameters of the response functions are restricted as follows (for linear and quadratic response functions):

$$\boldsymbol{\beta}_1^t\boldsymbol{\beta}_1 = \boldsymbol{\beta}_0^t\boldsymbol{\beta}_0 = 1$$
and for quadratic response functions:

$$\boldsymbol{\beta}_2^t\boldsymbol{\beta}_2 = 1$$

Again, only the $\psi$ parameters can grow in size to reflect the importance of the dimension, other paramters are normalized.

Since $\lambda_j$ already represents the offset for taxon $j$, should we fit the response functions without offsets (all $\beta_{0j} = 0$)? This means that if the environmental scores equals zero, there is no departure from independence. However, this does not give species fully the chance to react in a different way to the environmental gradient. For a certain value, some species might depart from independence while others might not. Therefore we leave $\beta_{0j}$ to be freely estimated, under the variance 1 restriction.

#### Scaling

To render the values of the continuous variables in the environmental gradient comparable, it is clear that they need to be centered and scaled prior to model fitting, as in PCA. This means that their corresponding elements of $\boldsymbol{\alpha}$ represent the contribution to the environmental score of one standard deviation away from the mean of this variable (in case of linear response functions). A perfect quantitative comparison to the magnitude of the values of the paramaters of the dummies of the categorical variables will never be possible. In our case, with 0-1 dummy coding, equal parameters for a dummy and a continuous variable imply that this level of the categorical variable contributes as much to the environmental score as one standard deviation away from the overall mean of the continuous variable.

#### Estimation of the constrained RC(M) model

##### Environmental gradient

$\boldsymbol{\alpha}$ could be estimated by ML as before, but in order to ensure maximum separation of the response functions of all the species we will consider a version of the log-likelihood ratio approach from Zhu _et al._ (2005):

$$LR(\boldsymbol{\alpha}) = log \frac{\prod_{i=1}^n \prod_{j=1}^p \big(p_j^{(\alpha)}(x_{ij};\boldsymbol{\alpha}^T \mathbf{c}_i,\boldsymbol{\beta}_j) \big)}{\prod_{i=1}^n \prod_{j=1}^p \big(p^{(\alpha)}(x_{ij};\boldsymbol{\alpha}^T \mathbf{c}_i,\boldsymbol{\beta}) \big)}$$

with $p^{(\alpha)}$ and $p^{(\alpha)}_j$ estimated probability density functions without and with taxon labels and $\boldsymbol{\beta}_l$ and $\boldsymbol{\beta}$ the parameters of the response functions. $p^{(\alpha)}$ and $p^{(\alpha)}_j$ are calculated under a certain error distribution (e.g. negative binomial). $LR(\boldsymbol{\alpha})$ actually compares two models. In the restricted model each species reacts in the same way to its environment, relative to its own baseline defined by $u_j$, and all do not depart from independence for a certain combination of environmental variables. In the extended model each species is left to react to the environment in its own way. In case of a linear response function it means that it gets a unique intercept and slope.

Naive approach

The naive approach to optimize $LR(\boldsymbol{\alpha})$ would be to use constrained optimization (constraining the $\boldsymbol{\alpha}$'s to have variance one etc.) and estimate $p_j^{(\boldsymbol{\alpha})}$ and $p^{(\boldsymbol{\alpha})}$ in every evaluation of this function. This implies a analytical jacobian cannot be estimated and requires a huge amount of re-estimation of the response functions. This naive approach is computationally unfeasible

Iterative approach

A smarter approach (also the one proposed by Zhu and Hastie, 2005) is to iterate between estimating $\boldsymbol{\alpha}$ by optimizing $LR(\boldsymbol{\alpha})$ given $p_j^{(\boldsymbol{\alpha})}$ and $p^{(\boldsymbol{\alpha})}$, and estimating $p_j^{(\boldsymbol{\alpha})}$ and $p^{(\boldsymbol{\alpha})}$ given $\boldsymbol{\alpha}$. This results in much fewer estimations of $p_j^{(\alpha)}$ and $p^{(\alpha)}$ and allows to optimize $LR(\boldsymbol{\alpha})$ by finding the roots of $\frac{\partial LR(\boldsymbol{\alpha})}{\partial \boldsymbol{\alpha}} = \mathbf{0}$ (and analytically specifying the second order derivative).

Convergence is assumed when every single change in $\alpha$ is below a tolerance level (infinity norm), or when the L2-norm of alpha is below a tolerance level.

Estimating $p^{(\alpha)}_k$ non-parametrically in this iterative algorithm will be computationally more demanding.

###### Lagrangian

Taking into account the restrictions mentioned above, the associated Lagrangian becomes

$$ Lag_{LR}(\boldsymbol{\alpha}_k| \boldsymbol{\beta}_{jk}, \boldsymbol{\beta}_{k}) =  \sum_{i=1}^n  \sum_{j=1}^p \Big(log \big(\frac{f_{NB}(x_{ij}|\psi_k, \mathbf{C}_i, \boldsymbol{\alpha}_k, \phi_j, f_j)}{f_{NB}(x_{ij}| \psi_k, \mathbf{C}_i, \boldsymbol{\alpha}_k, \phi_j, f_0)} \big) \Big) + \lambda_{1k} \boldsymbol{\alpha}_k^t\mathbf{M} + \lambda_{2k} (\boldsymbol{\alpha}_k^t \boldsymbol{\alpha}_k - 1) +  \boldsymbol{\alpha}_k^t \boldsymbol{\alpha}_{-k} \boldsymbol{\lambda}_{orth,k}$$

with $\boldsymbol{\alpha}_{-k}$ the $\boldsymbol{\alpha}$-matrix from the first to the (k-1)th column, and $f_{NB}$ the density function of the negative binomial distribution. $\boldsymbol{\lambda}_{orth,k}$ is a diagonal matrix with largangian multipliers for orthogonality on the diagonal.

###### Score equations

####### Linear response function

The first order derivatives becomes (in case of a linear response function, and remember we set $s_{jk} = 1$)

$$\frac{\partial Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\alpha}_k} = \psi_k \Big( \sum_{i=1}^n  \sum_{j=1}^p  \mathbf{C}_i \beta_{1j} \frac{x_{ij} - \mu_{ij}}{1 + \frac{\mu_{ij}}{\theta_j}} - \sum_{i=1}^n  \sum_{j=1}^p \mathbf{C}_i \beta_{1} \frac{x_{ij} - \mu_{ij0}}{1 + \frac{\mu_{ij0}}{\theta_j}} \Big) + \mathbf{M}\boldsymbol{\lambda}_{1k} + 2\lambda_{2k} \boldsymbol{\alpha}_k +  \boldsymbol{\alpha}_{-k}^t \boldsymbol{\lambda}_{orth,k} = \mathbf{0}$$

with $\mu_{ij0}$ the mean under the null model of equal response functions and $\mathbf{M}$ a q-by-(q-$d_cont$+1) centering matrix. M has ones in the first column, the next columns are all zero except for the entries corresponding to the positons of the dummies of this categorical variable in $\boldsymbol{\alpha}_k$, which are 1. As such, M forces the coefficients of each categorical variable to sum to zero.  In theory we should perhaps reestimate $\theta_j$ on every iteration between the estimation of $\boldsymbol{\alpha}_k$ and $\boldsymbol{\beta}_j$ and $\boldsymbol{\beta}$, and also estimate different overdispersion for the models with equal and species-wise response function, but we think it will complicate the estimation and not make a big difference anyway, so we ignore this for now.

####### Quadratic response function

For a quadratic response function ($f_j(z_i = \mathbf{C}_i\boldsymbol{\alpha}|\boldsymbol{\beta}) = \beta_0 + \beta_1 z_i + \beta_2 z_i^2$) this derivative becomes

$$\frac{\partial Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\alpha}_k} = \psi_k \Big( \sum_{i=1}^n  \sum_{j=1}^p\mathbf{C}_i (\beta_{1j}+2 \beta_{2j}\boldsymbol{\alpha}_k^t\mathbf{C}_i) \frac{x_{ij} - \mu_{ij}}{1 + \frac{\mu_{ij}}{\theta_j}} - \sum_{i=1}^n  \sum_{j=1}^p \mathbf{C}_i (\beta_{1}+2 \beta_{2}\boldsymbol{\alpha}_k^t\mathbf{C}_i) \frac{x_{ij} - \mu_{ij0}}{1 + \frac{\mu_{ij0}}{\theta_j}} \Big) + \mathbf{M}\boldsymbol{\lambda}_{1k} + \\ 2\lambda_{2k} \boldsymbol{\alpha}_k + \boldsymbol{\alpha}_{-k}^t \boldsymbol{\lambda}_{orth,k} = \mathbf{0}$$

For a non-parametric density function this derivative is not available, which which greatly increase the optimization time.

The other derivatives are (independent of the shape of the response function):

$$\frac{\partial Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\lambda_{1k}}} = \boldsymbol{\alpha}_k^t \mathbf{M} = \mathbf{0}$$
$$\frac{\partial Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \lambda_{2k}} = \boldsymbol{\alpha}_k^t \boldsymbol{\alpha}_k - 1 = 0$$
$$\frac{\partial Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\lambda_{3k}}} = \boldsymbol{\alpha}_k^t \boldsymbol{\alpha}_{-k} = \mathbf{0}$$

###### Jacobians

####### Linear response function

The second order derivatives are for the linear response function

$$\frac{\partial^2 Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\alpha}_k^2} = - {\psi_k}^2 \Big( \sum_{i=1}^n  \sum_{j=1}^p (\mathbf{C}_i \beta_{1j})^2 \frac{\mu_{ij}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij}}{\theta_j})^2} - \sum_{i=1}^n  \sum_{j=1}^p (\mathbf{C}_i \beta_{1})^2 \frac{\mu_{ij0}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij0}}{\theta_j})^2} \Big) + 2\lambda_{2k} $$

$$\frac{\partial^2 Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial {\alpha}_{yk}{\alpha}_{y'k}} = - {\psi_k}^2 \Big( \sum_{i=1}^n  \sum_{j=1}^p  \beta_{1j}^2 c_{iy} c_{iy'}\frac{\mu_{ij}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij}}{\theta_j})^2} - \sum_{i=1}^n  \sum_{j=1}^p \beta_{1}^2 c_{iy} c_{iy'} \frac{\mu_{ij0}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij0}}{\theta_j})^2} \Big) $$

####### Quadratic response function

For the quadratic response function these second order derivatives are:

$$\frac{\partial^2 Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial {\alpha}_{yk}{\alpha}_{y'k}} = - {\psi_k}^2 \Big[ \sum_{i=1}^n \sum_{j=1}^p \Big((\beta_{1j} + 2\beta_{2j}\boldsymbol{\alpha}_k^t\mathbf{C}_i)(\beta_{1j} + 2\beta_{2j}\boldsymbol{\alpha}_k^t\mathbf{C}_i) c_{iy}^2\frac{\mu_{ij}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij}}{\theta_j})^2}   \\ 
- (\beta_{1} + 2\beta_{2}\boldsymbol{\alpha}_k^t\mathbf{C}_i)(\beta_{1} + 2\beta_{2}\boldsymbol{\alpha}_k^t\mathbf{C}_i) c_{iy}^2\frac{\mu_{ij}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij}}{\theta_j})^2}  \Big) \Big] - 2\psi_k \Big[\sum_{i=1}^n \sum_{j=1}^p c_{iy}^2 \big(\beta_{2j} \frac{x_{ij} - \mu_{ij}}{1 + \frac{\mu_{ij}}{\theta_j}}-\beta_{2} \frac{x_{ij} - \mu_{ij0}}{1 + \frac{\mu_{ij0}}{\theta_j}}\big)\Big] + 2\lambda_{2k} $$
$$\frac{\partial^2 Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial {\alpha}_{yk}{\alpha}_{y'k}} = - {\psi_k}^2 \Big[ \sum_{i=1}^n \sum_{j=1}^p \Big((\beta_{1j} + 2\beta_{2j}\boldsymbol{\alpha}_k^t\mathbf{C}_i)(\beta_{1j} + 2\beta_{2j}\boldsymbol{\alpha}_k^t\mathbf{C}_i) c_{iy} c_{iy'}\frac{\mu_{ij}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij}}{\theta_j})^2}   \\ 
- (\beta_{1} + 2\beta_{2}\boldsymbol{\alpha}_k^t\mathbf{C}_i)(\beta_{1} + 2\beta_{2}\boldsymbol{\alpha}_k^t\mathbf{C}_i) c_{iy} c_{iy'}\frac{\mu_{ij}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij}}{\theta_j})^2}  \Big) \Big] - 2\psi_k \Big[\sum_{i=1}^n \sum_{j=1}^p c_{iy}c_{iy'} \big(\beta_{2j} \frac{x_{ij} - \mu_{ij}}{1 + \frac{\mu_{ij}}{\theta_j}}-\beta_{2} \frac{x_{ij} - \mu_{ij0}}{1 + \frac{\mu_{ij0}}{\theta_j}}\big)\Big] $$

<!---+ \\ 2\beta_{2j} c_{iy} c_{iy'}\frac{x_{ij} - \mu_{ij}}{1 + \frac{\mu_{ij}}{\theta_j}}--->

The other derivatives are again independent of the response function:

$$\frac{\partial^2 Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\alpha}_k \partial \boldsymbol{\lambda_{1k}}} =  \mathbf{M}$$

$$\frac{\partial^2 Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\alpha}_k \partial \lambda_{2k}} =  2\boldsymbol{\alpha}_k$$

$$\frac{\partial^2 Lag_{LR}(\boldsymbol{\alpha}_k)}{\partial \boldsymbol{\alpha}_{k} \partial \boldsymbol{\lambda}_{orth,k}} =  \boldsymbol{\alpha}_{-k}$$
All other second order derivatives equal zero. 

###### Non-parametric response function

The response function $f(\boldsymbol{\alpha}^t\mathbf{C})$ need not have a parametric form. We can also estimate it non-parametrically as in Zhu and Hastie (2005). Like these authross, we use the locfit library. We regress $\frac{X}{exp(\lambda_i+\lambda_j)}$ on $\boldsymbol{\alpha}^t\mathbf{C}$ with a log-link. We limit the surface between this curve and the x-axis to sum to one as a means of normalization, only the importance parameter $\psi$ can differ in size between the different dimensions.

##### Estimation of the parameters of the parametric response functions

We fit all parameters of the different taxa jointly, and enforce normalization through Lagrange multipliers again.

The derivative of the Lagrangian for $\beta_{jv}$ under the negative binomial model with $v$ a parameter index running along $\boldsymbol{\beta}_j$ (for k=1 and omitting dimension indices)

$$ \sum_{i=1}^n \frac{\partial Lag_{NB}(X_{.j}|\boldsymbol{\beta}_j, \theta_j)}{\partial \beta_{jv}} = 2\beta_{jv}\lambda_{2v} +  \sum_{i=1}^n \psi \frac{\partial f_j \big( \boldsymbol{\alpha}^t \mathbf{C}_{i.}|\boldsymbol{\beta}_j \big)}{\partial \beta_{jv}} \frac{x_{ij}-\mu_{ij}}{1+\frac{\mu_{ij}}{\theta_j}} = 0$$

The second order derivative for parameters from the same taxon is:

$$\sum_{i=1}^n \frac{\partial^2 Lag_{NB}(X_{.j}|\boldsymbol{\beta}_j, \theta_j)}{\partial \beta_{jv} \partial \beta_{jv'}} = -\psi^2 \sum_{i=1}^n \Big( \frac{\mu_{ij}(1+ \frac{x_{ij}}{\theta_j})}{(1 + \frac{\mu_{ij}}{\theta_j})^2} \frac{\partial f_j \big( \boldsymbol{\alpha}^t \mathbf{C}_{i.}|\boldsymbol{\beta}_j \big)}{\partial \beta_{jv}} \frac{\partial f_j \big( \boldsymbol{\alpha}^t \mathbf{C}_{i.}|\boldsymbol{\beta}_j \big)}{\partial \beta_{jv'}} \Big) + 2\lambda_{2v} \mathds{I}(v=v')$$

$$\sum_{i=1}^n \frac{\partial^2 Lag_{NB}(X|\boldsymbol{\beta}_j, \theta_j)}{\partial \beta_{jv} \partial \lambda_{2v'}} = 2\beta_{2v} \mathds{I}(v=v')$$

all other second order derivatives equal zero.

#### Algorithm

The whole estimation of the constrained RC(M)-model would then proceed as follows:

 1. Estimate the independence model ($u_i$'s, and $u_j$'s)
 2. (Optional): Filter out confounders by modifying offset
 3. Find starting values for $\alpha$, based on CCA
 4. Iterate between
  - Estimate $f_j$'s and $f_0$ given $\boldsymbol{\alpha}$
  - Find $\boldsymbol{\alpha}$ by optimizing $LR(\boldsymbol{\alpha})$
 5. Estimate overdispersions using empirical Bayes
 6. Estimate importance parameter $\psi_k$
 7. Repeat steps 4-6 for every dimension $k$ until convergence

Filtering of rare taxa is not needed in this case, as all covariates are summarized in one "continuous" row score. to check for convergencen, for $\psi_k$  we apply an infinity norm, for $\boldsymbol{\alpha}_k$ the default is an L2-norm.

#### Triplot

The interpretation of the triplot is different from the one from Zhu _et al._ (2005), and also different from that of the unconstrained RC(M).

 - The components of $\boldsymbol{\alpha}_k$ are plotted as arrows and represent the extent to which the covariates contribute to environmental score in dimension $k$. Categorical variables are centered, all their levels are shown. The continuous variables have been centered and scaled, so the magnitude of their compponents in $\boldsymbol{\alpha}_k$ can be compared to one another.
 - For the samples, the environmental scores are plotted as points and represent environmental scores of each sample. Samples close together in one dimension come from similar environments, considering that environmental gradient. This does _NOT_ mean that they have similar values of the environmental variables. 
 - The plot for the taxa depends on the response function used. 
    * In case of a linear response function we can plot arrows originating from $-\frac{\beta_{0j}}{\beta_{1j}}$, which represents the point of no departure from independence explained by the covariates. The direction and size of the arrow then depends purely on $\beta_{1j}\psi$. We can make two sorts of projections:
      - The projection of ($-\frac{\beta_{0j}}{\beta_{1j}}$, $-\frac{\beta_{0j}}{\beta_{1j}} + \beta_{1j}\psi$) onto $\boldsymbol{\alpha}^t\mathbf{C}_i$ (the environmental scores), then corresponds to the departure from independence for taxon $j$ in sample $i$ explained by its environmental variables $\mathbf{C}_i$. The arrow is appropriate since a larger environmental score always results in a larger departure from indepedence under this model. As often in statistics, linearity may not be a very credible assumption, but it leads to models that are easy to interpret.
      - The projection of ($-\frac{\beta_{0j}}{\beta_{1j}}$, $-\frac{\beta_{0j}}{\beta_{1j}} + \beta_{1j}\psi$) onto $(0,\alpha_y)$ reflects the sensitivity of taxon $j$ to changes in covariate $y$.
      We cannot make a direct connection between components of the environmental gradient $\boldsymbol{\alpha}$ and the observed environmental scores $\boldsymbol{\alpha}^t\mathbf{C}_i$, even though this may seem so intuitively. Often it is so that a large $\alpha_y$ points in the direction of samples with a large $c_{iy}$, but this is not guaranteed. Many different combinations of covariates $\mathbf{C}_{i}$ can lead to the same environmental score for a given $\boldsymbol{\alpha}$. This is because of the dimension reduction from $d$ to 1 resulting from the projection by $\boldsymbol{\alpha}$. We will have to warn the user for this pitfall, and facilitate the depiction of the environmental variables of each samples through colour codes.
    * In case of quadratic response function we should plot the extremum of the response functions in each dimension as a point (taking the importance parameter $\psi_k$ into account, which is multiplied by each of the parameters $\beta$ of the response functions). If this is a minimum, this represents the value of the environmental score with the largest drop in expected abundance compared to the independence model, in case of a maximum the largest rise. With colour and shape codes we can show which kind of extremum it is, and which kind of departure from indepedence (negative or positive effect on the expectation). To represent the tolerances, i.e. how steep these valleys and mountains of expected abundance are, we can use ellipses to designate e.g. points with 50% the departure from independence from the top. For CCA this feature is not needed since there all tolerances are assumed equal.
    * In case of non-parametric response function I do not see I clear way to plot the species. The benefit of this method will have to be the improved ordination of the environmental gradient and the samples. One can always look at the individual species response functions if one is interested.
  
It is clear that the triplot is not trivial to interpret, but then again: neither is a triplot based on CCA.

## Some general thoughts on the RC(M)

Contrary to most other methods to make biplots for microbiome data, the RC(M)-method is a truly statistical method that explicitly models the mean. This renders it very easy to interpret. It's main weaknesses are the reliance on a parametric distribution (which may be incorrect), and it's relatively long running times. However, many other biplot methods, such as CA ot PCoA with Hellinger distances, also make distributional assumptions, mostly without stating them explicitly.

\newpage

# Implementation

## Correspondence analysis

```{r Auxfuns Correspondence analysis, purl=TRUE}
## A function to perform correspondence analysis
# The unconstrained approach is the approximation to the Pearson's chi-squared, as defined by Gower _et al._ in their book "Understanding Biplots". The constrained approach is the one implemented in vegan, based on a weighted regression of the departure matrix on the environmental scores. We cannot reproduce its results by the code given by ter Braak based on eigenvlaue or singular value decompositions.
caSVD = function(X, Y = NULL, Z = NULL){
# @param X: the nxp count matrix
# @param Y(optional): a nxd matrix of covariates, or a vector of variable names if X is a phyloseq object. If null regular CA is performed
# @param Z(optional): a nxz matrix of variables to condition on, or a vector of variable names if X is a phyloseq object.
  
# @return: the singular value decomposition of the matrix of pearson residuals
  if(class(X)=="phyloseq"){
  if(!is.null(Y)){Y = model.matrix(data = data.frame(sample_data(object = X))[,Y], ~.)}
  if(!is.null(Z)){Z = model.matrix(data = sample_data(object = X)[,Z], ~.)} 
  X = if (taxa_are_rows(X)) t(otu_table(X)@.Data) else otu_table(X)@.Data
  }
  C = colSums(X)
  R = rowSums(X)
  E = outer(R,C)/sum(X) #Expected counts under independence
  if(is.null(Y) & is.null(Z)){ #Unconstrained analysis
  Goal = diag(1/sqrt(R)) %*% (X-E) %*% diag(1/sqrt(C))
  dimnames(Goal) = dimnames(X)
  svd(Goal)
  } else { #Constrained analysis
  vegan:::cca(X = X, Y=Y, Z=Z)
  # Ycent = t(t(Y) - rowMeans(t(Y)*R/sum(R) ))#Weighted centering  
  # Goal = Ycent %*% solve(a =  crossprod(Ycent,  diag(R)) %*% Ycent,
  #       b = crossprod(Ycent, diag(R) %*% diag(1/sqrt(R))) %*% (X-E) %*% diag(1/sqrt(C))) 
  # #The goal matrix consists of fitted values of weighted regression of the chi-square transformed data matrix on the covariates with row sums as weights
  # dimnames(Goal) = dimnames(X)
  # Svd = svd(Goal)
  # envScores = 1/sqrt(colSums(Ycent^2))*crossprod(Ycent, Svd$u)
  # 
  # # #Eigenvalue approach (ter Braak 1986)
  # EigGoal = t(X) %*% Y %*% solve(t(Y) %*% diag(R) %*% Y) %*% t(Y) %*% X
  # EigDec = eigen(EigGoal)
  # EigSitesWeighted = diag(1/R) %*% X %*% EigDec$vectors %*% diag(1/C)
  # alpha = solve(t(Y) %*% diag(R) %*% Y) %*% t(Y) %*% diag(R) %*% EigSitesWeighted
  # EigSitesConstr = Y %*% alpha
  # 
  # GeigDec = geigen(t(Y) %*% X %*% diag(1/C) %*% t(X) %*% Y, t(Y) %*% diag(R) %*% Y)
  # 
  # # Singular value approach (ter Braak 2014)
  # Ystar = sqrtm(solve(t(Y) %*% diag(R) %*% Y)) %*% t(Y) %*% X %*% diag(1/sqrt(C))
  # svdY = svd(Ystar)
  # B = sqrtm(solve(t(Y) %*% diag(R) %*% Y)) %*% svdY$u
  # U = diag(1/sqrt(C)) %*% svdY$v %*% diag(svdY$d)
  # sites = Y %*% B
  # 
  # # At least 
  # cor(B[,1:2], alpha[,1:2])
  # cor(sites[,1:2], EigSitesConstr[,1:2])
  # 
  # # Vegan
  # ccaVegan = cca(X, Y)$CCA
  # 
  # par(mfrow = c(2,2))
  # #Species plots
  # tmp = plot(cca(X, Y),display = "sp")
  # plot(U)
  # plot(EigDec$vectors)
  # plot(Svd$v)
  # cor(cbind(U[,1], EigDec$vectors[,1], Svd$v[,1], tmp[[1]][,1] ))
  # 
  # par(mfrow = c(2,2))
  # # Sample plots constrained
  # tmp = plot(cca(X, Y),display = "lc", type = "points")
  # plot(sites)
  # plot(EigSitesConstr)
  # plot(Svd$u)
  # cor(cbind(sites[,1], EigSitesConstr[,1], Svd$u[,1], tmp[[1]][,1] ))
  # 
  # # Ter Braaks eigen- and singular value decompositions are the same, but differ from the chi-square difference approach?
  }
}
#-------------------------------------------------#
cca2 = function (X, Y, Z, ...) 
{
  ZERO <- 1e-04
  CCA <- NULL
  pCCA <- NULL
  CA <- NULL
  weight.centre <- function(x, w) {
    w.c <- apply(x, 2, weighted.mean, w = w)
    x <- sweep(x, 2, w.c, "-")
    x <- sweep(x, 1, sqrt(w), "*")
    attr(x, "centre") <- w.c
    x
  }
  if (inherits(X, "dist") || NCOL(X) == NROW(X) && isTRUE(all.equal(X, 
    t(X)))) 
    stop("function cannot be used with (dis)similarities")
  X <- as.matrix(X)
  if (any(rowSums(X) <= 0)) 
    stop("All row sums must be >0 in the community data matrix")
  if (any(tmp <- colSums(X) <= 0)) {
    exclude.spec <- seq(along = tmp)[tmp]
    names(exclude.spec) <- colnames(X)[tmp]
    class(exclude.spec) <- "exclude"
    X <- X[, !tmp, drop = FALSE]
  }
  gran.tot <- sum(X)
  X <- X/gran.tot
  rowsum <- apply(X, 1, sum)
  colsum <- apply(X, 2, sum)
  rc <- outer(rowsum, colsum)
  Xbar <- (X - rc)/sqrt(rc)
  tot.chi <- sum(svd(Xbar, nu = 0, nv = 0)$d^2)
  if (!missing(Z) && !is.null(Z)) {
    Z <- as.matrix(Z)
    Z.r <- weight.centre(Z, rowsum)
    Q <- qr(Z.r)
    Z <- qr.fitted(Q, Xbar)
    tmp <- sum(svd(Z, nu = 0, nv = 0)$d^2)
    if (Q$rank) {
      pCCA <- list(rank = Q$rank, tot.chi = tmp, QR = Q, 
        Fit = Z, envcentre = attr(Z.r, "centre"))
      Xbar <- qr.resid(Q, Xbar)
    }
    if (tmp < ZERO) 
      pCCA$tot.chi <- 0
  }
  else Z.r <- NULL
  if (!missing(Y) && !is.null(Y)) {
    Y <- as.matrix(Y)
    Y.r <- weight.centre(Y, rowsum)
    Q <- qr(cbind(Z.r, Y.r), tol = ZERO)
    if (is.null(pCCA)) 
      rank <- Q$rank
    else rank <- Q$rank - pCCA$rank
    qrank <- rank
    Y <- qr.fitted(Q, Xbar)
    sol <- svd(Y)
    rank <- min(rank, sum(sol$d > ZERO))
    ax.names <- paste("CCA", seq_along(sol$d), sep = "")
    colnames(sol$u) <- ax.names
    colnames(sol$v) <- ax.names
    names(sol$d) <- ax.names
    rownames(sol$u) <- rownames(X)
    rownames(sol$v) <- colnames(X)
    if (rank) {
      CCA <- list(eig = sol$d[1:rank]^2)
      CCA$u <- sweep(as.matrix(sol$u[, 1:rank, drop = FALSE]), 
        1, 1/sqrt(rowsum), "*")
      CCA$v <- sweep(as.matrix(sol$v[, 1:rank, drop = FALSE]), 
        1, 1/sqrt(colsum), "*")
      wa.eig <- sweep(Xbar %*% sol$v[, 1:rank, drop = FALSE], 
        1, 1/sqrt(rowsum), "*")
      CCA$wa <- sweep(wa.eig, 2, 1/sol$d[1:rank], "*")
      oo <- Q$pivot
      if (!is.null(pCCA$rank)) 
        oo <- oo[-(1:pCCA$rank)] - ncol(Z.r)
      oo <- oo[1:qrank]
      if (length(oo) < ncol(Y.r)) 
        CCA$alias <- colnames(Y.r)[-oo]
      xx <- Y.r[, oo, drop = FALSE]
      CCA$biplot <- (1/sqrt(colSums(xx^2))) * crossprod(xx, 
        sol$u[, seq_len(rank), drop = FALSE]) #Standardize variables
      CCA$rank <- rank
      CCA$qrank <- qrank
      CCA$tot.chi <- sum(CCA$eig)
      CCA$QR <- Q
      CCA$envcentre <- attr(Y.r, "centre")
      CCA$Xbar <- Xbar
    }
    else {
      CCA <- list(eig = 0, rank = rank, qrank = qrank, 
        tot.chi = 0, QR = Q, Xbar = Xbar)
      u <- matrix(0, nrow = nrow(sol$u), ncol = 0)
      v <- matrix(0, nrow = nrow(sol$v), ncol = 0)
      CCA$u <- CCA$wa <- u
      CCA$v <- v
      CCA$biplot <- matrix(0, 0, 0)
      CCA$alias <- colnames(Y.r)
    }
    Xbar <- qr.resid(Q, Xbar)
    if (exists("exclude.spec")) {
      attr(CCA$v, "na.action") <- exclude.spec
    }
  }
  Q <- qr(Xbar)
  sol <- svd(Xbar)
  ax.names <- paste("CA", seq_along(sol$d), sep = "")
  colnames(sol$u) <- ax.names
  colnames(sol$v) <- ax.names
  names(sol$d) <- ax.names
  rownames(sol$u) <- rownames(X)
  rownames(sol$v) <- colnames(X)
  rank <- min(Q$rank, sum(sol$d > ZERO))
  if (rank) {
    CA <- list(eig = sol$d[seq_len(rank)]^2)
    CA$u <- sweep(as.matrix(sol$u[, seq_len(rank), drop = FALSE]), 
      1, 1/sqrt(rowsum), "*")
    CA$v <- sweep(as.matrix(sol$v[, seq_len(rank), drop = FALSE]), 
      1, 1/sqrt(colsum), "*")
    CA$rank <- rank
    CA$tot.chi <- sum(CA$eig)
    CA$Xbar <- Xbar
  }
  else {
    CA <- list(eig = 0, rank = rank, tot.chi = 0, Xbar = Xbar)
    CA$u <- matrix(0, nrow(sol$u), 0)
    CA$v <- matrix(0, nrow(sol$v), 0)
  }
  if (exists("exclude.spec")) {
    attr(CA$v, "na.action") <- exclude.spec
  }
  call <- match.call()
  call[[1]] <- as.name("cca")
  if (!is.null(pCCA) && pCCA$tot.chi == 0) 
    pCCA$rank <- 0
  sol <- list(call = call, grand.total = gran.tot, rowsum = rowsum, 
    colsum = colsum, tot.chi = tot.chi, pCCA = pCCA, CCA = CCA, 
    CA = CA)
  sol$method <- "cca"
  sol$inertia <- "mean squared contingency coefficient"
  class(sol) <- "cca"
  sol
}
#Our solution equals the plot.cca(display = "lc" option)

anacor2 = function (tab, ndim = 2, row.covariates, col.covariates, scaling = c("Benzecri", 
    "Benzecri"), ellipse = FALSE, eps = 1e-06) 
{
    tab <- as.matrix(tab)
    if (length(scaling) == 1) 
        scaling <- rep(scaling, 2)
    scaling[1] <- match.arg(scaling[1], c("standard", "Benzecri", 
        "Goodman"))
    scaling[2] <- match.arg(scaling[2], c("standard", "Benzecri", 
        "Goodman"))
    if (missing(row.covariates)) {
        row.covariates <- NULL
    }
    else {
        row.covariates <- as.matrix(row.covariates)
        if (nrow(row.covariates) != nrow(tab)) 
            stop("Matrix with row covariates does not match table dimensions!")
    }
    if (missing(col.covariates)) {
        col.covariates <- NULL
    }
    else {
        col.covariates <- as.matrix(col.covariates)
        if (nrow(col.covariates) != ncol(tab)) 
            stop("Matrix with column covariates does not match table dimensions!")
    }
    if (ndim > min(dim(tab)) - 1) 
        stop("Too many dimensions!")
    name <- deparse(substitute(tab))
    if (any(is.na(tab))) 
        tab <- reconstitute(tab, eps = eps)
    n <- dim(tab)[1]
    m <- dim(tab)[2]
    N <- sum(tab)
    tab <- as.matrix(tab)
    prop <- as.vector(t(tab))/N
    r <- rowSums(tab)
    c <- colSums(tab)
    qdim <- ndim + 1
    r <- ifelse(r == 0, 1, r)
    c <- ifelse(c == 0, 1, c)
    ROW <- FALSE
    COL <- FALSE
    z <- tab/sqrt(outer(r, c))
    chisq <- N * (sum(z^2) - 1)
    if (is.matrix(row.covariates)) {
        xr <- cbind(rep(1, n), row.covariates)
        nr <- dim(xr)[2]
        cx <- crossprod(xr, r * xr)
        or <- symSqrt(cx, inv = TRUE)
        ROW <- TRUE
    }
    if (is.matrix(col.covariates)) {
        yr <- cbind(rep(1, m), col.covariates)
        mr <- dim(yr)[2]
        cy <- crossprod(yr, c * yr)
        oc <- symSqrt(cy, inv = TRUE)
        COL <- TRUE
    }
    if (ROW) {
        z <- or %*% crossprod(xr, tab)
    }
    else {
        z <- tab/sqrt(r)
    }
    if (COL) {
        z <- z %*% yr %*% oc
    }
    else {
        z <- z/outer(rep(1, dim(z)[1]), sqrt(c))
    }
    sv <- svd(z, nu = qdim, nv = qdim)
    sval <- N * ((sv$d[-1])^2)
    sigmavec <- (sv$d)[2:qdim]
    dimlab <- paste("D", 1:ndim, sep = "")
    Tr <- Xstar <- isetCorRow <- NULL
    Tc <- Ystar <- isetCorCol <- NULL
    if (ROW) {
        x <- (xr %*% or %*% (sv$u))[, -1]
        U <- or %*% (sv$u)[, -1]
        Tr <- U %*% (diag(sigmavec))
        dimnames(Tr) <- list(paste("beta", 1:(dim(Tr)[1]), sep = ""), 
            dimlab)
        if (COL) 
            y <- (yr %*% oc %*% (sv$v))[, -1]
        else y <- ((sv$v)/sqrt(c))[, -1]
        Xstar <- (diag(1/r) %*% tab %*% y)
        dimnames(Xstar) <- list(rownames(tab), dimlab)
        isetCorRow <- cor(Xstar, row.covariates)
    }
    else {
        x <- ((sv$u)/sqrt(r))[, -1]
    }
    if (COL) {
        y <- (yr %*% oc %*% (sv$v))[, -1]
        V <- oc %*% (sv$v)[, -1]
        Tc <- V %*% (diag(sigmavec))
        dimnames(Tc) <- list(paste("beta", 1:(dim(Tc)[1]), sep = ""), 
            dimlab)
        Ystar <- (diag(1/c) %*% t(tab) %*% x)
        dimnames(Ystar) <- list(colnames(tab), dimlab)
        isetCorCol <- cor(Ystar, col.covariates)
    }
    else {
        y <- ((sv$v)/sqrt(c))[, -1]
    }
    x <- x * sqrt(N)
    y <- y * sqrt(N)
    if (scaling[2] == "Goodman") 
        y <- y * outer(rep(1, m), sqrt(sigmavec))
    if (scaling[1] == "Goodman") 
        x <- x * outer(rep(1, n), sqrt(sigmavec))
    if (scaling[2] == "Benzecri") 
        y <- y * outer(rep(1, m), sigmavec)
    if (scaling[1] == "Benzecri") 
        x <- x * outer(rep(1, n), sigmavec)
    benzres <- benzdist(scaling, x, y, z, tab, n, m, r, c, row.covariates, 
        col.covariates)
    prob <- sval/chisq
    pcum <- cumsum(prob)
    cs.mat <- cbind(sval, prob, pcum)
    eigenall <- sval/N
    if (ellipse) {
        res.deriv <- gsvdDer(tab, ndim)
        res.deriv.scaling <- gsvdScal(res.deriv, scaling)
        res.acov <- acovuv(res.deriv.scaling, prop, n, m, N, 
            ndim)
        se.sigma <- sqrt(diag(res.acov$acovd))
    }
    else {
        res.acov <- NULL
        se.sigma <- NULL
    }
    colnames(x) <- colnames(y) <- dimlab
    rownames(x) <- rownames(tab)
    rownames(y) <- colnames(tab)
    rownames(cs.mat) <- paste("Dimension", 1:dim(cs.mat)[1])
    colnames(cs.mat) <- c("Chisq", "Proportion", "Cumulative Proportion")
    if (!is.null(se.sigma)) {
        z.values <- sigmavec/se.sigma
        pvalues <- (1 - pnorm(abs(z.values)))
        stestmat <- data.frame(sigmavec, se.sigma, pvalues)
        colnames(stestmat) <- c("Singular Value", "Standard Error", 
            "p-value")
        rownames(stestmat) <- paste("Dimension", 1:ndim)
    }
    else {
        stestmat <- NULL
    }
    result <- list(datname = name, tab = tab, ndim = ndim, row.covariates = row.covariates, 
        col.covariates = col.covariates, row.scores = x, col.scores = y, 
        chisq.decomp = cs.mat, chisq = chisq, singular.values = sigmavec, 
        se.singular.values = se.sigma, stestmat = stestmat, left.singvec = sv$u[, 
            -1], right.singvec = sv$v[, -1], eigen.values = sigmavec^2, 
        eigenall = eigenall, scaling = scaling, bdmat = benzres[1:4], 
        rmse = benzres[5:6], row.acov = res.acov$acovu, col.acov = res.acov$acovv, 
        cancoef = list(rows = Tr, columns = Tc), sitescores = list(rows = Xstar, 
            columns = Ystar), isetcor = list(rows = isetCorRow, 
            columns = isetCorCol), call = match.call())
    class(result) <- "anacor"
    result
}
```

## Unconstrained RC(M)

### Negative binomial

```{r NBauxFuns, purl=TRUE}
#-------------------------------------#
#A score function for the row components of the independence model (library sizes (u_i))
dNBlibSizes = function(beta, X, reg, thetas){
  mu = exp(outer(beta,reg, "+"))
  rowSums((X-mu)/(1+(mu/thetas)))
}
#------------------------------------#
# The corresponding Jacobian
NBjacobianLibSizes = function(beta, X, reg, thetas){
       mu = exp(outer(beta,reg, "+"))
  diag(-rowSums((1+(X/thetas))*mu/(1+(mu/thetas))^2))
}
#-------------------------------------#
#A score function for the column components of the independence model (mean relative abundances (u_j))
dNBabunds = function(beta, X, reg, thetas){
    mu = exp(outer(reg,beta, "+"))
  score = colSums((X-mu)/(1+(mu/thetas)))
}
#------------------------------------#
# The corresponding Jacobian
NBjacobianAbunds = function(beta, X, reg, thetas){
    mu = exp(outer(reg,beta, "+"))
  -diag(colSums((1+(X/thetas))*mu/(1+(mu/thetas))^2))
}

#------------------------------#
#A small aux function for the length of the lambdas
seq_k = function(y){
  (y-1)*(2+(y-2)/2) + seq_len(y+1)
}

#--------------------------------------#
# A score function for the psis
dNBpsis = function(beta, X, reg, theta,  muMarg){
  # @param beta: the current guess for the psi parameter, a scalar
  # @param X: the nxp data matrix
  # @param muMarg: the nxp offset
  # @param reg: the regressor matrix, the outer product of current row and column scores
  # @param theta: a vector of length p with the dispersion parameters

  # @return The evaluation of the score function at beta, a scalar

  mu = muMarg * exp(reg* beta)
  
  sum(reg*(X-mu)/(1+(mu/theta)))

}
#--------------------------------------#
#A jacobian for the psi parameters
NBjacobianPsi = function(beta, X, reg, muMarg, theta){
  # @param beta: the current guess for the psi parameter, a scalar
  # @param X: the nxp data matrix
  # @param muMarg: the nxp offset
  # @param reg: the regressor matrix, the outer product of current row and column scores
  # @param theta: nxp matrix with the dispersion parameters (converted to matrix for numeric reasons)
  
  # @return The evaluation of the Jacobian function at beta, a 1x1 matrix
  
  mu = muMarg * exp(reg* beta)
  matrix(-sum(reg^2*(1+X/theta)*mu/(1+mu/theta)^2),1,1)

}
#-------------------------------------# 
#The influence function for the psis
NBpsiInfl = function(psi, X, cMat, rMat, muMarg, theta){
  reg = rMat %*% cMat
  mu = muMarg * exp(reg* psi)
  -((X-mu)*(theta+mu))/(reg*(theta+X)*mu)
}

#--------------------------------------#
#A score function of the NB for the column scores
dNBllcol = function(beta, X, reg, thetas, muMarg, k, p, n, colWeights, nLambda, cMatK, constrained = FALSE) {
  # @param beta: vector of length p+1+1+(k-1): p row scores, 1 centering, one normalization  and (k-1) orhtogonality lagrangian multipliers
  # @param X: the nxp data matrix
  # @param reg: a nx1 regressor matrix: outer product of rowScores and psis
  # @param theta: nxp matrix with the dispersion parameters (converted to matrix for numeric reasons)
  # @param mumarg: the nxp offset
  # @param k: a scalar, the dimension of the RC solution
  # @param p: a scalar, the number of taxa
  # @param n: a scalar, the number of samples
  # @param nLambda: the number of restrictions
  # @param colWeights: the weights used for the restrictions
  # @param cMatK: the lower dimensions of the colScores
  # @param constrained: a boolean, indicates if ordination is constrained.
  
  # @return A vector of length p+1+1+(k-1) with evaluations of the derivative of lagrangian
  
  cMat = matrix(beta[1:p], byrow=TRUE, ncol=p, nrow=1)
  
  mu = if (constrained){exp(rowMultiply(reg, c(cMat))) * muMarg} else {exp(reg %*% cMat) * muMarg}
  
  lambda1 = beta[p+1] #Lagrangian multiplier for centering restrictions sum(abunds*r_{ik}) = 0
  lambda2 = beta[p+2] #Lagrangian multiplier for normalization restrictions sum(abunds*r^2_{ik}) = 1
  
  lambda3 = if(k==1){0} else {beta[(p + 3):length(beta)]} #Lagrangian multiplier for orhtogonalization restriction
  
  score = if(constrained) {
    colSums(reg*((X-mu)/(1+mu/thetas))) + (colWeights*lambda1 + colWeights*lambda2*2*cMat + (lambda3 %*% cMatK)*colWeights)
  } else {
    crossprod(reg,((X-mu)/(1+mu/thetas))) + (colWeights*lambda1 + colWeights*lambda2*2*cMat + (lambda3 %*% cMatK)*colWeights)
    }

  center = sum(colWeights*cMat)
  unitSum = sum(colWeights*cMat^2)-1
  if(k==1) {
      return(c(score, center, unitSum))
    }
  orthogons = tcrossprod(cMatK, cMat*colWeights)
      return(c(score, center, unitSum, orthogons))
}
#--------------------------------------#
# A function to provide an analytical jacobian in the hope to speed up the calculations
NBjacobianCol = function(beta, X, reg, thetas, muMarg, k, n ,p, colWeights, nLambda, cMatK, constrained = FALSE){
  # @param beta: vector of length p+1+1+(k-1): p row scores, 1 centering, one normalization  and (k-1) orhtogonality lagrangian multipliers
  # @param X: the nxp data matrix
  # @param reg: a nx1 regressor matrix: outer product of rowScores and psis
  # @param theta: nxp matrix with the dispersion parameters
  # @param mumarg: the nxp offset
  # @param k: a scalar, the dimension of the RC solution
  # @param p: a scalar, the number of taxa
  # @param n: a scalar, the number of samples
  # @param nLambda: the number of restrictions
  # @param colWeights: the weights used for the restrictions
  # @param cMatK: the lower dimensions of the colScores
  # @param constrained: a boolean, indicates if ordination is constrained.

  #@return a symmetric jacobian matrix of size p + k +1
  cMat = matrix(beta[1:p], byrow=TRUE, nrow=1, ncol=p)

  #Calculate the mean
  mu = if (constrained){exp(rowMultiply(reg, c(cMat))) * muMarg} else {exp(reg %*% cMat) * muMarg}
  
  Jac = matrix(0, nrow= p + nLambda, ncol=p + nLambda)
  #The symmetric jacobian matrix. The upper part is filled first, then mirror image is taken for lower triangle
#   
  #dLag²/dr_{ik}dlambda_{1k}
  Jac[1:p,(p+1)] = colWeights
  #Jac[1:(p*k),(p*k+1):((p+1)*k)] = sapply(1:k, function(K){c(rep(0,(K-1)*p),colWeights,rep(0,(k-K)*p))})
  Jac[1:p,p+2] = colWeights*2 *cMat

  tmp = (1+X/thetas)*mu/(1+mu/thetas)^2

    #dLag²/ds_{ik}dlambda_{3kk'}
  if(k>1){
    Jac[1:p,(p+3):(p+nLambda)] = t(cMatK)*colWeights
  }

  #Symmetrize
   Jac = Jac + t(Jac)
  #dLag²/dr_{ik}²
    diag(Jac)[1:p] = if(constrained) {
          -colSums(tmp * reg^2) + 2*beta[p+2]*colWeights
    } else {
    -crossprod(tmp, reg^2) + 2*beta[p+2]*colWeights
    }
  Jac
}
#--------------------------------------#
# The influence function for the column scores
NBcolInfl = function(X, psis, cMat, rMat, thetas, colWeights, k, lambdaCol){
  reg = psis[k] *rMat[,k]
  muMarg = outer(rowSums(X), colSums(X)/sum(X)) * exp(rMat[,1:k, drop=FALSE] %*% (cMat[1:k,, drop=FALSE]*psis[1:k]))
  p=ncol(X)
  n = nrow(X)
  lambdaCol = lambdaCol[seq_k(k)]
  cMatK = cMat[1:(k-1),,drop=FALSE]
  nLambda = length(lambdaCol)
  tmp = if(k>1) lambdaCol[-(1:2)] %*% cMatK else 0
  
  score= t(t(reg*((X-muMarg)/(1+t(t(muMarg)/thetas)))) + c(colWeights*(lambdaCol[1] + lambdaCol[2]*2*cMat[k,] + tmp)))

  JacobianInv = solve(NBjacobianCol(beta = c(cMat[k,], lambdaCol), X = X, reg= reg, thetas = thetas, muMarg = muMarg, k = k, p = p, n=n ,colWeights = colWeights , nLambda = nLambda, cMatK = cMatK)) #Inverse Jacobian
  
  #After a long thought: The X's do not affect the estimation of the lambda parameters!
  #Matrix becomes too large: return score and inverse jacobian
  return(list(score=score, InvJac = JacobianInv[1:p,1:p]))
}
#-------------------------------------------#

#A function to extract the influence for a given parameter index
getInflCol = function(score, InvJac, parIndex){
  t(t(score)* InvJac[, parIndex])
}

#-------------------------------------------#
#A score function of the NB for the row scores
dNBllrow = function(beta, X, reg, thetas, muMarg, k, n , p, rowWeights, nLambda, rMatK) {
  # @param beta: a vector of k regression parameters to optimize: the k column scores
  # @param X: the data matrix of dimensions nxp
  # @param reg: a 1xp regressor matrix: outer product of rowScores and psis
  # @param thetas: nxp matrix with the dispersion parameters (converted to matrix for numeric reasons)
  # @param muMarg: an nxp offset matrix
  # @param k: a scalar, the dimension of the RC solution
  # @param p: a scalar, the number of taxa
  # @param n: a scalar, the number of samples
  # @param rowWeights: a vector of length n, the weights used for the restrictions
  # @param nLambda: an integer, the number of lagrangian multipliers
  # @param rMatK: the lower dimension row scores

  # @return A vector of length n + k +1 with evaluations of the derivative of the lagrangian
  rMat = matrix(beta[1:n], byrow=FALSE, ncol=1, nrow=n)
  mu = exp(rMat %*% reg)* muMarg
  
  lambda1 = beta[n+1] #Centering restrictions sum(abunds*r_{ik}) = 0
  lambda2 = beta[n+2] #normalization restrictions sum(abunds*r^2_{ik}) = 1
  lambda3 = if(k==1){0} else {beta[(n+3):length(beta)]}

  score = c(tcrossprod(reg, (X-mu)/(1+(mu/thetas)))) + (rowWeights*lambda1 + rowWeights*lambda2* 2*rMat + (rMatK %*% lambda3)*rowWeights)
  #
  center = sum(rMat*rowWeights)
  unitSum = sum(rMat^2*rowWeights)-1
  if(k==1){ return(c(score,center, unitSum))}
  orthogons = crossprod(rMatK, rMat*rowWeights)
  return(c(score,center, unitSum, orthogons))
}
#-------------------------------------------#
# A function to provide an analytical jacobian to speed up the calculations
NBjacobianRow = function(beta, X, reg, thetas, muMarg, k, n ,p, nlambda, rowWeights, nLambda, rMatK){
  # @param beta: a vector of k regression parameters to optimize: the k column scores
  # @param X: the data matrix of dimensions nxp
  # @param reg: a 1xp regressor matrix: outer product of rowScores and psis
  # @param thetas: nxp matrix with the dispersion parameters (converted to matrix for numeric reasons)
  # @param muMarg: an nxp offset matrix
  # @param k: a scalar, the dimension of the RC solution
  # @param p: a scalar, the number of taxa
  # @param n: a scalar, the number of samples
  # @param rowWeights: a vector of length n, the weights used for the restrictions
  # @param nLambda: an integer, the number of lagrangian multipliers
  # @param rMatK: the lower dimension row scores
  
  #@return a symmetric jacobian matrix of size (n+2) *k + k(k-1)/2
  rMat = matrix(beta[1:n], byrow=FALSE, ncol=1, nrow=n)

  mu = exp(rMat %*% reg)* muMarg

  Jac = matrix(0, nrow= n + nLambda, ncol= n + nLambda)
  #The symmetric jacobian matrix. The upper part is filled first, then mirror image is taken for lower triangle

  #dLag²/dr_{ik}dlambda_{1k}
  Jac[1:n, n+1] = rowWeights
  #dLag²/dr_{ik}dlambda_{2k}
  Jac[1:n, n+2] = 2 *rMat*rowWeights
  tmp = (1+X/thetas)*mu/(1+mu/thetas)^2

  #dLag²/dr_{ik}dlambda_{3kk'}
  if(k>1){
    Jac[1:n,(n+3):(n+nLambda)] = rMatK*rowWeights
  }
  #Symmetrize
  Jac = Jac + t(Jac)
#dLag²/dr_{ik}²
 diag(Jac)[1:n] = -tcrossprod(reg^2 ,tmp) + 2*rowWeights*beta[n+2]
    
  Jac
}
#--------------------------------------#
# The influence function for the row scores
NBrowInfl = function(X, psis, cMat, rMat, thetas, rowWeights, k, lambdaRow){
  reg = psis[k] *cMat[k,]
  muMarg = outer(rowSums(X), colSums(X)/sum(X)) * exp(rMat[,1:k, drop=FALSE] %*% (cMat[1:k,, drop=FALSE]*psis[1:k]))
  p=ncol(X)
  n = nrow(X)
  lambdaRow = lambdaRow[seq_k(k)]
  rMatK = rMat[,1:(k-1),drop=FALSE]
  nLambda = length(lambdaRow)
  tmp = if(k>1) rMatK %*% lambdaRow[-(1:2)] else 0
  
  score= reg*(X-muMarg)/(1+t(t(muMarg)/thetas)) + c(rowWeights*(lambdaRow[1] + lambdaRow[2]*2*rMat[,k] + tmp))

  JacobianInv = solve(NBjacobianRow(beta = c(rMat[,k], lambdaRow), X = X, reg= reg, thetas = thetas, muMarg = muMarg, k = k, p = p, n=n, rowWeights = rowWeights , nLambda = nLambda, rMatK = rMatK)) #Inverse Jacobian
  
  #After a long thought: The X's do not affect the estimation of the lambda parameters!
  #Matrix becomes too large: return score and inverse jacobian
  return(list(score=score, InvJac = JacobianInv[1:n,1:n]))
}
#-------------------------------------------#

#A function to extract the influence for a given parameter index
getInflRow = function(score, InvJac, parIndex){
  score* InvJac[, parIndex]
}

#-------------------------------------------#
# A function to estimate the overdispersion of the NB with empirical Bayes using the edgeR pacakage
estDisp = function (X, cMat = NULL, rMat = NULL, muMarg, psis, trended.dispersion, prior.df = 10, dispWeights = NULL, rowMat = NULL)
{
  # @param X: the data matrix of dimensions nxp
  # @param cMat: a 1xp colum scores matrix
  # @param rMat(optional): a nx1 rowscores matrix, if unconstrained
  # @param muMarg: an nxp offset matrix
  # @param psis: a scalar, the current psi estimate
  # @param trended.dispersion: a vector of length p with pre-calculated trended.dispersion estimates. They do not vary in function of the offset anyway
  # @param prior.df(optional): an integer, number of degrees of freedom of the prior for the Bayesian shrinkage
  # @param dispWeights (optional): Weights for estimating the dispersion in a zero-inflated model
  # @param rowMat: matrix of row scores in case of constrained ordination
  
  # @return A vector of length p with dispersion estimates
  
  require(edgeR)
  logMeansMat = 
    if(!is.null(rMat)){ #Unconstrained
      t(rMat %*% (cMat * psis) + log(muMarg))
    } else { #Constrained
      t(log(muMarg) + psis* rowMat)
    }
  thetaEsts <- 1/estimateGLMTagwiseDisp(y = t(X), design = NULL,
                                        prior.df = prior.df, offset = logMeansMat, dispersion = trended.dispersion, weights = dispWeights)
  if (anyNA(thetaEsts)) {
    idNA = is.na(thetaEsts)
    thetaEsts[idNA] = mean(thetaEsts[!idNA])
    warning(paste(sum(idNA), "dispersion estimations did not converge!"))
  }
  return(thetas = thetaEsts)
}
#-------------------------------------------#
# A function to trim based on confounders to ensure there are no taxa with only zero counts in one of the subgroups defined by the confounders. Returns a trimmed data matrix, should be called prior to fitting the independence model
trimOnConfounders = function(confounders, X, prevCutOff, minFraction, n){
  # @param confounders: a nxt confounder matrix
  # @param X: the nxp data matrix
  # @param prevCutOff: a scalar, the prevalence cut off
  # @param minFraction: a scalar, each taxon's total abundance should equal at leat the number of samples n times minFraction, otherwise it is trimmed
  # @param n: the number of samples
  
  # @return A trimmed data matrix nxp'
  
  trimmingID = apply(X, 2, function(x){ #Over taxa
    any(apply(confounders, 2, function(conf){ #Over confounding variables
      tapply(X = x, INDEX = conf, FUN = function(y){mean(y!=0) <= prevCutOff | sum(y)<(n*minFraction)}) #Any all-zero subgroup?
    }))
  })
  
  if(all(trimmingID)){
    stop("All taxa would be trimmed, please provide a covariate with less levels! \n")
  }
  
  X[, !trimmingID] #Return trimmed X
}

#-------------------------------------------#
#A function to filter out the effect of known confounders. This is done by fitting interactions of every taxon with the levels of the confounders. 

filterConfounders = function(muMarg, confMat, X, thetas, p, n, nleqslv.control, trended.dispersion, tol = 1e-3, maxIt = 20){
  # @param muMarg: a nxp matrix, the current offset
  # @param confounders: a nxt confounder matrix
  # @param X: the nxp data matrix
  # @param thetas: a vector of length p with the current dispersion estimates
  # @param p: a scalar, the number of columns of X
  # @param tol: a scalar, the convergence tolerance
  # @param maxIt (optional): maximum number of iterations, defaults to 20

  # @return a list with components:
  # -thetas: new theta estimates
  # -muMarg: a new offset matrix
  # -NB_params: the estimated parameters of the interaction terms

  NB_params = matrix(0, ncol(confMat), p)

  iter = 1
  while((iter ==1) || ((iter <= maxIt) && (!convergence))){
    
    NB_params_old = NB_params

     NB_params = sapply(seq_len(p), function(i){
     nleq = try(nleqslv(NB_params[,i] , reg = confMat,  fn = dNBllcol_constr, theta = thetas[i], muMarg = muMarg[,i], cReg = 1, X = X[,i], control = nleqslv.control, jac = JacCol_constr)$x)  # , ... Fit the taxon-by taxon NB with given overdispersion parameters and return predictions
     if(class(nleq)=="try-error"| anyNA(nleq)| any(is.infinite(nleq))){
     nleq = nleqslv(NB_params[,i] , reg = confMat,  fn = dNBllcol_constr, theta = thetas[i], muMarg = muMarg[,i], cReg =1, X = X[,i], control = nleqslv.control)$x # If fails try with numeric jacobian
     }
     return(nleq)
   }) #Estimate response functions

     thetas = estDisp(X = X, cMat = matrix(0, ncol=p), rMat = matrix(0, nrow=n), psis = 0, muMarg = muMarg * exp(confMat %*% NB_params), trended.dispersion = trended.dispersion) #Estimate overdispersion
     
     iter = iter +1
     
  convergence = sqrt(mean((NB_params_old-NB_params)^2)) < tol #Check for convergence, L2-norm
  
  }
  
  list(muMarg = muMarg * exp(confMat %*% NB_params), thetas = thetas, NB_params = NB_params)
  }
```

## Constrained RC(M)

```{r LRnb, purl=TRUE} 
# DON'T USE "p" as variable name, partial matching in the grad-function in the numDeriv package!!!

# A function that returns the value of the log-likelihood ratio of alpha, keeping the response functions fixed
LR_nb <- function(Alpha, X, CC, responseFun = c("linear","quadratic","nonparametric"), muMarg, psi, nleqslv.control = list(trace=FALSE), NB_params, NB_params_noLab, thetaMat, n, ncols, nonParFit, nonParamRespFun,...){#
  # @param Alpha: a vector of length d, the environmental gradient
  # @param X: the n-by-p count matrix
  # @param CC: the n-by-d covariate matrix
  # @param responseFun: a character string indicating the type of response function
  # @param muMarg: an n-by-p offset matrix
  # @param thetas: a vector of length p with estimated dispersion parameters (they will not be re-estimated to keep computation time down)
  # @param ncols: a scalar, the number of columns of X
  # @param psi: a scalar, an importance parameter
  # @param nleqslv.control: the control list for the nleqslv() function
  # @param n: number of samples
  # @param NB_params_start: Starting values for the NB_params
  # @param NB_params_start_noLab: Starting values for the NB_params without label
  # @param v: an integer, number of parameters to be estimated

  # @return: a scalar, the evaluation of the log-likelihood ratio at the given alpha
  
	# alpha<-alpha/sqrt(sum(alpha^2))#
    sampleScore = CC %*% Alpha #A linear combination of the environmental variables yields the sampleScore (the rMat vector)
    #cat(sampleScore, "\n \n")
    design = switch(responseFun,
                     linear = model.matrix(~  sampleScore),#Without intercept
                     quadratic = model.matrix(~ sampleScore + I(sampleScore^2))
                    ) 
    if(responseFun %in% c("linear","quadratic")){
    muT = muMarg * c(exp(design %*% NB_params *psi))
    mu0 = muMarg * c(exp(design %*% NB_params_noLab *psi))
    } else { #Non-parametric response function
    muT = muMarg * exp(nonParamRespFun$taxonWise*psi)
    mu0 = muMarg * exp(nonParamRespFun$overall*psi)
    }
    logDensj = dnbinom(X, mu = muT, size = thetaMat, log = TRUE) #Likelihoods under species specific model
    #Immediately return log likelihoods

    logDens0 = dnbinom(X, mu = mu0, size = thetaMat, log = TRUE) #Likelihoods of null model

 
    lr <- sum(logDensj-logDens0) # The likelihood ratio
    return(-lr) # opposite sign for the minimization procedure

}
#--------------------------------------#
# A function that returns the value of the partial derivative of the log-likelihood ratio to alpha, keeping the response functions fixed
dLR_nb <- function(Alpha, X, CC, responseFun = c("linear", "quadratic", "nonparametric"), psi, NB_params, NB_params_noLab, d, alphaK, k, centMat, nLambda, nLambda1s, thetaMat, muMarg, ncols, ...){
  # @param Alpha: a vector of length d + k*(2+(k-1)/2), the environmental gradient plus the lagrangian multipliers
  # @param X: the n-by-p count matrix
  # @param CC: a n-by-d covariate vector
  # @param responseFun: a character string indicating the type of response function
  # @param NB_params: Starting values for the NB_params
  # @param NB_params_noLab: Starting values for the NB_params without label
  # @param d: an integer, the number of covariate parameters
  # @param k: an integer, the current dimension
  # @param centMat: a nLambda1s-by-d centering matrix

  # @return: The value of the lagrangian and the constraining equations
  
  #Extract the parameters
  alpha = Alpha[seq_len(d)]
  lambda1s = Alpha[d+seq_len(nLambda1s)] #Multiple centering requirements now!
  lambda2 = Alpha[d+nLambda1s+1]
  lambda3 = if(k==1) {0} else {Alpha[(d+nLambda1s+2):(d+nLambda)]}
  
  sampleScore = CC %*% alpha #A linear combination of the environmental variables yields the sampleScore 
  design = switch(responseFun,
                     linear = model.matrix(~ sampleScore),#With intercept
                     quadratic = model.matrix(~ sampleScore + I(sampleScore^2)),
                    stop("Non-parametric model not yet implemented")
                    ) 
  
  mu = muMarg * exp(design %*% NB_params *psi)
  mu0 = muMarg * c(exp(design %*% NB_params_noLab*psi))
  tmp = (X-mu)/(1+mu/thetaMat)
  tmp0 = (X-mu0)/(1+mu0/thetaMat)

  lag = switch(responseFun, #The lagrangian depends on the shape of the response function
               "linear" = psi * (crossprod(CC, tmp) %*% (NB_params[2,]) - rowSums(crossprod(CC, tmp0*NB_params_noLab[2]))) ,
               "quadratic" = psi * (
                  c(crossprod(CC, tmp) %*% (NB_params[2,])) +
                  c(crossprod(CC * c(sampleScore), tmp) %*% (NB_params[3,]) * 2)  -  
                  rowSums(crossprod(CC, tmp0) *NB_params_noLab[2]) - 
                  rowSums(crossprod(CC * c(sampleScore), tmp0) * NB_params_noLab[3])*2),
               stop("Unknown response function provided! \n")) + #Restrictions do not depend on response function
                 c(lambda1s %*% centMat) +
                 lambda2 * 2 * alpha + 
                if(k>1) alphaK %*% lambda3 else 0
  
  centerFactors = centMat %*% alpha #Includes overal centering
  size = sum(alpha^2)-1
  if(k==1) { return(c(lag, centerFactors, size))}
  ortho = crossprod(alphaK ,alpha)
  c(lag, centerFactors, size, ortho)
}
#In the quadratic case the fit is really dependent on starting values
#--------------------------------------#
# A function that returns the Jacobian of the likelihood ratio
LR_nb_Jac = function(Alpha, X, CC, responseFun = c("linear", "quadratic", "nonparametric"), psi, NB_params, NB_params_noLab, d, alphaK, k, nLambda, nLambda1s, centMat, thetaMat, muMarg, n, v, ncols, ...){
  # @param Alpha: a vector of length d + k*(2+(k-1)/2), the environmental gradient plus the lagrangian multipliers
  # @param X: the n-by-p count matrix
  # @param CC: a n-by-d covariate vector
  # @param responseFun: a character string indicating the type of response function
  # @param NB_params_start: Starting values for the NB_params
  # @param NB_params_start_noLab: Starting values for the NB_params without label
  # @param d: an integer, the number of covariate parameters
  # @param k: an integer, the current dimension

  # @return: The jacobian, a d-by-d symmetric matrix
  
  did = seq_len(d)
  #Extract the parameters
  alpha = Alpha[did]
  lambda1s = Alpha[d+seq_len(nLambda1s)] #Multiple centering requirements now!
  lambda2 = Alpha[d+nLambda1s+1]
  lambda3 = if(k==1) {0} else {Alpha[(d+nLambda1s+2):(d+nLambda)]}
  
  sampleScore = CC %*% alpha #A linear combination of the environmental variables yields the sampleScore 
  design = switch(responseFun,
                     linear = model.matrix(~ sampleScore),#With intercept
                     quadratic = model.matrix(~ sampleScore + I(sampleScore^2)),
                    stop("Non-parametric model not yet implemented")
                    ) 
  
  mu = muMarg * exp(design %*% NB_params *psi)
  mu0 = muMarg * c(exp(design %*% NB_params_noLab*psi))
  
  Jac = matrix(0, nrow= d + nLambda, ncol = d + nLambda)
  #dLag²/dalpha_{yk}dlambda_{1k}
  Jac[(d+seq_len(nLambda1s)), did] = centMat
 
  Jac[did,d+nLambda1s+1] = 2 * alpha

  if(responseFun=="linear"){
    tmp = rowMultiply((1+X/thetaMat)*mu/(1+mu/thetaMat)^2,NB_params[2,]^2)
    tmp0 = (1+X/thetaMat)*mu0/(1+mu0/thetaMat)^2 * NB_params_noLab[2]^2
  } else if (responseFun=="quadratic"){
    tmp = (1+X/thetaMat)*mu/(1+mu/thetaMat)^2
    tmp0 = (1+X/thetaMat)*mu0/(1+mu0/thetaMat)^2 
  }
  #dLag²/ds_{ik}dlambda_{3kk'}
  if(k>1){
    Jac[(d+nLambda1s+2):(d+nLambda), did] = alphaK
  }
  
  #Symmetrize
   Jac = Jac + t(Jac)
   cSam = c(sampleScore)
   cSam2 = cSam^2
   
  Jac[did,did] = switch(responseFun,
                           "linear" = - psi^2 *(colSums(tensor(vapply(did, FUN.VALUE = tmp, function(x){CC[,x]*(tmp-tmp0)}), CC,1,1))),
                          "quadratic" = colSums((tensor(vapply(did, FUN.VALUE = tmp, function(x){CC[,x]*(-psi^2*(tmp*(matrix(NB_params[2,]^2,n,ncols, byrow =TRUE)+4*outer(cSam,NB_params[2,]*NB_params[3,])+4*outer(cSam2,NB_params[3,]^2))-
tmp0*(NB_params_noLab[2]+NB_params_noLab[3]*2*cSam)^2) + 2*psi*(rowMultiply((X-mu)/(1+mu/thetaMat),NB_params[3,])-(X-mu0)/(1+mu0/thetaMat)*NB_params_noLab[3]))}),CC,1,1))
))
  
  diag(Jac)[did] = diag(Jac)[did] + 2*lambda2 #Correct the diagonal

  Jac
}
#--------------------------------------#
# A function to estimate the taxon-wise NB-params
estNBparams = function(design, thetas, muMarg, psi, X, nleqslv.control, ncols, initParam,v){
vapply(seq_len(ncols), FUN.VALUE = vector("numeric",v), function(i){
     nleq = nleqslv(initParam[,i] , reg = design,  fn = dNBllcol_constr, theta = thetas[i], muMarg = muMarg[,i], psi = psi, X = X[,i], control = nleqslv.control, jac = JacCol_constr)$x
     return(nleq)
})
  }
#--------------------------------------#
# A function to estimate the NB-params ignoring the taxon labels
estNBparamsNoLab = function(design, thetas, muMarg, psi, X, nleqslv.control, initParam, n, v){
     #Without taxon Labels
nleqslv(x = initParam , reg = design,  fn = dNBllcol_constr_noLab, thetas = thetas, muMarg = muMarg, psi = psi, X = X, control = nleqslv.control, jac = JacCol_constr_noLab, n=n, v=v)$x
}
#--------------------------------------#
# A function to estimate the taxon-wise response functions non-parametrically, and return their fitted values
estNPresp = function(sampleScore, muMarg, X, ncols, psi, ...){
  taxonWiseFits = lapply(seq_len(ncols), function(i){
  locfit.raw(x = sampleScore, y = X[,i]/muMarg[,i], link = "log",...)
})
  overall = matrix(normCurve(locfit.raw(x = rep(sampleScore,ncols), y = c(X/muMarg), link = "log",...)),ncol = ncols)
  taxonWise = sapply(taxonWiseFits, normCurve)
  list(taxonWise = taxonWise, overall = matrix(overall,ncol = ncols), taxonWiseFits = taxonWiseFits)
}
#--------------------------------------#
# A function to normalize the non-parametric curves to total area 1. Only the psi parameter can grow in size
normCurve = function(fit, newDataLength = 200){
  x = seq(min(fit$eva$xev), max(fit$eva$xev), length.out = newDataLength)
  auc = flux::auc(x = x, y = predict(fit, newdata = x))
  fitted(fit)/auc
}
#--------------------------------------#
# A function to evaluate the score function for 1 taxon at the time
dNBllcol_constr = function(betas, X, reg, theta, muMarg, psi) {
  # @param beta: a vector of regression parameters (\boldsymbol{beta}_j)
  # @param X: the nx1 data vector
  # @param reg: a matrix of regressors with the same length as beta (\mathbf{C} \boldsymbol{\alpha})
  # @param theta: The dispersion parameter
  # @param mumarg: offset of length n
  # @param cReg:a scalar psi_k * s_{jk}

  # @return The evaluation of the score functions

  mu = exp(c(reg %*% betas)*psi) * muMarg
  crossprod((X-mu)/(1+mu/theta) , reg) * psi
}
#--------------------------------------#
# The jacobian of this function
JacCol_constr = function(betas, X, reg, theta, muMarg, psi) {
  # @param beta: a vector of regression parameters (\boldsymbol{beta}_j)
  # @param X: the nx1 data vector
  # @param reg: a matrix of regressors with the same length as beta (\mathbf{C} \boldsymbol{\alpha})
  # @param theta: The dispersion parameter
  # @param mumarg: offset of length n
  # @param cReg:  a scalar psi_k * s_{jk}

  # @return The jacobian

  mu = exp(c(reg %*% betas)* psi) * muMarg
  tmp = (1+X/theta)*mu/(1+mu/theta)^2
  -crossprod(tmp*reg,reg) * psi^2 #Don't forget to square psi!

}
#--------------------------------------#
# A function to evaluate the score function ignoring taxon labels
dNBllcol_constr_noLab = function(betas, X, reg, thetas, muMarg, psi, n, ncols,...) {
  # @param beta: a vector of regression parameters with length v (\boldsymbol{beta}_j)
  # @param X: the nxp data matrix
  # @param reg: a matrix of regressors of dimension nxv (\mathbf{C} \boldsymbol{\alpha})
  # @param thetas: The dispersion parameters (a vector of length p)
  # @param muMarg: offset matrix of dimension nxp
  # @param cReg: a vector of length p with the column scores * importance parameter

  # @return The evaluation of the score functions (a vector length v)

  mu = c(exp(reg %*% betas*psi)) * muMarg
  colSums(crossprod((X-mu)/(1+(mu/thetas)),reg) * psi)

}
#--------------------------------------#
# The jacobian of this function
JacCol_constr_noLab = function(betas, X, reg, thetas, muMarg, psi, n ,v) {
  # @param beta: a vector of regression parameters with length v (\boldsymbol{beta}_j)
  # @param X: the nxp data matrix
  # @param reg: a matrix of regressors of dimension nxv (\mathbf{C} \boldsymbol{\alpha})
  # @param thetas: The dispersion parameters (a vector of length p)
  # @param muMarg: offset matrix of dimension nxp

  # @return The jacobian (a v-by-v matrix)

  mu = c(exp(reg %*% betas*psi)) * muMarg
  tmp = (1+(X/thetas))*mu/(1+(mu/thetas))^2*psi^2 #Don't forget to square psi!
  -crossprod(reg, vapply(seq_len(v), FUN.VALUE = vector("numeric", n), function(x){rowSums(reg[,x]*tmp)}))
  }

#--------------------------------------#
# A function to estimate alpha based on the LR-criterion
estAlpha = function(Alpha, X, CC, d, k, ncols, centMat, n, v, nLambda1s, nLambda, muMarg, thetas, psi, alphaK, maxItAlpha = 200, responseFun = c("linear", "quadratic", "nonparametric"),  control.outer = list(trace=FALSE), control.optim = list(trace=FALSE), nleqslv.control = list(), convNorm = 2, tol = 1e-3, NB_params = NULL, NB_params_noLab = NULL, analytic = FALSE, lambdasAlpha = NULL, verbose = TRUE){
  # @param alpha: a vector of length d, the environmental gradient
  # @param alphaK: a matrix with d columns d, the environmental gradient of the previous dimensions
  # @param ncols: a scalar, the number of columns of X
  # @param k: a scalar, dimension index
  # @param v: an integer, the number of parameters per taxon of the response function
  # @param nLambdas1: an integer, number of centering equations
  # @param nLambdas: an integer, total number of lambda parameters
  # @param X: the n-by-ncols count matrix
  # @param C: the n-by-d covariate matrix
  # @param responseFun: a character string indicating the type of response function
  # @param muMarg: an n-by-ncols offset matrix
  # @param thetas: nxp matrix with with estimated dispersion parameters (they will not be re-estimated within the alpha estimation to keep computation time down)
  # @param cReg: a vector of length p, column scores * importance parameter
  # @param maxItAlpha: an integer, the maximum number of iterations
  
  # @return: a vector of length  with the new alpha estimate
  
  iterA = 1

  if(responseFun %in% c("linear","quadratic")){
    
 while ((iterA == 1) || ((iterA <= maxItAlpha) &&                           
             (!convergenceA))) { 
  AlphaOld = Alpha
  if(verbose) cat("Environmental gradient iteration ",iterA, "\n")
  #New strategy: estimate the parameters and then optimize alpha, iterate between both.
  sampleScore = CC %*% Alpha
  design = switch(responseFun,
                     linear = model.matrix(~sampleScore),
                     quadratic = model.matrix(~sampleScore + I(sampleScore^2) )) #With intercept
  
NB_params = estNBparams(design = design, thetas = thetas[1,], muMarg = muMarg, psi = psi, X = X, nleqslv.control = nleqslv.control, ncols =ncols, initParam = NB_params, v = v)

NB_params = NB_params/sqrt(rowSums(NB_params^2)) #Normalize the parameters

NB_params_noLab = nleqslv(x = NB_params_noLab , reg = design,  fn = dNBllcol_constr_noLab, thetas = thetas, muMarg = muMarg, psi = psi, X = X, control = nleqslv.control, jac = JacCol_constr_noLab, n=n, v=v)$x
  
 if(analytic){
   AlphaTmp = nleqslv(x = c(Alpha,lambdasAlpha), fn = dLR_nb, jac = LR_nb_Jac, X = X, CC = covariates, responseFun = responseFun, cMat = cMat, psi = psi, NB_params = NB_params, NB_params_noLab = NB_params_noLab, alphaK = alphaK, k = k, d = d, centMat = centMat, nLambda = nLambda, nLambda1s = nLambda1s, thetaMat = thetas, muMarg = muMarg, control = nleqslv.control, n=n, v=v, ncols = ncols)$x
  Alpha = AlphaTmp[seq_len(d)]
  lambdasAlpha = AlphaTmp[d+seq_len(nLambda)]
} else {
   AlphaTmp = constrOptim.nl(par = Alpha, fn = LR_nb, gr = NULL, heq = heq_nb, heq.jac = heq_nb_jac, alphaK = alphaK, X=X, CC=CC, responseFun = responseFun, muMarg = muMarg, psi = psi, d = d, ncols=ncols, control.outer = control.outer, control.optim = control.optim, nleqslv.control = nleqslv.control, k =k, centMat = centMat, NB_params = NB_params, NB_params_noLab = NB_params_noLab, v = v, thetaMat = thetas, n=n) #
  Alpha = AlphaTmp$par
  lambdasAlpha = AlphaTmp$lambda
}
  
  iterA = iterA + 1
  convergenceA = sum((AlphaOld - Alpha)^convNorm)^(1/convNorm) < tol
  } 
  
  return(list(Alpha = Alpha, NB_params = NB_params, NB_params_noLab = NB_params_noLab, convergence = convergenceA, lambdasAlpha = lambdasAlpha ))
  } else {
     while ((iterA == 1) || ((iterA <= maxItAlpha) &&                           
             (!convergenceA))) { 
    
  AlphaOld = Alpha
  cat("Environmental gradient iteration ",iterA, "\n")
  #New strategy: estimate the parameters and then optimize alpha, iterate between both.
  sampleScore = CC %*% Alpha
  nonParamRespFun = estNPresp(sampleScore = sampleScore, muMarg = muMarg, X = X, ncols = ncols, psi = psi)
  AlphaTmp = constrOptim.nl(par = Alpha, fn = LR_nb, gr = NULL, heq = heq_nb, heq.jac = heq_nb_jac, alphaK = alphaK, X=X, CC=CC, responseFun = responseFun, muMarg = muMarg, d = d, ncols=ncols, control.outer = control.outer, control.optim = control.optim, nleqslv.control = nleqslv.control, k =k, centMat = centMat, n=n, nonParamRespFun = nonParamRespFun, psi = psi, thetaMat = thetas)
  Alpha = AlphaTmp$par
  lambdasAlpha = AlphaTmp$lambda
  
  iterA = iterA + 1
  convergenceA = sum((AlphaOld - Alpha)^convNorm)^(1/convNorm) < tol
     }
    return(list(Alpha = Alpha, convergence = convergenceA, nonParamRespFun = nonParamRespFun))  
  }
}
#--------------------------------------#
# A linear equality constraints function for estAlpha (see ?alabama)
heq_nb = function(Alpha, alphaK, d, k,  centMat, ...){
  centerFactors = centMat %*% Alpha #Includes overal centering
  size = sum(Alpha^2)-1
  if(k==1) { return(c( centerFactors, size))}
  ortho = crossprod(alphaK,  Alpha)
  c(centerFactors, size, ortho)
}
#--------------------------------------#
# It's jacobian, returns a matrix with d columns and at least 2 rows (numerically checked with nloptr::check.derivatives)
heq_nb_jac = function(Alpha, alphaK, d, k, centMat, ...){
  if(k==1) {return(rbind( centMat, 2*Alpha))
    } else {rbind(centMat, 2*Alpha, t(alphaK))}
}
```

$$-(2hb_2cy_2 + b_1cy_2)(2hb_2 +
b_1)cy \psi^2
\frac{\mu}{(\frac{\mu}{\theta} + 1)} \\ + (2hb_2cy_2 + b_1cy_2)(2hb_2 + b_1)(\mu - x) cy\psi^2 \frac{\mu}{\theta(\mu)/\theta + 1)^2}) - 2(\mu) - x)b_2cycy_2\frac{\psi}{\frac{\mu}{\theta} + 1})$$

## Wrapper for all distributions

```{r Wrapper all dists, purl = TRUE}
#A simple wrapper function for phyloseq objects and matrices, for all posible distributions it passes all argument onto the correct function.
#Trim on prevalence and total abundance to avoid instability of the algorithm. We cannot conlcude much anyway on lowly abundant taxa
RCM = function(dat, k, round=FALSE, distribution= "NB", prevCutOff = 0.025, minFraction = 0.1, rowWeights = "uniform", colWeights = "marginal", covariates = NULL, confounders = NULL, prevFit = NULL, ...){
  # @param dat: an nxp count matrix or a phyloseq object with an otu_table slot
  # @param k: an integer, the number of dimensions of the RCM solution
  # @param round (optional): a boolean, whether to round to nearest integer. Defaults to FALSE.
  # @param distribution (optional): a character string, the error distribution of the RC(M) model. Defaults to NB
  # @param prevCutoff (optional): a scalar, the prevalance cutoff for the trimming. Defaults to 2.5e-2
  # @param minFraction (optional): a scalar, each taxon's total abundance should equal at leat the number of samples n times minFraction, otherwise it is trimmed. Defaults to 10%
  # @param rowWeights, colWeights (optional): character strings, the weighting procedures for the normalization of row and column scores. Defaults to "uniform" and "marginal" respectively
  # @param covariates (optional): In case "dat" is a phyloseq object, the names of the sample variables to be used as covariates in the constrained analysis, or "all" to indicate all variables to be used. In case dat is a matrix, a nxf matrix or dataframe of covariates. Character variables will be converted to factors, with a warning. Defaults to NULL, in which case an unconstrained analysis is carried out.
  # @param confounders (optional): In case "dat" is a phyloseq object, the names of the sample variables to be used as confounders to be filtered out. In case dat is a matrix, a nxf matrix or dataframe of confounders Character variables will be converted to factors, with a warning. Defaults to NULL, in which case no filtering occurs.
  # @param prevFit(optional) An object with a previous fit, normally from a lower dimension, that should be extended.
  
  classDat = class(dat) #The class provided
  
  ##The count data##
  if (classDat=="matrix"){
  X = dat
  }else  if(classDat=="phyloseq"){
  X = if (taxa_are_rows(dat)) t(otu_table(dat)@.Data) else otu_table(dat)@.Data
  } else {stop("Please provide a matrix or a phyloseq object! \n")}
  p=ncol(X); n=nrow(X)
  
  if(is.null(colnames(X))){colnames(X)=1:ncol(X)}
  if(is.null(rownames(X))){rownames(X)=1:nrow(X)}
  colNames = colnames(X); rowNames =rownames(X)
  # names(rowWeightsNum) =rowNames; names(colWeightsNum) = colNames
  if(round) {X=round(X, 0) }#Round to integer
  
  #Check X type
  if(!all(sapply(X, function(x){(x%%1)==0}))){stop("Please provide integer count matrix! \n")
    } else{X=matrix(as.integer(X), ncol=ncol(X), nrow=nrow(X))}

  colnames(X)=colNames; rownames(X)=rowNames
  X=X[rowSums(X)>0, ] #Remove empty samples
  X=X[, (colMeans(X==0)<(1-prevCutOff)) & (colSums(X)>(n*minFraction))] #Remove txa with low prevalence and those who do not meet the minFraction requirement
  if (distribution %in% c("ZIP","ZINB")) X = X[rowSums(X==0)>0, colSums(X==0)>0] #For a zero-inflated model, make sure every row and column has zeroes
  
  ##Build confounder matrix if applicable. For the preliminary trimming, we do not include an intercept, but we do include all the levels of the factors using contrasts=FALSE: we want to do the trimming in every subgroup, so no hidden reference levels. For the filtering we just use a model with an intercept and treatment coding, here the interest is only in adjusting the offset##
  if(!is.null(confounders)){
    if(class(confounders) %in% c("vector","matrix")){
      if(NROW(X)!=NROW(confounders)){ #Check dimensions
        stop("Data and confounder matrix do not have the same number of samples! \n")
      }
      if(is.vector(confounders)){
      confounders = as.matrix(confounders) #Convert to matrix if only 1 variable
      }
      if(is.null(colnames(confounders))){ #assign names if needed
        colnames(confounders) = paste("var",seq_len(NCOL(confounders)))
      }
      if(class(confounders[,1]) %in% c("character", "factor")){ #If character or factor, build model matrix
        if(class(confounders[,1])=="character"){
          warning("Converting character vector to factor! \n")
        }
        confModelMatTrim = model.matrix(object = as.formula(paste("~" , paste(colnames(confounders), collapse="+"),"-1")), contrasts.arg = apply(colnames(confounders),2,contrasts, contrasts=FALSE)) #No intercept for preliminary trimming
        confModelMat = model.matrix(object = as.formula(paste("~", paste(colnames(confounders), collapse="+"))), contrasts.arg = apply(colnames(confounders),2,contrasts, contrasts=TRUE)) #With intercept for filtering
      } else {
        confModelMat = confModelMatTrim = confounders
      }
    } else if(class(confounders) == "data.frame"){
      if(NROW(X)!=NROW(confounders)){ #Check dimensions
        stop("Data and confounder matrix do not have the same number of samples! \n")
      }
      confModelMatTrim = model.matrix( #No intercept for preliminary trimming
        object = as.formula(paste("~", paste(names(confounders), collapse="+"),"-1")),
        data = datFrame, 
        contrasts.arg = lapply(confounders[sapply(confounders, is.factor)],
                               contrasts, contrasts=FALSE))
      confModelMat = model.matrix( #With intercept for filtering
        object = as.formula(paste("~", paste(names(confounders), collapse="+"))), 
        data = datFrame, 
        contrasts.arg = lapply(confounders[sapply(confounders, is.factor)],
                               contrasts, contrasts = TRUE))
    } else if(class(confounders)=="character"){
      if(classDat != "phyloseq"){
        stop("Providing confounders through variable names is only allowed if phyloseq object is provided! \n")
      }
      datFrame = data.frame(sample_data(dat)) # The dataframe with the confounders
      if(anyNA(datFrame)){stop("Confounders contin missing values!\n")}
    confModelMatTrim = model.matrix(
        object = formula(paste("~", paste(confounders, collapse="+"),"-1")), 
        data = datFrame, 
        contrasts.arg = lapply(datFrame[sapply(datFrame, is.factor), drop=FALSE][, confounders,drop=FALSE],
                               contrasts, contrasts=FALSE)) 
    confModelMat = model.matrix(
        object = formula(paste("~", paste(confounders, collapse="+"))), 
        data = datFrame, 
        contrasts.arg = lapply(datFrame[sapply(datFrame, is.factor), drop=FALSE][, confounders,drop=FALSE],
                               contrasts, contrasts = TRUE)) 
    } else{
      stop("Please provide the confounders either as matrix, dataframe, or character string! \n")
    }
  } else{
    confModelMat = confModelMatTrim = NULL
  }

  ##Build covariate matrix if applicable##
  # In this case we will 1) Include dummy's for every level of the categorical variable, and force them to sum to zero. This is needed fro plotting and required for normalization. 2) Exclude an intercept. The density function f() will provide this already. See introduction.
  if(!is.null(covariates)){
    if(class(covariates) == "data.frame"){
      if(NROW(X)!=NROW(covariates)){ #Check dimensions
        stop("Data and covariate matrix do not have the same number of samples! \n")
      }
    datFrame = covariates
    covariatesNames = names(covariates)
    } else if(class(covariates)=="character"){
      if(classDat != "phyloseq"){
        stop("Providing covariates through variable names is only allowed if phyloseq object is provided! \n")
      }
      if(covariates[[1]]=="all"){covariates = sample_variables(dat)} #Enable the "all" option if phyloseq object is provided
      datFrame = data.frame(sample_data(dat))[,covariates] # The dataframe with the covariates
      covariatesNames = covariates
    } else{
      stop("Please provide the covariates either as matrix, dataframe, or character string! \n")
    }
    
    if(any(sapply(datFrame, is.logical))){
      datFrame[,sapply(datFrame, is.logical)] = lapply(datFrame[sapply(datFrame, is.logical)], as.factor) #Convert logicals to factors
    #warning("Logicals converted to factors! \n", immediate. = TRUE). No warning needed
    }
   if(any(sapply(datFrame, is.integer))){
      datFrame[,sapply(datFrame, is.integer)] = lapply(datFrame[sapply(datFrame, is.integer)], as.numeric) #Convert integers to numeric
    warning("Integer values treated as numeric! \n", immediate. = TRUE)
      }
    if(any(sapply(datFrame, is.character))){
      datFrame[,sapply(datFrame, is.character)] = lapply(datFrame[sapply(datFrame, is.character)], factor) #Convert characters to factor
      warning("Character vectors treated as factors! \n", immediate. = TRUE)
    }
    nFactorLevels = sapply(datFrame, function(x){if(is.factor(x)) nlevels(x) else 1}) #Number of levels per factor
    datFrame[,sapply(datFrame, is.factor) & (nFactorLevels < 2)] = NULL #Drop factors with one level
    if(any(sapply(datFrame, is.factor) & (nFactorLevels < 2))){
        warning("The following variables were not included in the analyses because they are factors with only one level: \n", paste(covariates[sapply(datFrame, is.factor) & (nFactorLevels < 2)], sep = " \n"),immediate. = TRUE)
      }
      if(anyNA(datFrame)){
        NArows = apply(datFrame, 1, anyNA)
        if(all(NArows)){ stop("All samples have missing covariates")}
        X = X[!NArows,]
        datFrame = datFrame[!NArows,]
       if(!is.null(confModelMat)){
          confModelMat = confModelMat[!NArows,]
          confModelMatTrim = confModelMatTrim[!NArows,]
        }
        warning(paste("Some covariates contain missing values. We removed samples \n", paste(which(NArows), collapse = ", "), "\n prior to analysis." ),immediate. = TRUE)
      }
    
    # Center and scale the continuous covariates
    datFrame[sapply(datFrame, is.numeric)] = scale(datFrame[sapply(datFrame, is.numeric)])
      
      covModelMat = model.matrix(
        object = formula(paste("~", paste(covariatesNames, collapse="+"),"-1")), 
        data = datFrame, 
        contrasts.arg = lapply(datFrame[sapply(datFrame, is.factor)],
                               contrasts, contrasts=FALSE)) 
    #Already prepare the matrix that defines the equations for centering the coefficients of the dummy variables
    centMat  = t(sapply(seq_along(nFactorLevels), function(i){
      c(rep.int(0, sum(nFactorLevels[seq(0,i-1)])), #Zeroes before
        rep.int(if(nFactorLevels[i]==1) 0 else 1, nFactorLevels[i]), #Ones within the categorical variable
        rep.int(0, sum(nFactorLevels[-seq(1,i)]))) #Zeroes after
    })) 

    centMat = if(all(rowSums(centMat)==0)) {matrix(0, 1,sum(nFactorLevels))} else {centMat[rowSums(centMat)>0,, drop = FALSE]} #Remove zero rows, corresponding to the non-factors
    #centMat = rbind(1, centMat) # Add a row of ones for the overal sum. This turned out to be a bad idea!
  } else {
    covModelMat = centMat = NULL
  }
  
tic = proc.time() #Time the calculation
  tmp = switch(distribution, 
               NB=RCM_NB(X, rowWeights = rowWeights, colWeights = colWeights, k = k, 
                         confounders  = list(confounders = confModelMat, confoundersTrim = confModelMatTrim), 
                         covariates = covModelMat, prevCutOff = prevCutOff, minFraction = minFraction, centMat = centMat, NBRCM = prevFit,...),
               ZIP=RCM_ZIP(X, rowWeights = rowWeights, colWeights = colWeights, k = k, 
                            confounders  = list(confounders = confModelMat, confoundersTrim = confModelMatTrim), 
                           covariates = covModelMat, prevCutOff = prevCutOff, minFraction = minFraction, ZIPRCM = prevFit,...),
               ZINB=RCM_ZINB(X,  rowWeights = rowWeights, colWeights = colWeights, k = k, 
                              confounders  = list(confounders = confModelMat, confoundersTrim = confModelMatTrim), 
                             covariates = covModelMat, prevCutOff = prevCutOff, minFraction = minFraction, ZINBRCM = prevFit, ...))
  if(classDat=="phyloseq"){tmp$physeq = prune_samples(rownames(X),prune_taxa(colnames(X),dat)) } 
  tmp = within(tmp, {
    runtimeInMins = (proc.time()-tic)[1]/60 + if(is.null(prevFit)) {0} else {prevFit$runtimeInMins} # Sum the runtimes
    k = k #Store number of dimensions
  })
  class(tmp) = "RCM"
  tmp
}

#Likelihoods-------------------------------------
# A function to calculate the likelihoods of
#-The independence model
#-The saturated model
#-The fitted model
#-All models with dimension k 0<k<K
#Which overdispersions to use is a non-trivial problem.  One option would be to use the estimated dispersions of the full model for all calculations. Another is to estimate the overdispersions of the independence and lower dimensional models separately and use them. The problem is that if we use the edgeR machinery again, we get stable estimates but not MLE's, so that the likelihood of the independence model can sometimes be larger than that of a RC model. We provide three options, specified through the Disp parameter:
# - "MLE" Use the MLE's of the separate models where possible,
# - "edgeR" Use the edgeR robust estimate separately for every model
liks = function(rcm, Disp=c("edgeR","MLE")){
  require(MASS)
  #@param rcm: a list, the output of the outerLoop function
  
  #@return a list with components
    #-indLL: likelihood of the indepence model
    #-LL1,..., LL[K-1]: likelihood of intermediate models
    #-LLK: The likelihood of the fitted model
  #Independence model
  C = colSums(rcm$X)
  R = rowSums(rcm$X)
  onesn =rep.int(1, nrow(rcm$X))
  onesp = rep.int(1, ncol(rcm$X))
  E = diag(R) %*% outer(onesn ,onesp) %*% diag(C)/sum(C)
  
  if(Disp=="MLE"){
    
  #Estimate dispersions for the independence model
  # thetasInd = estDisp(rcm$X, cMat=matrix(0,ncol=ncol(rcm$X)), rMat=matrix(0,nrow=nrow(rcm$X)), libSizes=rowSums(rcm$X), abunds=colSums(rcm$X)/sum(rcm$X), psis=0)
  thetasInd =sapply(1:ncol(rcm$X), function(i){theta.ml(y=rcm$X[,i],mu=E[,i])})
  #The overdispersions of the independence model are larger: This is logical since less variation has been explained in this model
  
    #Estimate the overdispersions for the intermediate models and the Full RC model
    LLintDisp = vapply(1:(ncol(rcm$rMat)),FUN.VALUE=rep(0, ncol(rcm$X)), function(k){
      mu=E * exp(rcm$rMat[, 1:k, drop=FALSE] %*% (rcm$cMat[1:k,,drop=FALSE]*rcm$psis[1:k]))
    sapply(1:ncol(rcm$X), function(i){theta.ml(y=rcm$X[,i],mu=mu[,i])})
  })

  } else if(Disp=="edgeR"){
  #Independence model
  trended.dispersion.ind  <- estimateGLMTrendedDisp(y = t(rcm$X), design = NULL, method = "bin.loess",offset=t(log(E)))
  thetasInd = 1/estimateGLMTagwiseDisp(y = t(rcm$X), design = NULL,  offset=t(log(E)), dispersion = trended.dispersion.ind)
  
  #RCM models
  thetasInt = vapply(1:(ncol(rcm$rMat)),FUN.VALUE=rep(0, ncol(rcm$X)), function(k){
  mu=E * exp(rcm$rMat[, 1:k, drop=FALSE] %*% (rcm$cMat[1:k,,drop=FALSE]*rcm$psis[1:k]))
  trended.dispersion  <- estimateGLMTrendedDisp(y = t(rcm$X), design = NULL, method = "bin.loess",offset=t(log(mu)))
  1/estimateGLMTagwiseDisp(y = t(rcm$X), design = NULL,  offset=t(log(mu)), dispersion = trended.dispersion)
  })
  LLintDisp=cbind(thetasInd, thetasInt)
  }  else{stop("No valid dispersion estimation paradigm provided! Choose either MLE or edgeR")}
    names(LLintDisp) = paste0("dispLL",1:(ncol(rcm$rMat)-1))
#Now we have the overdispersions, estimate the likelihoods
  
  #Estimate the likelihoods
  LLintList = mapply(1:(ncol(rcm$rMat)-1),1:(ncol(LLintDisp)-2), FUN = function(k, ThetasI){
    sum(dnbinom(rcm$X, mu = E * exp(rcm$rMat[, 1:k, drop=FALSE] %*% (rcm$cMat[1:k,,drop=FALSE]*rcm$psis[1:k])), size=LLintDisp[,ThetasI], log=TRUE))
  })
  names(LLintList) = paste0("LL",1:(ncol(rcm$rMat)-1))
  indLL = sum(dnbinom(rcm$X, mu=E, size=thetasInd, log=TRUE))
  
#Full RC model
  LLK = sum(dnbinom(rcm$X, mu=E * exp(rcm$rMat %*% (rcm$cMat*rcm$psis)), size=LLintDisp[, ncol(LLintDisp)], log=TRUE))
    
  c(indLL=indLL, LLintList,  LLK=LLK)
}
```

## Plotting function

```{r plot, purl=TRUE}
plotRCM = function(RCM, X = NULL, thetas = NULL, abunds = NULL, arrowFrac = 0.04, biplot = TRUE,
                   libLoc ="topright", libInset = c(-0.7,-0.1), libCex = 0.85, libInterSp = 0.75, libLeg = TRUE, samColour=NULL,
                   dispInset = c(0,-0.4), abInset = c(0,-0.4), abundLeg=FALSE, stressSpecies = NULL, taxColour = NULL, taxLegPos = "bottomright", taxCol = NULL, taxInset = libInset, taxIntersp = libInterSp,
                   asp=1, xpd=TRUE, mar=c(4,5,5,5), Dim=c(1,2), ...){
  
  # @param psis: vector of length k with psi estimates
  # @param rMat: a nxk matrix with final row scores
  # @param cMat: a pxk with matrix with final column scores
  # @param Dim (optional): a vector of length 2, indicating which dimensions to fit, defaults to the first two
  # @param X (optional): the nxp data matrix
  # @param thetas (optional): a vector of length p with estimates for the overdispersion
  # @param abunds(optional): a vector of length p with abundance estimates (otherwise it is estimated from the data)
  # @param arrowFrac(optional): Fraction of largest species to plot. defaults to 0.1
  # @param biplot(optional): A boolean indicating if species should be added to the plot, defaults to TRUE
  # @param libLoc(optional): a string, location of the library size legend. Defaults to "topleft"
  # @param libLegend(optional): a boolean, should library size legend be displayed? defaults to TRUE
  # @param libInset, dispInset, abInset(optional): numeric vectors of length 2, insets for library size, dispersion and abundance legends
  # param stressSpecies(optional): names of species to be highlighted
  # @param ... additional arguments, passed on to the plot() function
  
  cMat = RCM$cMat
  rMat = RCM$rMat
  psis = RCM$psis
  
  # @return: NONE,  plots the result in the plotting window
  tmp = par(no.readonly = TRUE)
  par(mar=mar, pty="s")
  if(!(length(psis)== NCOL(rMat) && length(psis) == NROW(cMat))){
    stop("Dimensions don't match")
  }
  #Reorder dimensions according to importance

  a = Dim[1]
  b = Dim[2]
  
  if(!is.null(RCM$physeq) && is.character(samColour)){
    tmp = get_variable(RCM$physeq, samColour)
    samColour = if(is.factor(tmp)) tmp else factor(tmp)
  }
  
  plot(x = rMat[,a] * psis[a],
  y = rMat[,b] *psis[b],
  xlab=paste0("Dim",a),
  ylab=paste0("Dim",b),
  xlim = max(abs(rMat[,a] * psis[a]))*c(-1,1),
  ylim = max(abs(rMat[,b] * psis[b]))*c(-1,1),
  col = samColour,
  asp=asp,
  ...)
  if(libLeg){
legend(pch = 1, col = seq_along(levels(samColour)), legend = levels(samColour), x=libLoc, inset=libInset, xpd=xpd, cex = libCex, y.intersp = libInterSp, x.intersp = 0.7)
}
# {legend(libLoc,legend=c("Small library size","Large library size"),
#       pch=c(1,1), col=c("blue","red"), inset=libInset, xpd=xpd)
#   }
  
    if (biplot){
      #Arrows
        arrowLengths = apply(cMat[Dim,],2,function(x){sqrt(sum(x^2))})
    id = arrowLengths >= quantile(arrowLengths,1-arrowFrac)
        scalingFactor = max(abs(apply(t(t(rMat[,Dim])*psis[Dim]),2, range)))/
        max(abs(cMat[Dim,id]))*0.99
    arrows(x0=0,y0=0,x1=cMat[a,id]*scalingFactor,y1=cMat[b, id]*scalingFactor, col=taxColour[id]
           )
    # legend(x = taxLegPos, legend = levels(taxColour),pch = 1, col = seq_along(levels(taxColour)) )
    if( abundLeg){
      legend(x = taxLegPos, legend = levels(taxColour), xpd=xpd, cex = libCex,lty = 1, col = taxCol, inset = taxInset, y.intersp = taxIntersp)
    }
    #if(!is.null(stressSpecies))
    }
return(invisible())
  par(tmp)
}
# A ggplot2-version of the biplot function
plot.RCM = function(RCMfit, Dim = c(1,2), 
                    samColour = NULL, colLegend = samColour, samShape = NULL, shapeLegend = samShape, samSize = 3,
                    taxNum = 10, scalingFactor = NULL, plotTaxa = TRUE, quadDrop = 0.995, nPoints = 1e3, plotEllipse = TRUE, taxaScale = 0.5,
                    Palette = NULL, arrowCol = "blue", nudge_y = -0.08,...) {
  # @param RCMfit: an RCM object
  # @param Dim: two dimensions to plot, defaults to the first two
  
  # @param samColour: a character string, the variable to use for the colour of the sample dots
  # @param colLegend: a character string, the legend text for the sample colour. Defaults to the name of the colour variable
  # @param samShape: a character string, the variable to use for the shape of the sample dots
  # @param shapeLegend: a character string, the text to use for the shapeLegend. Defaults to the name of the shape variable
  # @param samSize: a scalar, the size of the sample dots, defaults to 3
  
  # @param taxNum: an integer, the number of taxa to be plotted
  # @param scalingFactor: a scalar, a user supplied scaling factor for the taxon location. If not supplied it will be calculated to make sample and taxon plots on the same scale
  # @param plotTaxa: a boolean, whether to add the taxa to the plot to make a bi- or triplot
  # @param quadDrop: a scalar, the value at which the response function has dropped to this fraction of its maximum on the count scale to be plotted as a halo
  # @param nPoints: an integer, number of points to evaluate for the ellipse
  # @param plotEllipse: a boolean, whether to add the ellipses
  # @param taxaScale: a scalar, by which to scale the rectangles of the quadratic taxon plot
  
  # @param Palette: the colour palette
  # @param arrowCol: a character string: the colour of the arrows of the environmental gradient.
  # @param nudge_y: a scalar, the offet for the taxon labels
  
  #Retrieve dots (will be passed on to aes())
    dotList = list(...)
    constrained = !is.null(RCMfit$covariates)
    if(constrained){
      dataSam <- data.frame(RCMfit$covariates %*% RCMfit$alpha[,Dim] %*% diag(RCMfit$psis[Dim]))
    } else {
        dataSam <- data.frame(RCMfit$rMat[, Dim] %*% diag(RCMfit$psis[Dim]))
      }
    names(dataSam)=paste0("Dim", Dim)
    
    #Get the colours
    if(!is.null(samColour) && !is.factor(samColour)){
      dataSam$colourPlot = get_variable(RCMfit$physeq, samColour)
      if(is.character(dataSam$colourPlot)) dataSam$colourPlot  = factor(dataSam$colourPlot )

    } else if(is.factor(samColour)){
      dataSam$colourPlot = samColour} else {dataSam$colourPlot=factor(rep(1, nrow(dataSam)))}
    
    #     #Set colour palette
    if(is.null(Palette)){
Palette=rainbow(length(unique(dataSam$colourPlot)))
    }  
  taxFrac = taxNum/ncol(RCMfit$X)
    
    #Construct dataframe for taxa
    
      if(constrained) {
        if(RCMfit$responseFun=="linear"){
          dataTax = data.frame(
            origin1 = -RCMfit$NB_params[1,,Dim[1]]/RCMfit$NB_params[2,,Dim[1]] ,
            origin2 = -RCMfit$NB_params[1,,Dim[2]]/RCMfit$NB_params[2,,Dim[2]] ,
            slope1 = RCMfit$NB_params[2,,Dim[1]] * RCMfit$psis[Dim[1]], #The gradient
            slope2 = RCMfit$NB_params[2,,Dim[2]] * RCMfit$psis[Dim[2]]
            ) 
          dataTax = within(dataTax, {
            end1 = origin1 + slope1 
            end2 = origin2 + slope2
          })
    arrowLengths = apply(dataTax[, c("slope1","slope2")],1,function(x){sqrt(sum(x^2))})
    id = arrowLengths >= quantile(arrowLengths,taxFrac)
    #Filter out small arrows
    dataTax = dataTax[id,] 
    if(is.null(scalingFactor)){
    scalingFactor = min(abs(apply(dataSam[, paste0("Dim", Dim)],2,range)))/max(abs(dataTax[, c("slope1","slope2")]))*0.925
    }
    dataTax = within(dataTax, { #Scale the arrows
            end1 = origin1 + slope1 * scalingFactor
            end2 = origin2 + slope2 * scalingFactor
          })
     } else if (RCMfit$responseFun == "quadratic"){
          dataTax = data.frame(
              apply(RCMfit$NB_params[c(2,3),,Dim],c(2,3), function(x){
                a = x[2]; b=x[1]
                -b/(2*a)
            })) #The location of the extrema
          names(dataTax) = c("end1","end2")
          dataTax$colour = apply(RCMfit$NB_params[3,,Dim],1, function(x){
              if(all(x>0)) {
                return("green")
              } else if (all(x<0)){
                return("red")
              } else if(x[1] > 0){
                return("brown")
              } else{
                return("purple")
              }
            })
          dataEllipseTmp = vapply(seq_along(taxa_names(RCMfit$physeq)), FUN.VALUE = matrix(0,nPoints,3),function(tax){
              x = RCMfit$NB_params[,tax,Dim]
              cbind(ellipseCoord(a = x[3,] * RCMfit$psis, b = x[2,] * RCMfit$psis, c = x[3,] * RCMfit$psis, quadDrop = quadDrop, nPoints = nPoints), taxon = tax)
              })
  
          peakHeights = apply(RCMfit$NB_params[,,Dim],2, function(x){
            A = x[3,]; B = x[2,]; C = x[1,];
            sapply(exp(B^2 * RCMfit$psis-4*A*C)/(4*A), function(y){max(y,1/y)}) #select largest relative departure
          })
          rownames(peakHeights) = c("peak1","peak2")
          meanPeakHeights = colMeans(peakHeights)
          dataTax = cbind(dataTax, t(peakHeights))
           #Pick taxa with largest extrema, within observed values of the envrionmental scores (otherwise it is almost extrapolation)
          dataID = data.frame(
            meanPeakHeights = meanPeakHeights,
            id = seq_len(nrow(dataTax)),
            dataTax
          )
          envScores = RCMfit$covariates %*% RCMfit$alpha
          id = dataID[with(dataID, 
                               order(end1 > max(envScores[,Dim[1]]) | end1 < min(envScores[,Dim[1]]), 
                                     end2 > max(envScores[,Dim[2]]) | end2 < min(envScores[,Dim[2]]),
                                     -meanPeakHeights)),
                      ]$id[seq_len(ceiling(taxFrac*nrow(dataTax)))]
          dataTax = dataTax[id,] 
          #Rescale peak heights for plotting
          dataTax[, c("peak1","peak2")] = taxaScale* rowMultiply(dataTax[, c("peak1","peak2")], apply(dataTax[, c("end1","end2")],2,function(x){max(abs(x))})/apply(dataTax[, c("peak1","peak2")], 2,max))
          dataTax[, c("peak1","peak2")] = apply(dataTax[, c("peak1","peak2")], c(1,2), max, 0.0075) # Make sure a line always appears
          #Unfold into two dimensions
          dataEllipse = data.frame(
            apply(dataEllipseTmp[,,id], 2, c), colour = as.character(dataTax$colour)
          )
        } else if(RCMfit$responseFun  == "nonparametric"){ #For non-parametric response function we cannot plot the taxa
          plotTaxa = FALSE
        } else {
          stop("No valid response function present in this RCM object!")
        }
           } else { #If not constrained
    dataTax = data.frame(cbind(t(RCMfit$cMat[Dim,]),0,0))
    names(dataTax) = c("end1","end2", "origin1","origin2")
    }
    if(!constrained){ #Scaling needed
    arrowLengths = apply(dataTax[, c("end1","end2")],1,function(x){sqrt(sum(x^2))})
    id = arrowLengths >= quantile(arrowLengths,1 - taxFrac)
    #Filter out small arrows
    dataTax = dataTax[id,] 
    if(is.null(scalingFactor)){
    scalingFactor = min(abs(apply(dataSam[, paste0("Dim", Dim)],2,range)))/max(abs(dataTax[, c("end1","end2")]))*0.925
    }
    dataTax = within(dataTax, { #Scale the arrows
            end1 = end1 * scalingFactor
            end2 = end2 * scalingFactor
          })
     } # End scaling needed
    if(plotTaxa) dataTax$labels = sub(" ", "\n", colnames(RCMfit$X)[id])

    plot = ggplot(dataSam, aes_string(x=names(dataSam)[1], y=names(dataSam)[2], col = "colourPlot", dotList)) + 
      geom_point(size = samSize) + #point size
      xlab(paste0(names(dataSam)[1],": ", paste0("psi",Dim[1]), " = ",round(RCMfit$psis[Dim[1]],1))) + #xlabel
      ylab(paste0(names(dataSam)[2],": ", paste0("psi",Dim[2]), " = ",round(RCMfit$psis[Dim[2]],1))) + #ylabel
      if(is.null(samColour)) {guides(color=FALSE)} #Legend

    #add legend names
    if(!is.null(colLegend) & is.factor(dataSam$colourPlot) ){
      plot = plot + scale_colour_manual(name=colLegend, values =Palette)
    }    else if(!is.null(colLegend) & !is.factor(dataSam$colourPlot) ){
      plot = plot + scale_colour_continuous(name=colLegend)
    } 
    if(!is.null(shapeLegend)){
      plot = plot + scale_shape_discrete(name=shapeLegend)
    }
  
    #Add arrows or labels
    if(length(arrowCol)>1){
     arrowCol = Palette[c(arrowCol[id])] 
    }
    if(plotTaxa){
      if(!constrained || RCMfit$responseFun=="linear"){
    plot <- plot + geom_segment(data=dataTax, aes_string(x='origin1', y='origin2', xend="end1", yend = "end2"), arrow=arrow(length=unit(0.2,"cm")), alpha=0.75, color=arrowCol, show.legend=FALSE)
    } else if(RCMfit$responseFun=="quadratic"){ #quadratic response functions
      plot <- plot + 
      geom_tile(data=dataTax, aes_string(x='end1', y='end2', fill="colour", width = "peak1", height = "peak2" ), pch = 21, show.legend=FALSE, inherit.aes = FALSE) + #The centers
        if(plotEllipse) {geom_path(inherit.aes = FALSE, data = dataEllipse, mapping = aes_string(x = "x", y = "y", group = "taxon"), colour = "grey50", show.legend = FALSE)}
    } else {
    plot <- plot + geom_point(data=dataTax, aes_string(x='end1', y='end2', fill="colour"), pch = 21, show.legend=FALSE, inherit.aes = FALSE) 
    }

      plot <- plot + geom_text(data=dataTax, aes_string(x="end1", y = "end2", label = "labels"),  alpha=0.75, color=arrowCol, show.legend=FALSE, nudge_y = nudge_y)
    } else {}
    if(constrained){
      #Add variable labels
      varData = data.frame(RCMfit$alpha)
      varData$label = rownames(RCMfit$alpha)
      scalingFactorAlpha = min(abs(apply(dataSam[, paste0("Dim", Dim)],2, range)))/max(abs(varData[, paste0("Dim", Dim)]))*0.99
      varData[, paste0("Dim", Dim)] = varData[, paste0("Dim", Dim)]*scalingFactorAlpha
      plot = plot + geom_text(data = varData, mapping = aes_string(x = names(dataSam)[1], y = names(dataSam)[2], label = "label"), inherit.aes = FALSE)
    }
    squarePlot(plot)
}
#--------------------------------#
# A function to enforce squareness
squarePlot <- function(plt){
    return(plt+coord_fixed()+
            expand_limits(x=ggplot_build(plt)$panel$ranges[[1]]$y.range,
                          y=ggplot_build(plt)$panel$ranges[[1]]$x.range))
}
#--------------------------------#
# A function that returns the coordinates of an ellipse, given the parameters of teh quadratic function and the fraction of the peak at which to draw the ellipse
ellipseCoord = function(a,b,c, quadDrop = 0.95, nPoints = 100){
  center = -b/(2*a)
  roots = mapply(a,b,c, FUN = function(A,B,C){
     Re(polyroot(c(C - log(ifelse(A>0,1/quadDrop,quadDrop)) + (B^2-4*A*C)/(4*A), B, A))[1]) #Log to go to count scale. Since function is symmetric, does not matter which root we pick, only distance to vertex matters so we just take the first
  })
  anglesEval = seq(0,2*pi, length.out = nPoints)
  ab = center-roots
  radii = prod(ab)/sqrt(sin(anglesEval)^2*ab[1]^2 + ab[2]^2* cos(anglesEval)^2)
  cbind(x = radii*cos(anglesEval) + center[1], y = radii * sin(anglesEval) + center[2])
}
```

## Auxiliary functions

```{r auxFuns}
# A function to efficiently row multiply a a-by-b matrix by a vector of length b. More memory intensive but that does not matter with given matrix sizes
rowMultiply = function(matrix, vector){
  matrix * rep(vector, rep(nrow(matrix),length(vector)))
}
#An auxiliary R function to "array" multiply an array with a vector, kindly provided by Joris Meys
#-------------------------------------------#
arrayprod <- function(x,y){
  # @param x: a axbxc array
  # @param y: a vector of length c matrix
  
  # @return a axb matrix
  xdim <- dim(x)
  outdim <- xdim[1:2]
  outn <- prod(outdim)
 
  yexpand <- rep(y, each = outn)

  tmp <- x * yexpand
 
  dim(tmp) <- c(outn, xdim[3])
  out <- rowSums(tmp)
 
  dim(out) <- outdim
 
  out
}
#-------------------------------------------#
# A function to drop the recorded parameter estimates throughout the iterations to save space
dropArrayRec = function(RCM){
  class(RCM)= "list"
 tmp =  within(RCM, {
    rowRec = colRec = psiRec = thetaRec = NULL
  })
 class(tmp)= "RCM"
 tmp
}
```

#### Test the software

Some preliminary test to make sure everything runs without errors

```{r tests, purl = FALSE, eval = FALSE}
n=50;p=200
counts = matrix(rnbinom(n*p, mu = 5, size = 3), n, p)
covars = data.frame(age = rnorm(n, 35,5), gender = sample(c(TRUE, FALSE), n , replace = TRUE), country = factor(sample(c("France","Belgium","Germany"),n, replace=TRUE)), BMI = rnorm(n , 27,6), status = sample(c("Treatment","Control"),n, replace=TRUE), siblings = rpois(n, lambda = 0.9))
physeq = phyloseq(otu_table(counts, taxa_are_rows = FALSE), sample_data(covars))
confounderVec = get_variable(physeq, "country")
d = 30
v = 2
X = matrix(rpois(n*p,2), n, p)
thetaMat = abs( matrix(rnorm(p), n, p, byrow = TRUE))
psi = 3
muMarg = outer(rowSums(X), colSums(X))/sum(X)
NB_params = betas = matrix(rnorm(p*v),v,p)
reg = model.matrix(~rnorm(n))


#Unconstrained one dimension matrix
un1Dmat = RCM(counts, k=1)
# Add two dimensions
un3Dmat = RCM(dat = counts, prevFit = un1Dmat, k=3)

#Unconstrained one dimension phyloseq
un1DmatPhy = RCM(physeq, k=1)
# Add two dimensions (even accepts a matrix!)
un3DmatPhy = RCM(dat = counts, prevFit = un1DmatPhy, k=3)

#Unconstrained one dimension matrix with confounder filtering
un1DmatConf = RCM(counts, k=1, confounders = data.frame(confounderVec))
# Add two dimensions
un3DmatConf = RCM(dat = counts, prevFit = un1DmatConf, k=3)

#Check some requirements
str(un3DmatConf$confParams)
colSums(un3Dmat$rMat^2)
crossprod(un3Dmat$rMat)
colSums(t(un3Dmat$cMat) * un3Dmat$abunds)
rowSums(un3Dmat$cMat^2)
sum(un3Dmat$cMat[1,]* un3Dmat$cMat[2,] * un3Dmat$abunds)
# All is well!

#Unconstrained one dimension physeq with confounder filtering
un1DmatConfPhy = RCM(physeq, k=1, confounders = "country")
# Add two dimensions
un3DmatConfPhy = RCM(dat = counts, prevFit = un1DmatConf, k=3)

#Constrained one dimension matrix without confounder filtering
con1Dmat = RCM(physeq, k=1, covariates =  sample_variables(physeq), analytic = TRUE)
# Add two dimensions
con3Dmat = RCM(dat = physeq, prevFit = con1Dmat, k=3, covariates =  sample_variables(physeq))

con2DmatLin = RCM(physeq, k = 2, covariates =  sample_variables(physeq), analytic = TRUE, responseFun = "linear")

con2DmatQuad = RCM(physeq, k = 2, covariates =  sample_variables(physeq), analytic = TRUE, responseFun = "quadratic")
# The zeller data?
con2DmatNon = RCM(physeq, k = 2, covariates =  sample_variables(physeq), analytic = TRUE, responseFun = "nonparametric")

plot.RCM(con2DmatLin, colour = "BMI")
plot.RCM(con2DmatQuad, colour = "BMI")
plot.RCM(con2DmatNon, colour = "BMI")

con2DmatLin$alpha
con2DmatQuad$alpha
con2DmatNon$alpha
```

#### Real data examples

Finally we also apply the method to real datasets. First the AGP and HMP datasets

```{r RC(M)_NB real datasets, purl = FALSE}
load("/home/stijn/PhD/American Gut/AGphylo.RData")
load("/home/stijn/PhD/Simulations/data/physeqListV13.RData")
realNames = c("AGP", "Hard.palate","Buccal.mucosa","Anterior.nares","Left.Antecubital.fossa", "Posterior.fornix", "Mid.vagina", "Left.Retroauricular.crease", "Vaginal.introitus", "Stool", "Hard.palate" )
if(!file.exists(file = "resListRCMagphmp.RData")){
fileNames = paste(realNames, "RCM.RData")#"Kostic", "zeller16S", "zellerMeta",
resListRCM = mapply(fileNames, c(AGphylo, physeqListV13[realNames[-1]]),  SIMPLIFY = FALSE, FUN=function(x,y){
  load(x)
  RCMres$physeq=y
  RCMres
})
names(resListRCM) = realNames
save(resListRCM, file = "resListRCMagphmp.RData")
} else {load(file = "resListRCMagphmp.RData")}
```

##### The HMP dataset: sample plots

```{r HMPplot, purl = FALSE, include = FALSE, fig.cap="Sample plots of the RC(M) ordination of the HMP dataset. Colours indicate gender. \\label{fig:samHMP}"}
SClevels  = unique(c(unlist(sapply(physeqListV13, function(x){
  levels(get_variable("RUNCENTER",physeq=x))
}))))
#Give similar colours for the same sequencing center
cols = c("cadetblue","cyan", "darkblue","blue","magenta","orange","yellow","brown","black",  "grey80","grey70","pink") 
palette(cols)
physeqListV13_SC = lapply(physeqListV13, function(x){
  sample_data(x)[["RUNCENTER"]] = factor(sample_data(x)[["RUNCENTER"]], levels = SClevels, labels = SClevels)
  x
})
phyList = c(AGP = AGphylo, physeqListV13_SC)
resListRCM2 = lapply(realNames, function(x){
  tmp = resListRCM[[x]]
  tmp$physeq = prune_samples(rownames(tmp$X),prune_taxa(colnames(tmp$X), phyList[[x]]))
  tmp
}) 

names(resListRCM2) = realNames

par(mfrow = c(3,4))
lapply(names(resListRCM2)[-1], function(x){plotRCM(resListRCM2[[x]], biplot=FALSE, samColour = "sex", main=x)})
```

```{r HMPsamRuncenter, fig.cap="Sample plots of the RC(M) ordination of the HMP dataset. Colours indicate sequencing center. it is clear that there is a strong effect of sequencing center \\label{fig:samHMP}", purl = FALSE}
par(mfrow = c(3,4))
foo = lapply(names(resListRCM2)[-1], function(x){plotRCM(resListRCM2[[x]], biplot=FALSE, samColour = "RUNCENTER", main=x, libInset = c(1.1,-0.1))})
rm(resListRCM, resListRCM2)
```

\clearpage

##### The HMP dataset: sample plots after conditioning

Now let's filter out the impact of this variable using conditioning

```{r HMP filter out variable effect, purl=FALSE, fig.cap="Sample plots of the RC(M) ordination of the HMP dataset after conditioning. Colours indicate sequencing center. Even when fitted with a rough grouping of sequencing centers, it is clear that we have conditioned out its effect. \\label{fig:samHMP}"}
# antNarJob = mcparallel(  RCM(RCMhmpFilt[["Anterior.nares"]], distribution="NB", k = K, nleqslv.control= nleqslv.control, maxItOut=5e3, prevCutOff=0.01, colWeights="marginal", rowWeights = "uniform", marginEst = "MLE", round=TRUE, confounders = "RUNCENTER2"))
# antNar = mccollect(antNarJob, wait=FALSE)[[1]]
# save(antNar, file="antNarConfounders.RData")

if(!file.exists(file="/home/stijn/PhD/Biplots/filtResHMP.RData")){
  K=2
nleqslv.control = list()
#Group the runcenters together

RCMhmpFilt = lapply(physeqListV13, function(x){
  runcenters = data.frame(sample_data(x))[ "RUNCENTER"]
runcenters[["RUNCENTER"]][runcenters[["RUNCENTER"]] %in% c("BCM","BI,BCM","BCM,BI","BCM,WUGC", "BCM,JCVI")] = "BCM"
runcenters[["RUNCENTER"]][runcenters[["RUNCENTER"]] %in% c("JCVI","JCVI,BI","JCVI,WUGC")] = "JCVI"
runcenters[["RUNCENTER"]][runcenters[["RUNCENTER"]] %in% c("WUGC,JCVI","WUGC","WUGC,BCM")] = "WUGC"
sample_data(x) = data.frame(sample_data(x) ,RUNCENTER2 = runcenters[[1]])
x})
filtResJob = mcparallel(lapply(RCMhmpFilt, function(x){
  RCM(x, distribution="NB", k = K, nleqslv.control= nleqslv.control, maxItOut=5e3, prevCutOff=0.01, colWeights="marginal", rowWeights = "uniform", marginEst = "MLE", round=TRUE, confounders = "RUNCENTER2")
  }))
filtRes = mccollect(filtResJob, wait=FALSE)[[1]]
save(filtRes, file="/home/stijn/PhD/Biplots/filtResHMP.RData")
} else {load(file="/home/stijn/PhD/Biplots/filtResHMP.RData")}
par(mfrow = c(3,3))
foo = lapply(names(filtRes), function(x){plotRCM(filtRes[[x]], biplot=FALSE, samColour = "RUNCENTER", main=x, libInset = c(-1.9,-0.1))})
# lapply(names(filtRes), function(x){plotRCM(filtRes[[x]], biplot=FALSE, samColour = "RUNCENTER2", main=x, libInset = c(-0.9,-0.1))})
par(mfrow = c(1,1))
rm(filtRes, physeqListV13_SC, phyList)
```


```{r filterTwoConfounders, purl = FALSE, eval = FALSE}
#With two confounders
confTestJob = mcparallel(RCM(RCMhmpFilt[["Anterior.nares"]], distribution="NB", k = K, nleqslv.control= nleqslv.control, maxItOut=5e3, prevCutOff=0.01, colWeights="marginal", rowWeights = "uniform", marginEst = "MLE", round=TRUE, confounders = c("sex","RUNCENTER2")))
confTest = mccollect(confTestJob, wait=FALSE)[[1]]
save(confTest, file="/home/stijn/PhD/Biplots/confTestHMP.RData")
```

\clearpage

##### The AGP dataset: sample plots

```{r AGPplots, purl = FALSE, eval = FALSE}
par(mfrow = c(3,3))
AGPvars = c("IBD","SEX","PREGNANT","LACTOSE","ASTHMA","DIABETES", "DIET_TYPE","COUNTRY")
sapply(AGPvars, function(x){
plotRCM(resListRCM[["AGP"]], biplot=FALSE, samColour = x, main=x, libInset = c(-3.9,-0.1), libLeg=TRUE)
})
```

The AGP dataset is presumably too noisy to find much of a signal

```{r fitting process, purl = FALSE, eval = FALSE}
lapply(names(resListRCM), function(x){with(resListRCM[[x]],{ plot(main=x, rowRec[2,1,1:iter[1]])})})
```

##### The Zeller data: sample plots

```{r load zeller RC(M) NB results, purl = FALSE, include = FALSE}

zellers = c("zellerMeta","zeller16S")
# zellerRCM  = lapply(paste(zellers, "RCM.RData"), function(x){
#   load(x)
#   RCMres
# })
if(!file.exists("zellerSphyRCM.RData")){
zellerSphyRCM = RCM(zellerSphy, k=2, round = TRUE) # Less than one minute!
save(zellerSphyRCM, file="zellerSphyRCM.RData")
} else {load("zellerSphyRCM.RData")}

if(!file.exists("zellerMphyRCM.RData")){
zellerMphyRCM = RCM(zellerMphy, k=2, round = TRUE, prevCutOff = 0.05, minFraction = 0.15) # Less than one minute! More trimming needed here
save(zellerMphyRCM, file="zellerMphyRCM.RData")
} else {load("zellerMphyRCM.RData")}

zellerRCM  =list(zellerMphyRCM, zellerSphyRCM)
names(zellerRCM) = zellers

sapply(zellerRCM,function(x){x$converged}) # All converged
sapply(zellerRCM,function(x){x$runtime})
sapply(zellerRCM,function(x){x$iter})

cols = c("darkblue","orange","darkgreen", "brown","black", "blue","magenta","cadetblue","cyan", "yellow", "grey80","grey70","pink") 
palette(cols)
```

```{r zellerOldPlots, purl = FALSE, eval = FALSE, echo = FALSE}
par(mfcol = c(2,3))
sapply(c("Country","Diagnosis","Gender"), function(y){
sapply(names(zellerRCM), function(x){plotRCM(zellerRCM[[x]], biplot=FALSE, samColour = y, main=paste(x,y, collapse="_"))})
})
```

For the 16S data we see signals for Country and Cancer diagnosis. For the metagenomics data we do not see a trend. Notice we are doing indirect gradient analysis again, but this is just to look for confirmation that the unconstrained analysis is detecting true groups. 

```{r plotZellerSphy, purl = FALSE, eval = FALSE}
lapply(c("Country","Diagnosis","Gender", "BMI","Age"), function(x){
  plot.RCM(zellerSphyRCM, colour = x)
})
```

```{r plotZellerSphyExec, purl = FALSE}
lapply(c("Country","Diagnosis"), function(x){
  plot.RCM(zellerSphyRCM, samColour = x)
})
rm(zellerRCM, zellerSphyRCM, zellerMphyRCM)
```

```{r zeller country vs diagnosis, purl = FALSE, eval = FALSE}
#Are country and diagnosis related?
with(sample_data(zellerMphyRCM$physeq),table(Country, Diagnosis))
#Yes, only cancer patients from Germany!
```

\clearpage

##### The Zeller data: Constrained RC(M)

```{r zellerConstrained, purl =FALSE}
if(!file.exists("zellerSphyRCMconstr.RData")){
zellerSphyRCMconstrLin = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "linear") # Use smallest possible cndtol!
zellerSphyRCMconstrQuad = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "quadratic")
zellerSphyRCMconstrNonP = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "nonparametric")
save(zellerSphyRCMconstrLin, zellerSphyRCMconstrQuad, file="zellerSphyRCMconstr.RData")
} else {load("zellerSphyRCMconstr.RData")}  #It looks like we have to drop the centering restriction on the alphas. It forces one continuous variable to be 0, two to be each others opposite
```

```{r ZellerConstrainedBiplot, purl=FALSE, eval = FALSE}
plot.RCM(zellerSphyRCMconstrLin)
plot.RCM(zellerSphyRCMconstrLin, samColour = "Diagnosis")
plot.RCM(zellerSphyRCMconstrLin, samColour = "Gender")
plot.RCM(zellerSphyRCMconstrLin, samColour = "Country")
plot.RCM(zellerSphyRCMconstrLin, samColour = "Age")
#apply(zellerSphyRCMconstr$NB_params^2, c(1,3) ,sum)

plot.RCM(zellerSphyRCMconstrQuad, samColour = "Diagnosis", Palette = rainbow(5))
```

```{r ZellerConstrainedBiplotExec, purl=FALSE, fig.cap="Constrained RC(M) analysis on zeller data with linear response functions. Origins of species arrows indicate points of no departure from independence. \\label{fig:triZellerDiagnosis}"}
plot.RCM(zellerSphyRCMconstrLin, samColour = "Diagnosis", plotTaxa = TRUE)
```

```{r ZellerConstrainedBiplotExec2, purl=FALSE, fig.cap="Constrained RC(M) analysis on zeller data with linear response functions. Origins of species arrows indicate points of no departure from independence. \\label{fig:triZellerAge}"}
plot.RCM(zellerSphyRCMconstrLin, samColour = "Age", plotTaxa = TRUE)
rm(zellerSphyRCMconstrLin)
```

Cancer turns out to be the main driver of explained variability in the first dimension, country in the second

```{r AGPconstr, eval=FALSE}
##### Constrained analysis of the AGP data
AGPconstr = RCM(AGphylo, k=2, round = TRUE, covariates = "all")
```

##### Turnbaugh et al. (2008)

Also try the Turnbaugh (2008) dataset: lean and obese twins

```{r turnTwins, eval = FALSE, purl = FALSE}
if(!file.exists("turntwinsRCm.RData")){
  load(file = "/home/stijn/PhD/Datasets/TurnbaughObeseTwins/TurnTwins.RData")
turntwinsRCm = RCM(TurnTwins, k=3)
#turntwinsRCmConstr = RCM(TurnTwins, k=3, covariates = c("Family_p","Status_p")) #Not enough variables for constrained analysis
save(turntwinsRCm, file = "turntwinsRCm.RData")
} else {load("turntwinsRCm.RData")}
plot.RCM(turntwinsRCm, colour = "Family_p")
plot.RCM(turntwinsRCm, colour = "twin")
plot.RCM(turntwinsRCm, colour = "Status_p")
rm(turntwinsRCm)
```

##### Turnbaugh et al. (2009)

15 gnotobiotic mice were inoculated with human feces, and for one group the diet was switched to a Western diet after one month. Then a second generation of mice was inoculated with cecal samples from the previous groups, and here also diet was varied.

- Diet: current diet
- Generation_p: "Recipient1" if second generation, otherwise diet of cecal samples with which they were inoculated
- Time: age of mouse at sampling
- DGS: Diet and previous diet
- DTG: combines sampling diet, time, and previous generation (if any) 

```{r turnmice, purl = FALSE}
if(!file.exists("turnMiceRCM.RData")){
  load(file = "/home/stijn/PhD/Datasets/Humanized mouse/HumMicePhy.RData")
HumMiceRCM = RCM(HumMicePhy, k=3)
HumMiceRCM = RCM(HumMicePhy, k=3, covariates = c("Diet","Generation_p","DTG","DTS"))
save(HumMiceRCM, file = "turnMiceRCM.RData")
} else {load("turnMiceRCM.RData")}
```

```{r humMicePlot, purl = FALSE, eval = FALSE}
plot.RCM(HumMiceRCM) # Two or three distinct groups
plot.RCM(HumMiceRCM, samColour = "Diet") #Diet is a main factor
plot.RCM(HumMiceRCM, samColour = "Individual") #Individuals play less of a role
plot.RCM(HumMiceRCM, samColour = "Generation_p") #Microbiome changes quickly with feed pattern
plot.RCM(HumMiceRCM, samColour = "Time") #Later samples also form a cluster
```

```{r humMicePlotExec, purl = FALSE}
plot.RCM(HumMiceRCM, samColour = "Diet", main =  "Humanized mice") #Diet is a main factor
plot.RCM(HumMiceRCM, samColour = "Time", main =  "Humanized mice") #Later samples also form a cluster
rm(HumMiceRCM)
```

```{r humMicePlot13D, purl = FALSE, eval = FALSE}
plot.RCM(HumMiceRCM, samColour = "Diet", Dim = c(1,3)) #Diet is a main factor
plot.RCM(HumMiceRCM, samColour = "Individual", Dim = c(1,3)) #Individuals play less of a role
plot.RCM(HumMiceRCM, samColour = "Generation_p", Dim = c(1,3)) #Microbiome changes quickly with feed pattern
plot.RCM(HumMiceRCM, samColour = "Time", Dim = c(1,3)) #Later samples also form a cluster
plot.RCM(HumMiceRCM, samColour = "DTG", Dim = c(1,3))
plot.RCM(HumMiceRCM, samColour = "DGS", Dim = c(1,3))
```

##### Dr. armpit data

The dataset contains 16S sequencing data from armpit swabs of volunteers.

```{r ArmPit, purl =  FALSE}
if(!file.exists("armpit.RData")){
load("Armpit/okseldata.rda") 
load("Armpit/div.rda") 
load("Armpit/odor.rda")
db<-data.frame(odor,info,food,washing)
db$subject = db$subject.1
db$subject.2 = db$subject.1 = NULL
load("Armpit/okseldata_20160229.rda") # data in object counts
Counts<-counts[,-903]
rownames(Counts) = rownames(db) = db$subject
OTUa = Counts[,-903]
armpit = phyloseq(otu_table(apply(as.matrix(OTUa), 1:2, as.numeric), taxa_are_rows =FALSE), sample_data(data.frame(db)))
save(armpit, file = "armpit.RData")
} else {load(file = "armpit.RData")}

if(!file.exists(file = "armpitRCM.RData")){
  armpitRCM = RCM(armpit, k=3)
  armpitRCMconstrLin = RCM(armpit, covariates = sample_variables(armpit)[!sample_variables(armpit) %in% c("subject")], k=3, responseFun = "linear")
  armpitRCMconstrQuad = RCM(armpit, covariates = sample_variables(armpit)[!sample_variables(armpit) %in% c("subject")], k=3, responseFun = "quadratic")
  save(armpitRCM, armpitRCMconstrLin, armpitRCMconstrQuad, file = "armpitRCM.RData")
} else {load(file = "armpitRCM.RData")}
```

```{r ArmpitPlot1, purl = FALSE}
plot(armpitRCM)
plot(armpitRCM, samColour = "shaving")
plot(armpitRCM, samColour = "odor_panel")
plot(armpitRCM, samColour = "gender")
plot(armpitRCM, samColour = "age")
plot(armpitRCM, samColour = "antimicrobial")
rm(armpitRCM)
```

##### CMET data

This is a longitudinal dataset on growth in a water cooling system. It has the unique property of providing relative counts trhough 16S sequencing that can be multiplied by absolute counts from flow cytometry to get truly absolute abundances that reflect the number of cells. This type of data does not qualify for the term _compositional data_ anymore. The advent of these kind of approaches support the use of count models rather than compositional data approaches.

```{r readInCMET, purl = FALSE, eval = FALSE}
if(!file.exists(file = "ata.RData")){
  library(openxlsx)
  samData = read.xlsx("/home/stijn/PhD/Datasets/CMET/Metadata.xlsx")
  rownames(samData) = samData$sample_title
  samData$Reactor.cycle = as.character(samData$Reactor.cycle)
  samData$cellDensity = samData$`Cell.density.(cells/mL)`
  samData$`Cell.density.(cells/mL)` = NULL
  otuTab = read.csv("/home/stijn/PhD/Datasets/CMET/OTU_table_raw.csv", sep = ";")
  rownames(otuTab) = otuTab$Sample_ID
  otuTab$Sample_ID = NULL
  #Select only samples present in samData
  otuTab = otuTab[as.character(samData$sample_title),]
  #Multiply relative abundances by cell counts in samData
  otuTab = round(as.matrix(otuTab)/rowSums(as.matrix(otuTab))*samData$cellDensity)
  taxTab = read.csv("/home/stijn/PhD/Datasets/CMET/tax_table_raw.csv", sep = ";")
  rownames(taxTab) = taxTab$X
  taxTab$X = NULL
  samData = samData[rowSums(otuTab)>0,]
  taxTab = taxTab[ colSums(otuTab)>0,]
  otuTab = otuTab[rowSums(otuTab)>0, colSums(otuTab)>0] #Number of zeroes is not related to rounding
  CMETwater = phyloseq(otu_table(otuTab, taxa_are_rows = FALSE), sample_data(samData), tax_table(as.matrix(taxTab)))
  save(CMETwater, file = "CMETdata.RData")
}else {load(file = "CMETdata.RData")}
```

```{r analysCMET, purl = FALSE}
if(!file.exists(file = "CMETRCM.RData")){
  CMETRCM = RCM(CMETwater, k = 3)
  envVarsCMET = c("ph","temp","Conductivity","cellDensity", "Reactor.phase", "collection_date", "Timepoint", "Reactor.cycle")
  CMETRCMlin = RCM(CMETwater, k = 3, responseFun = "linear", covariates = envVarsCMET)
  CMETRCMquad = RCM(CMETwater, k = 3, responseFun = "quadratic", covariates = envVarsCMET)
  save(CMETRCM, CMETRCMlin, CMETRCMquad, file = "CMETRCM.RData")
} else {load(file = "CMETRCM.RData")}
```

```{r plotCMET, purl = FALSE}
plot(CMETRCM)
plot(CMETRCM, samColour = "ph")
plot(CMETRCM, samColour = "temp")
plot(CMETRCM, samColour = "Conductivity")
plot(CMETRCM, samColour = "cellDensity")
plot(CMETRCM, samColour = "Reactor.phase")
plot(CMETRCM, samColour = "Timepoint")
plot(CMETRCM, samColour = "collection_date")
plot(CMETRCM, samColour = "Reactor.cycle")
```

We clearly have two groups in the second dimension, partly explained by the _Reactor.phase_ and _collection.date_ variables. In the first dimension we have a strong time signal.

# Competing methods

## Unconstrained RC(M)

### PCoA/MDS

Principal coordinates analysis (PCoA) or multidimensional scaling (MDS) are dimension reduction techniques specifically for samples. The consist in calculating dissimilarity matrices with dissimilarities between samples based on some chosen dissimilarity measure. These dissimilarity matrices are than decomposed through eigenvalue decomposition, and the first two dimensions are plotted. More formally, for a n-by-p data matrix $\mathbf{X}$, all pairwise distances between samples $i$ and $i'$ $D_{ii'} = d(\mathbf{X_{i.}}, \mathbf{X_{i'.}})$ are calculated and summarized in a n-by-n dissimilarity matrix $\mathbf{D}$. If the dissimilarity measure is a true distance, then $d_{ii'} = d_{i'i}$ and $\mathbf{D}$ is symmetric. Next $\mathbf{D}$ is squared and undergoes double centering:

$$B = -1/2 J D^2 J$$

with $\mathbf{J} = \mathbb{I} - \frac{1}{n}\mathbf{11'}$ a centering matrix. This transformed matrix __B__ then undergoes eigendecomposition into

$$ \mathbf{B} = \mathbf{V}^t\boldsymbol{\Lambda}\mathbf{V} $$
  with the eigenvectors of $\mathbf{D}$ in the columns of $\mathbf{V}$ and $\boldsymbol{\Lambda}$ a diagonal matrix with eigenvalues of $\mathbf{D}$ on the diagonal. $\mathbf{v}_1\lambda_1$ can then be plotted _vs_ $\mathbf{v}_2\lambda_2$ in a 2D scatterplot to represent distances between the samples optimally in 2D. In the calculation of $\mathbf{D}$, the information on the individual taxa is lost. Taxa scores can be calculated by weighted averaging of the sample scores for each taxon, weighted by the taxon abundances in these samples. The taxon score of taxon _j_ in dimension 1 then becomes
   
   $$t_{j1} = \sum_{i=1}^n v_{i1}*\frac{x_{ij}}{x_{.j}}$$
   
  This is a two step approach: the taxon scores are only calculated post-hoc and not as part of the eigendecomposition.
   
  In the following paragraphs we discuss a number of popular distance measures for microbiome data that are commonly used for PCoA.
   
#### Bray-Curtis dissimilarity

This distance measure was invented in 1957 in the context of forest ecology. It is calculated as

$$BC_{12} = 1-\frac{2*S_{12}}{S_1 + S_2}$$

with $S_{12}$ the counts or relative abundances of all species observed in \textbf{both samples}, and $S_{i}$ the counts or relative abundances of species only observed in sample $i$. This dissimilarity measure varies between 1 when no species are shared (maximal dissimilarity), and 0 when all species shared, and the community composition is identical. It is not a distance measure since it does not fulfill the triangle inequality $d(a,b) \leq d(a,c) + d(b,c)$. For use in microbiome research see e.g. Barfod _et al._, 2017 and many others. You can apply it to eihter absolute or relative counts, although the latter appears to be the most popular. In Callahan _et al._ (2016) they also apply it to log-transformed counts, even though they note that this leads to correlations with the library sizes.

#### UniFrac distance

The UniFrac distance was developed specifically for the microbiome by Lozupone and coworkers (2005), and includes phylogenetic information in the distance calculation. It is defined as the fraction of the phylogenetic tree length that leads to species found in only one of the two samples. As opposed to Bray-Curtis dissimilarity, it is a true distance metric.

##### Weighted UniFrac

To account for differences in sequencing depth, in 2007 weighted UniFrac was proposed. This weights every branch by the absolute difference in relative abundances the species of both samples as follows:

$$d_{UniFrac}(1,2) = \sum_{i=1}^b l_i |\frac{x_{1b}}{x_{1.}} - \frac{x_{2b}}{x_{2.}}|$$
with $x_{zb}$ the total abundance of species in sample _z_ pertaining to branch _b_. This means that also branches with species from both samples can contribute to the distance between the samples, if their compositions are uneven.

#### Jenssen-Shannon distance
   
We include the Jenssen-Shannon distance because of its role in the enterotypes publication (Arumugam _et al._, 2011). It is the square root of the Jenssen-Shannon _divergence_, a symmetrized version of the Kullback-Leibler divergence.

$$d_{Jenssen-Shannon}(a,b) = \sqrt{0.5 \sum_{j=1}^p x_{aj} log \frac{x_{aj}}{\frac{x_{aj}+x_{bj}}{2}} + 0.5 \sum_{j=1}^p x_{bj} log \frac{x_{bj}}{\frac{x_{aj}+x_{bj}}{2}}}$$

It is a true distance.

#### Hellinger distance

This came up during our own researhc into a version of correspondence analysis that would accurately model the variance of our count data. Still it is designed for distances between distributions rather than between real data ratios.

The Hellinger distance divides by the square root of the observation as means of normalization:

$$\frac{X}{\sqrt{Var(X)}} \approx \frac{X}{\sqrt{X}} = \sqrt{X}$$

The rationale is that each observation x is the best estimate of its expectation E(X), and under the Poisson model also of its variance. Still also these plots are susceptible to differences in library sizes. See Rao (1997), who suggests this method as an alternative to correspondence analysis. The drawback of correspondence analysis is that the estimate for the expected abundances used to standardize departures from independence $X-E$ is based on the taxon sums, and thus on other samples. With the Hellinger distance only, the sample itself is used to estimate the expected abundances, namely by the count itself. It is defined as 

$$d_{hellinger}^2(a,b) = \sum_{j=1}^p (\sqrt{\frac{x_{aj}}{x_{a.}}}-\sqrt{\frac{x_{bj}}{x_{b.}}})^2 $$

. It does works on the relative abundances. It has been used in publications on microbiome, e.g. Kreisinger _et al._, 2015, Hollister _et al._ 2015, Vital _et al._, 2015.

#### Transformations

In Callahan _et al._ (2016), which serves as a manual for the popular _phyloseq_ package, it is suggested to first log-transform the counts(after adding pseudocount 1) as a variance stabilizing transformation, prior to applying methods like PCoA

#### Related material

   * See Warton _et al._ (2012): Distance-based multivariate analyses confound
location and dispersion effects
   * See technical report bio BioRankings: "Dealing with High-Dimensional Data". Pairwise distances all become similar in high dimensions.
   
#### Non-metric multidimensional scaling

Non-metric multidimensional scaling (NMDS) can be considered the little brother of MDS, since it also represents distance matrices in lower dimensions. Still there are some important differences: 

 - NMDS is fitted numerically instead of through singular value decomposition
 - The number of axes is chosen beforehand. There is no ordering in the axes, the solution depends on the number of axes chosen
 - It makes no assumption on the distribution of the data, in contrast to PCA which assumes linear relationships and CCA which assumes unimodal ones.
 
 The ordination is fitted iteratively, by minimizing the stress between the original distances and the distances of the actual ordination. See the tutorial by Steven M. Holland for more information.

### PCA

#### PCA on ranks

### Correspondence analysis

This is the classical, statistical way of analysing contingency tables of count data. For discussion see the beginning of this document.

### CoDa

In recent years there have been many publications on analysing microbiome data as compositional. Gloor and Reid (2016) and Gloor _et al._ propose a method to extend this philosophy to biplots. In a frist step, zero counts are replaced by a non-zero pseudocount, either 0.5 or through a Bayesian method. Next the data are normalized to realtive abundances, divided by the geometric mean and log-transformed. This is called the central logratio transform (clr). PCA is applied to these transformed data (through eigendecomposition of the variance-covariance matrix) and sample scores mutliplied by eigenvalues are plotted. Row scores can be added as weighted sample means.

Theoretical concerns I have about this method are the following:

 - Heteroscedasticity of count data is not addressed. Authors claim this should be handled "downstream".
 - Negative correlations due to compositionality might be swamped by true biological correlations
 - The library sizes are not always of merely technical nature, and very often measures of biomass or cell count are including. This questions the compositional nature of these data

### Bayesian Nonparametric Ordination 

(Ren _et al._ (2016))
 
```{r CoDA}
# The compositional approach:
 # Zeroes are fixed somehow
 # log-ratio transform of relative abundances
 # PCA on transformed values
# Main concern is that this approach ignores the fact that we have count data that are heteroscedastic

# This function is build after the supplementary files of Gloor and Reid, 2016. Only the abundance filtering is made less severe
CoDa = function(dat, abCutOff = 1e-6, scale = FALSE){
  require(zCompositions)
  classDat = class(dat) #The class provided
  
  ##The count data##
  if (classDat=="matrix"){
  X = dat
  }else  if(classDat=="phyloseq"){
  X = if (taxa_are_rows(dat)) t(otu_table(dat)@.Data) else otu_table(dat)@.Data
  } else {stop("Please provide a matrix or a phyloseq object! \n")}
  
Xzero <- as.matrix(cmultRepl(X = X, method="CZM", output="counts")) #Correct zeroes in a Bayesian way

# convert to proportions 
Xnorm = Xzero/rowSums(Xzero)

#Trim on abundance in any sample
Xtrim = Xnorm[,apply(Xnorm,2,min)>abCutOff]

# make our compositional dataset
Xcrl <- t(apply(Xtrim, 1, function(x){log(x) - mean(log(x))}))

#Column center
Xsca = scale(Xcrl, center = TRUE, scale = FALSE) #Shouldn't this be scaled? It is not the default, and was not done by Gloor and Reid

# Apply PCA
Svd = svd(Xsca)
rowScores = Svd$u %*% diag(Svd$d) #=Xsca %*% Svd$v #%*% diag(Svd$d)
colScores = Svd$v #scale = 0 in biplot
rownames(rowScores) = rownames(Xsca)
rownames(colScores) = colnames(Xsca)

# rowScores2 = Svd$u
# colScores2 = Svd$v %*% diag(Svd$d) #scale = 1 in biplot

list(rowScores = rowScores, colScores = colScores)
}
```

```{r nonparametricBayesian}
# A bayesian approach, this is quite slow
#data(JRphylo) # A vaginal dataset
nonParamBayes = function(dat, hyper = list( nv = 3, a.er = 1, b.er = 0.3, a1 = 3, a2 = 4, m = 10, alpha = 10, beta = 0 )){
  require(DirFactor)
  classDat = class(dat) #The class provided
  ##The count data##
  if (classDat=="matrix"){
  X = dat
  }else  if(classDat=="phyloseq"){
  X = if (taxa_are_rows(dat)) t(otu_table(dat)@.Data) else otu_table(dat)@.Data
  } else {stop("Please provide a matrix or a phyloseq object! \n")}
  DirFactor(t(X), hyper, step = 100)
}
```
   
## Constrained

  - Constrained PCoA?
  - Constrained correspondence analysis (CCA)

# Method evaluation

After implementing the method and proving its concept, it is time to compare it with existing methodologies. We need to define a way to obtain datasets, and a way to evaluate the ordinations in a high-throughput way. For the simulation we will need multiple repetitions to prove the superiority of our method, so we need an automatic evaluation of the clustering accuracy.

## overall vs. within distance

A first intuitive approach would be to compare the within cluster distances on the plot to the distances to the origin. Clusters are defined as samples generated with the same taxon distribution, either parametrically or non-parametrically. Say we have $m = 1,...,l$ groups with (multidimensional) means $\mathbf{M}_m$.  The within distance is then defined as $d_{m, within} = \frac{1}{n_m}\sum_{i=1}^{n_m} d_{euclidean}(\mathbf{M}_m, Y_i)$, with $d_{euclidean}$ the euclidean distance. The overall distance of a group then equals $d_{m, overall} = \frac{1}{n_m}\sum_{i=1}^{n_m} d_{euclidean}(\mathbf{0}, Y_i)$ with $\mathbf{0}$ the origin. This calculation can be done for each dimension separately or in the multidimensional space.

We could perhaps look for the distance based method that maximizes $\frac{d_{m, overall}}{d_{m, within}}$? This is related to Ward's minimum variance method, but with scaling to the overall distance. This takes differences in the scales of solutions of different methods into account.

This approach can only be applied to the row scores (samples), since there the importance parameters are multiplied and thus distances are represented correctly.

```{r within vs. between distances}
# A function to calculate these within and overall distances
distanceFun = function(coord, clusters, Dim = c(1,2)){
  # @param coord: an nxk matrix of coordinates
  # @param clusters: a factor of length n with cluster memberships
  # @param Dim: The dimensions to look at, defaults to the first two
  
  # @return: a list with
    # -withinDist: within distances
    # -overalDist: the n overal distances
    # -ratio: 
  
  coord = coord[,Dim, drop=FALSE]
  
  withinDist = tapply(seq_len(nrow(coord)), clusters, function(x){
    coordsX = coord[x,]
    centerX = colMeans(coordsX)
    dists = apply(coordsX, 1, function(y){
      dist(rbind(y, centerX))
    })
    return(dists)
  })
  
  overalCenter = colMeans(coord)
  overalDist = apply(coord, 1, function(y){
    dist(rbind(y, overalCenter))
  })
  
  ratio = mean(unlist(sapply(unique(clusters), function(x){overalDist[clusters==x]/withinDist[[x]]})))
  list(withinDist = withinDist, overalDist =  overalDist, ratio = ratio)
  }
```

## Silhouette

See "Silhouettes: a Graphical Aid to the Interpretation and Validation of Cluster Analysis", Rousseeuw, 1987. The silhouette can be calculated with any distance measure, but we will just use the euclidean distance again. For each point $i$, calculate the distance to each other point in the ordination, and average these distances within the cluster. Call $a(i)$ the average distance to its own cluster and $b(i)$ the smallest of the average distances to the other cluster. The the silhouette of observation i $s(i)$ is defined as:

$$s(i) = \frac{b(i) - a(i)}{max\big(a(i), b(i)\big)}$$

The silhouette can take values between +1 for optimal separation and -1 for wrong classification. If a poin i close to the centroid of its own cluster but very far from all the others the methods has performed well and is rewarded with a high silhouette. We can average the silhouttes over the ordination or jus look at their distribution to compare the ordination methods.

```{r silhouette}
silhouette = function(coord, clusters, Dim = c(1,2)){
  # @param coord: an nxk matrix of coordinates
  # @param clusters: a factor of length n with cluster memberships
  # @param Dim: The dimensions to look at, defaults to the first two
 
  # @return a vector of n silhouettes
  
  coord = coord[,Dim, drop=FALSE]
  
  distMat = as.matrix(dist(coord)) # all distances
  avDists = apply(distMat, 1 , function(x){
    tapply(x[x!=0], clusters[x!=0], mean) #Exclude observation itself
  }) # Average distances within clusters
  ais = sapply(seq_along(clusters), function(i){
    avDists[clusters[i], i]
  }) #Average distance to own cluster
  bis = sapply(seq_along(clusters), function(i){
    min(avDists[clusters[i]!=levels(clusters), i])
  }) #Average distance to nearest cluster
  
  silhouette = mapply(ais, bis, FUN = function(a,b){(b-a)/max(a,b)}) #The silhouette
  
  return(silhouette)
}
```

## Correlation with library sizes and abundances

The motivating problem to develop the whole method was to find an approach that would not show correlation between the row scores and the library sizes, which is a technical artefact.

Pearson correlations of row scores with library sizes could be another criterion to evaluate the quality of the biplot. Equivalently correlations with average relative abundance can be investigated.

```{r libSizeCorrelation}
libCor = function(coord, margins, Dim = c(1,2)){
  if(length(margins)!=nrow(coord)) {return(rep(NA, length(Dim)))}
  apply(coord[, Dim, drop=FALSE], 2, cor, margins)
}
```

<!---
The between distance between groups m and m' is then defined as $d_{mm'} = d_{euclidean}(\mathbf{M}_m, \mathbf{M}_{m'})$ with $d_{euclidean}$ the euclidean distance.
--->

## Contribution of taxa to the separation of the clusters

For the methods that do yield taxon scores, we can also verify if the correct taxa contribute to the separation of the clusters. A useful measure in this case seems the relative contribution of the taxa with signal in the direction of the cluster centroid. It is calculated as the ratio of the average contribution in the direction of the centroid of taxa with signal to all the average contribution of all other taxa. With $\mathbf{H}_l$ the vector from the origin to the centroid of group $l$, $\mathbf{s}_{l,sig}$ the $n_{l,sig}$-by-k matrix with scores of the $n_{l,sig}$ taxa in group l with signal and $\mathbf{s}_{l,noSig}$ the $n_{l,noSig}$-by-k matrix with scores of the $n_{l,noSig}$ taxa without signal, the quantity we are looking for is:

$$ \frac{n_{l,noSig} \mathbf{s}_{l,sig}\mathbf{H}_l \mathbb{1}}{n_{l,sig} \mathbf{s}_{l,noSig}\mathbf{H}_l \mathbb{1}}$$
with $\mathbb{1}$ a column matrix of one of the appropriate dimension. However, this will very often lead to negative and/or very large ratios. Therefore we look at the ratio of the signal taxa to the average signal of the 50\% non-signal taxa with highest signal. Because the mean is very sensitive to outliers, we also include the median, whcih makes our method come out as best. it is very important to mention this in a publication and show both results!

```{r contrTaxa}
contrTaxa = function(rowMatPsi, colMat, groupFactor, idTax, sigFrac = 0.5, Dim = 1:3, upDown = FALSE, centerFun = median){
  #FIX ME! Hellinger!
  rowMatPsi = rowMatPsi[,Dim]
  colMat = t(colMat)[Dim,]
  if(upDown){
  idTaxNum = matrix(-1,nrow(idTax) , ncol(idTax))
  idTaxNum[idTax=="up"] = 1
  }
  centroids = matrix(unlist(tapply(seq_len(nrow(rowMatPsi)), groupFactor, function(i){
    colMeans(rowMatPsi[i,])
  })), nrow = ncol(rowMatPsi)) #The group centroids
  colnames(centroids) = levels(groupFactor)
  overalCenterSam = colMeans(rowMatPsi) # The overall sample centroid(not always c(0,0,0)!!)
  overalCenterTax = colMeans(colMat)
  contrib = sapply(seq_along(levels(groupFactor)), function(i){
        crossprod(colMat, centroids[, i])
  }) # The inner product of all species' arrows with the vectors locating the centroids
  sigContr = if(upDown) {sapply(seq_along(levels(groupFactor)), function(i){
    centerFun(contrib[match(rownames(idTax[[i]]), colnames(colMat), nomatch = 0),i]*idTaxNum[,i])
  })} else {
    sapply(seq_along(levels(groupFactor)), function(i){
    centerFun(contrib[match(idTax[[i]], colnames(colMat), nomatch = 0),i])
  })
  } # The mean/median contribution of the taxa with signal
  allContr = if(upDown) {sapply(seq_along(levels(groupFactor)), function(i){
    centerFun(
      apply(contrib[!colnames(colMat) %in% rownames(idTax[[i]]),],2,sort, decreasing = TRUE)[seq_len(round(sigFrac*(ncol(colMat)-length(idTax[[i]])))), ])
  })} else {
    sapply(seq_along(levels(groupFactor)), function(i){
    centerFun( apply(contrib[!colnames(colMat) %in% idTax[[i]],],2,sort, decreasing = TRUE)[seq_len(round(sigFrac*(ncol(colMat)-length(idTax[[i]])))), ])
  })
  } # The mean/median contribution of the 50% taxa with highest contribution among non-significant taxa
  ratios = sigContr/allContr # The ratio of these mean contributions
  mean(ratios) # Mean ratio
}
```

## Auxiliary functions

```{r extract row vectors}
# A function to perform all analyses except for RC(M)
otherAnalyses = function(RCMlist, unifrac = FALSE, treeList = NULL, cores = 1){
    if(unifrac) {
      RCMlist = mapply(SIMPLIFY = FALSE,RCMlist, treeList, FUN = function(x,tree){
        x$physeq = phyloseq(otu_table(x$X, taxa_are_rows = FALSE), tree)
        sample_names(x$physeq) = seq_len(nsamples(x$physeq))
        x
      })
  }
mclapply(mc.cores = cores, RCMlist, function(x){
  tmp = x$X; rownames(tmp) = seq_len(nrow(tmp))
  R = rowSums(tmp)
  E = outer(R, colSums(tmp)/sum(tmp))
  otuTab = otu_table(tmp, taxa_are_rows = FALSE)
  otuTabRel = otu_table(tmp/R, taxa_are_rows = FALSE)
  otuTabLog = otu_table(log(tmp+1), taxa_are_rows = FALSE)
  list(
    RCM = x, CA = caSVD(tmp), DCA = ordinate(otuTab, method = "DCA"), CoDa = CoDa(tmp),
    BC = ordinate(otuTab, method = "PCoA", distance = "bray"),
    BClog = ordinate(otuTabLog, method = "PCoA", distance = "bray"),
    JSD = ordinate(otuTab, method = "PCoA", distance = "jsd"),
    BCrel = ordinate(otuTabRel, method = "PCoA", distance = "bray"),
    BCrelNMDS = ordinate(otuTabRel, method = "NMDS", distance = "bray", k=3),
    Hellinger = svd(diag(1/sqrt(R)) %*% (sqrt(tmp)-sqrt(E))),
    UniFrac = if(unifrac) ordinate(x$physeq, method = "PCoA", distance = "unifrac") else NULL, 
    wUniFrac = if(unifrac) ordinate(x$physeq, method = "PCoA", distance = "wunifrac") else NULL
    )
})
} #JSD works on relative abundances by default
#---------------------------------------#
#A function to extract the row vectors multiplied by importance elements
extractRows = function(x){
tmpList = with(x, list(
    RCM = RCM$rMat %*% diag(RCM$psis),
    CA = CA$u %*% diag(CA$d),
    DCA = DCA$rproj,
    CoDa = CoDa$rowScores,
    BC = BC$vectors %*% diag(BC$values$Eigenvalues[seq_len(ncol(BC$vectors))]),
    JSD = JSD$vectors %*% diag(JSD$values$Eigenvalues[seq_len(ncol(JSD$vectors))]),
    BCrel = BCrel$vectors %*% diag(BCrel$values$Eigenvalues[seq_len(ncol(BCrel$vectors))]),
    BCrelNMDS = BCrelNMDS$points,
    UniFrac = if(is.null(UniFrac)) NULL else UniFrac$vectors %*% diag(UniFrac$values$Eigenvalues[seq_len(ncol(UniFrac$vectors))]),
    wUniFrac = if(is.null(wUniFrac)) NULL else wUniFrac$vectors %*% diag(wUniFrac$values$Eigenvalues[seq_len(ncol(wUniFrac$vectors))]),
    Hellinger = Hellinger$u %*% diag(Hellinger$d)
    ))
tmpList = tmpList[!sapply(tmpList, is.null)]
lapply(tmpList, function(y){
  colnames(y) = NULL; y
})
}
#---------------------------------------#
#A function to extract the column vectors. For PCoA and CoDa this uses weighted averaging
extractCols = function(x, row){
  tmpList = with(x, list(
    RCM = t(RCM$cMat),
    CA = CA$v,
    DCA = DCA$cproj,
    CoDa = CoDa$colScores,
    BC = weightedTaxonScores(RCM$X, row$BC),
    JSD = weightedTaxonScores(RCM$X, row$JSD),
    BCrel = weightedTaxonScores(RCM$X, row$BCrel),
    UniFrac = if(!is.null(UniFrac))  weightedTaxonScores(RCM$X, row$UniFrac) ,
    wUniFrac = if(!is.null(wUniFrac)) weightedTaxonScores(RCM$X, row$wUniFrac) ,
    BCrelNMDS = BCrelNMDS$species,
    Hellinger = Hellinger$v
    ))
  tmpList = tmpList[!sapply(tmpList, is.null)]
  rowNames = rownames(tmpList$RCM)
  lapply(tmpList, function(x){
    colnames(x) = NULL
    if(nrow(x)==length(rowNames)) rownames(x) = rowNames
    x
  })
}
#---------------------------------------#
# An auxiliary function to get weighted taxon scores
weightedTaxonScores = function(X, rowScores){
  crossprod(X %*% diag(1/colSums(X)),rowScores)
}
#-----------------------------------------#
# A functions to plot the correlations of the library sizes with the row scores, or of the abundances with the column scores
plotCor = function(scores, datList, Dims = 1:3, scoreDim = "rows", dataMat = TRUE, yloc = 0.2, groupMeth = factor(c(as.character(groupsMeth[names(groupsMeth) %in% names(scores[[1]])]), "Control"), levels = c(levels(groupsMeth), "Control"))){
if(!dataMat){
  datList = lapply(datList, function(x){x$dataMat})
}
  margins = switch(scoreDim, "rows" = rowSums(datList[[1]]), "columns" = colSums(datList[[1]]))
cor0 = lapply(Dims, function(Dim){
  rbind(
  mapply(scores, datList, FUN = function(x,z){
  margins = switch(scoreDim, "rows" = rowSums(z), "columns" = colSums(z))
  sapply(x, function(y){libCor(y, margins = margins, Dim = Dim)}) 
 }, SIMPLIFY = TRUE),
 "Control" = replicate(length(datList),cor(rnorm(length(margins)), margins)))}) #Include a control

cor0df = lapply(cor0, melt, value.name = "Correl", varnames = c("Method"))
par(mfrow = c(1,max(Dims)))
for (i in Dims){
  if(i==max(Dims)){
  parTmp = par(no.readonly = TRUE)
  par(mar = c(4,4,4,5))
}
data = orderDF(cor0df[[i]])
boxplot(Correl ~ Method, main = paste(
  switch(scoreDim, "rows" = "Correlations of library sizes \n with row scores of dimension",
         "columns" = "Correlations of species abundances \n with column scores of dimension"
           ),i), 
  data = data, ylim = c(-1,1), ylab = "Pearson correlation", col = groupMeth, las = 2)
meansCor = tapply(data$Correl, data$Method, mean)
points(meansCor, col="red", pch=18, cex = 1.6)
abline(h=0, lty = "dashed")
if(i==max(Dims)){
  par(parTmp)
}
}
addLegend(groupMeth = groupMeth, x = length(groupMeth), cex = 0.6)
par(mfrow = c(1,1))

}
#----------------------------------#
# A function to plot the silhouettes
plotSil = function(resListRows, groupFactor, method = NULL, groupMeth = droplevels(groupsMeth[names(groupsMeth) %in% names(resListRows[[1]])]),las = 2,...){
sil1  = lapply(resListRows, function(x){
  sapply(x, silhouette, clusters = groupFactor)
})
meanSil1 = sapply(sil1, function(x){
  colMeans(x)
})
meanSil1df = orderDF(melt(meanSil1, value.name = "Silhouette", varnames = c("Method")))

if(!is.null(method)){
  meanSil1df = subset(meanSil1df, meanSil1df$Method %in% method)
  }
boxplot(Silhouette~Method, main = "Mean silhouettes per Monte Carlo simulation by method", data = meanSil1df, col = groupMeth,las = las, ...)
meansSil = tapply(meanSil1df$Silhouette, meanSil1df$Method, mean)
points(meansSil, col="red", pch=18, cex = 1.6)
abline(h=0, lty = "dashed")
addLegend(groupMeth = groupMeth)
}
#---------------------------------#
distRatioPlot = function(resListRows, groupFactor, method = NULL, groupMeth = droplevels(groupsMeth[names(groupsMeth) %in% names(resListRows[[1]])]),log = FALSE, las = 2,...) {
dist = sapply(resListRows, function(x){
  sapply(x, function(y){distanceFun(y, clusters = groupFactor)$ratio})
})
meanDist1df = orderDF(melt(dist, value.name = "DistanceRatio", varnames = c("Method")))
if(!is.null(method)){
  meanDist1df = subset(meanDist1df, meanDist1df$Method %in% method)
}
boxplot(DistanceRatio~Method, main = "Mean ratios of overal to within distance by method", data = meanDist1df, col = groupMeth, ylim = if(log) NULL else c(0, ceiling(max(meanDist1df$DistanceRatio))), las = las,log = ifelse(log,"y",""),...)
meansDist = tapply(meanDist1df$DistanceRatio, meanDist1df$Method, mean)
points(meansDist, col="red", pch=18, cex = 1.6)
abline(h=1, lty = "dashed")
addLegend(groupMeth = groupMeth, yloc = 5)
}
#---------------------------#
#A function to make a scatterplot of a randomly selected MC isntance
randomScatter = function(list, Dim = c(1,2), col = "black"){
  parTmp = par(no.readonly = TRUE)
  if(length(list)<=9) {
    par(mfrow = c(3,3), mar = c(3,3,3,3))
  } else {
     par(mfrow = c(3,4), mar = c(3,3,3,3))
  }
ID = sample(size=1, seq_along(list))
obj = list[[ID]]
lapply(names(obj),function(x){plot(obj[[x]][,Dim], main=x, col = col, xlab = paste("Dim", Dim[1]),ylab = paste("Dim", Dim[2]), pty = "s")})
par(parTmp)
}
#------------------------------#
#A function to wrap the calculation of the contribution of the taxa
contrTaxaWrap = function(resListRows, resListCols, IDlist, groupFactor, upDown = FALSE, centerFun = median){
dat = mapply(resListRows, resListCols, IDlist, SIMPLIFY = TRUE, FUN = function(row, col, id){
  mapply(row, col, SIMPLIFY = TRUE, MoreArgs = list(idTax = id, groupFactor = groupFactor, upDown = upDown, centerFun = centerFun), FUN = contrTaxa)
})
as.data.frame(t(dat))
}
#-----------------------------#
# A function to plot the result of this as boxplots. Horizontal boxplots have been tried but the overview over the best methods is a bit lost
contrPlot = function(contr, method = NULL, las = 2, expFac = 1.2, groupMeth = droplevels(groupsMeth[names(groupsMeth) %in% names(contr)]),...) {
moltenContr = orderDF(melt(contr, variable.name = "Method", value.name = "contrRatio", id.vars = NULL))
if(!is.null(method)){
  moltenContr = subset(moltenContr, moltenContr$Method %in% method)
}
meansDist = tapply(moltenContr$contrRatio, moltenContr$Method, mean, na.rm = TRUE)
at = seq_along(meansDist)*expFac
boxplot(contrRatio~Method, main = "Mean ratios of contribution of DA taxa \n relative to 50% most contributing non-DA taxa by method", data = moltenContr, ylab = "Mean contribution ratio", col = groupMeth, las = las, at = at,...)
points(y = meansDist, x = at, col="red", pch=18, cex = 1.6)
abline(h=1, lty = "dashed")
addLegend(groupMeth = groupMeth, expFac = expFac)
}
#-----------------------------#
# A function to calculate correlation with the starting values
startFun = function(list){
  X = list$RCM$X
  R = rowSums(X)
  C = colSums(X)
  E = outer(R, C)/sum(R)
  svd1 = svd(diag(1/sqrt(R)) %*% (X-E) %*% diag(1/sqrt(C)))
  svd2 = svd(diag(1/R) %*% (X-E) %*% diag(1/C))
  svd3 = svd(diag(R^(-3/2)) %*% (X-E) %*% diag(C^(-3/2)))
  svd4 = svd(diag(R^(-3)) %*% (X-E) %*% diag(C^(-3)))
  svd5 = svd(diag(1/R^2) %*% (X-E) %*% diag(1/C^2))
  cMat = list$RCM$cMat
  rMat = list$RCM$rMat
  list(corRows1 = getCor(svd1$u, rMat), corRows2 = getCor(svd2$u, rMat), corRows3 = getCor(svd3$u, rMat), corRows4 = getCor(svd4$u, rMat), corRows5 = getCor(svd5$u, rMat) ,corCols1 = getCor(svd1$v, t(cMat)), corCols2 = getCor(svd2$v, t(cMat)), corCols3 = getCor(svd3$v, t(cMat)), corCols4 = getCor(svd4$v, t(cMat)), corCols5 = getCor(svd5$v, t(cMat)))
} 
getCor = function(a,b){
  sapply(1:ncol(b), function(x){
    cor(a[,x], b[,x])
  })
}
#--------------------------------#
# A function to reorder the methods
orderDF = function(df){
  df$Method = factor(df$Method, labels = labelsMeth, levels = levelsMeth, ordered = TRUE)
  df$groupMethod = groupsMeth[match(df$Method, levelsMeth)]
  droplevels(df)
}
#-------------------------------#
#A function to add the legend
addLegend = function(yloc = 0.2, groupMeth, x = length(groupMeth)+1, expFac = 1,...){
legend(legend = levels(groupMeth), col = factor(levels(groupMeth), levels = levels(groupMeth), ordered=TRUE),pch = 16, x = x*expFac, y = yloc, xpd = TRUE,...)
}
```

# Simulations

## Unconstrained RC(M)

As ways of generating test datasets we see three different options

  1) Parametric simulation with the negative binomial with known abundances. 
  2) Non-parametric simulation with SimSeq
  3) Real datasets with biological signal allegedly known
  
  Parametric simulation is needed as a first check but evidently favours our method (if we use the negative binomial distribution). SimSeq is a reasonably neutral tool, we will only use WMW to obtain the lfdr. Real datasets will need to complement our analysis but it may not be possible to evaluate them in a high-throughput way, this will need to happen empirically.
  
### Parametric simulations

#### Negative binomial

The negative binomial is the assumed distribution of our method. As a result we should outperform all other methods in this setting.

\clearpage

##### Null setting

In this setting we simulate all distributions from the same template, thus there is no signal in this dataset

```{r NBnull, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions
load("/home/stijn/PhD/Simulations/data/MLES.RData") #Load observed parameter values
rhoEst = unlist(rhoMLEs) #observed relative abundances, pooled
thetaEst = 1/unlist(phiMLEs) #observed dispersions abundances, pooled

#Library sizes
libSizesEst = unlist(sapply(physeqListV13, sample_sums)) #Resample
if(!file.exists(file = "RCMsim/resList0.RData")){
  if(!file.exists(file = "RCMsim/datList0.RData")){
datList0 = mclapply(mc.cores=1, 1:reps, function(i){
## Sample parameters
rhosSampled = sample(rhoEst, p)
thetasSampled = sample(thetaEst, p) #Sampled dispersions are the same in every population, this is the ideal case assumed by our method.
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes

## Create dataset
meanMat = outer(libSizesSampled, rhosSampled)
dataMat = matrix(rnbinom(prod(dim(meanMat)), mu = meanMat, size = matrix(thetasSampled,n,p,byrow = TRUE)),n,p)
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, thetasSampled = thetasSampled, rhosSampled = rhosSampled)
})
save(datList0,  file = "RCMsim/datList0.RData")
} else {load(file = "RCMsim/datList0.RData")}

# Results
  if(!file.exists("RCMsim/resList0RCM.RData")){
RCMresList0 = mclapply(mc.cores=4, datList0, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresList0, file = "RCMsim/resList0RCM.RData")
} else { load("RCMsim/resList0RCM.RData")}
resList0 = otherAnalyses(RCMresList0)
resList0Rows = lapply(resList0, extractRows)
resList0Cols = mapply(resList0, resList0Rows, FUN=extractCols, SIMPLIFY=FALSE)
save(resList0, resList0Rows,resList0Cols,  file = "RCMsim/resList0.RData")
} else {load(file = "RCMsim/resList0.RData")}
```

```{r reslist0Runningtimes, purl = FALSE, eval = FALSE, fig.cap = "Boxplot of running times of the RC(M) method in minutes for 100 Monte Carlo simulations. \\label{fig:runtimes}"}
boxplot(sapply(resList0, function(x){x$RCM$runtimeInMins}))
```

```{r resList0 sampleBiplot, purl = FALSE, include=FALSE}
#A sample biplot
randomScatter(resList0Rows)
# Our method and CoDa show only white noise, CA and PCoA show artificial clusters. Confounding mean and dispersion effects?
#randomScatter(resList0Rows, Dim = c(1,3))
```

```{r libSizeCorr0, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions. Red diamonds indicate means.\\label{fig:libCor1}"}
datList0 = lapply(resList0, function(x){x$RCM$X})
plotCor(resList0Rows, datList0, scoreDim = "rows", dataMat = TRUE, groupMeth = groupsMeth)
```

Hellinger distance, CoDa ana Bray-Curtis on absolut abundances show correlations with library sizes.

The taxa

```{r resList0cols sampleBiplot, purl = FALSE, include = FALSE}
#A sample biplot
randomScatter(resList0Cols)
# CA has outliers, CoDa and Hellinger reveal weird patterns
#randomScatter(resList0Cols, Dim = c(1,3))
```

```{r AbsCorr0, purl = FALSE, fig.cap = "Correlations of column scores with average relative abundances, results of 100 simulations shown for first two dimensions. Red diamonds indicate means.\\label{fig:libCor1}"}
plotCor(resList0Cols, datList = datList0,scoreDim = "columns", dataMat = TRUE)
rm(resList0Rows, resList0, datList0)
```

Only our RC(M) method has correlations at the level of random noise.

```{r Starting Values, purl = FALSE, eval = FALSE}
startList= lapply(resList0, startFun)
moltStart = melt(startList, value.name = "correlation")
names(moltStart) = c("correlation", "scores","run")
boxplot(correlation ~ scores, data = moltStart, las=2)
abline(h=0, col ="red", lty = "dashed")
#Squared weighing clearly gives the best results, although no fantastic correlations.
```

\clearpage

##### Simulate from different distributions

In this setting we will simulate data from 4 unrelated negative binomial distributions. Because of this clustering should be very good.

```{r NBunrelatedDistr, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions

#Load observed parameter values
rhoEst = unlist(rhoMLEs) #observed relative abundances, pooled
thetaEst = 1/unlist(phiMLEs) #observed dispersions abundances, pooled

groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactor1 = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor

#Library sizes

libSizesEst = unlist(sapply(physeqListV13, sample_sums)) #Resample
if(!file.exists(file = "RCMsim/resList1.RData")){
  if(!file.exists(file = "RCMsim/datList1.RData")){
datList1 = mclapply(mc.cores=1, 1:reps, function(i){
## Sample parameters
rhosSampled = sapply(1:nPop, function(x){tmp=sample(rhoEst, p); tmp/sum(tmp)})[,groupInd]
thetasSampled = sample(thetaEst, p) #Sampled dispersions are the same in every population, this is the ideal case assumed by our method.
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes

## Create dataset
meanMat = libSizesSampled * t(rhosSampled)
dataMat = matrix(rnbinom(prod(dim(meanMat)), mu = meanMat, size = matrix(thetasSampled,n,p, byrow = TRUE)),n,p)
list(dataMat = dataMat, group = groupFactor1, libSizesSampled = libSizesSampled, thetasSampled = thetasSampled, rhosSampled = rhosSampled)
})
save(datList1,groupFactor1,  file = "RCMsim/datList1.RData")
} else {load(file = "RCMsim/datList1.RData")}

# Results

  if(!file.exists("RCMsim/resList1RCM.RData")){
RCMresList1 = mclapply(mc.cores = 4, datList1, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresList1, file = "RCMsim/resList1RCM.RData")
} else {load("RCMsim/resList1RCM.RData")}
resList1 = otherAnalyses(RCMresList1)
resList1Rows = lapply(resList1, extractRows)
resList1Cols = mapply(resList1, resList1Rows,FUN = extractCols, SIMPLIFY = FALSE)
save(resList1,resList1Rows, resList1Cols, groupFactor1, file = "RCMsim/resList1.RData")
} else {load(file = "RCMsim/resList1.RData")}
```

```{r resList1 sampleBiplotE, purl = FALSE}
#A sample biplot
datList1 = lapply(resList1, function(x){x$RCM$X})
randomScatter(resList1Rows)
# It seems like our method performs better, only two groups do not get separated very well. But then again this is under the model assumption of the NB, although one might expect the other methods to indentify the clusters. Jensen Shannon divergence also performs rather well, separating all (but one) clusters in two dimensions. Look in first and third dimension. Our method gets outperformed by PCoA according to this statistic. Would this be because PCoA focusses explicitly on samplewise distances?
```

```{r resList1Evalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette1}"}
plotSil(resList1Rows, groupFactor1)
```

```{r resList1Evaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions. Red diamonds indicate means. \\label{fig:dist1}"}
distRatioPlot(resList1Rows, groupFactor1)
```

```{r libSizeCorr1, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions. Red diamonds indicate means. \\label{fig:libCor1}"}

plotCor(resList1Rows, datList1, dataMat = TRUE)
rm(resList1Rows, resList1, datList1)
```

The library sizes of the CoDa approach do negatively correlate with the row scores, confirming our concern about the heteroscedasticity.

\clearpage

##### Simulate from related distributions

In this setting we will simulate data from 4 related negative binomial distributions, all obtained from the same mother distributions but 10\% of the taxa have modified abundances

```{r NBrelatedDistr, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions
TPR = 0.1
FC = 5

groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactor2 = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor
if(!file.exists(file = "RCMsim/resList2.RData")){
  if(!file.exists(file = "RCMsim/datList2.RData")){
datList2 = mclapply(mc.cores = 1, 1:reps, function(i){
## Sample parameters
rhosSampled0 = sample(rhoEst, p)
rhosSampled0 = rhosSampled0/sum(rhosSampled0)
#For each population, modify a fraction TPR of the abundances (compensation is not so badly needed here, we do not test individual species). Record the taxa that have been modified.
rhosIDSampled = lapply(1:nPop, function(x, rhos){
  id = sample(seq_len(p), round(TPR*p))
  rhos[id] = rhos[id]*FC
  cbind(rhos = rhos/sum(rhos), id = id) #renormalize
}, rhos = rhosSampled0)

rhosSampled00 = sapply(rhosIDSampled, function(x){x[, "rhos"]})
rhosSampled = rhosSampled00[,groupInd]
IDsampled = sapply(rhosIDSampled, function(x){x[, "id"]})

thetasSampled = sample(thetaEst, p) 
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes

## Create dataset
meanMat = libSizesSampled * t(rhosSampled)
dataMat = matrix(rnbinom(prod(dim(meanMat)), mu = meanMat, size = matrix(thetasSampled,n,p,byrow = TRUE)),n,p)
colnames(dataMat) = rownames(rhosSampled) 
rownames(dataMat) = groupFactor2
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, thetasSampled = thetasSampled, rhosSampled = rhosSampled00, IDsampled = IDsampled)
})
save(datList2,groupFactor2,  file = "RCMsim/datList2.RData")
} else {load(file = "RCMsim/datList2.RData")}

# Results

  if(!file.exists("RCMsim/resList2RCM.RData")){
RCMresList2 = mclapply(mc.cores=4, datList2, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresList2, file = "RCMsim/resList2RCM.RData")
} else {load("RCMsim/resList2RCM.RData")}
resList2 = otherAnalyses(RCMresList2)
IDlist2 = lapply(datList2, function(x){
   id = apply(x$IDsampled, 2, unique)
   taxa = apply(id, 2, function(y){rownames(x$rhosSampled)[y]})
   })
resList2Rows = lapply(resList2, extractRows)
resList2Cols = mapply(resList2, resList2Rows,FUN = extractCols, SIMPLIFY = FALSE)
datList2 = lapply(resList2, function(x){x$RCM$X})
save(datList2, resList2Rows, resList2Cols, IDlist2, groupFactor2, file = "RCMsim/resList2.RData")
} else {load(file = "RCMsim/resList2.RData")}
```

```{r resList2 sampleBiplot, purl = FALSE}
#A sample biplot
randomScatter(resList2Rows, col = groupFactor2)
# In this setting, the RCM really appears to outperform the other methods
# par(mfrow = c(2,4), mar = c(3,3,3,3))
# ID = sample(size=1, seq_along(resList2))
# obj = resList2Rows[[ID]]
# lapply(names(obj),function(x){plot(obj[[x]][,c(1,3)], col = groupFactor2, main=x, xlab = "Dim 1",ylab = "Dim 3", pty = "s")})
# rm(ID, obj)
# par(mfrow = c(1,1))
```

```{r resList2Evalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette2}"}
plotSil(resList2Rows, groupFactor2)
```

The RC(M) method has the highest silhouettes, PCoA is second best but far below.

```{r resList2Evaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:dist2}"}
distRatioPlot(resList2Rows, groupFactor2)
```

RCM comes out as best, CA as worst

```{r libSizeCorr2, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:libCor2}"}
plotCor(resList2Rows, datList2, dataMat = TRUE)
```

```{r AbCorr2, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCor2}"}
plotCor(resList2Cols, datList2, dataMat = TRUE, scoreDim = "columns")
```

The CoDa method is clearly affected by the library sizes, as is CA and Bray-Curtis on absolute abundances.

```{r taxRat2, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRat2}"}
contr2 = contrTaxaWrap(resList2Rows, resList2Cols, IDlist2, groupFactor2)
contrPlot(contr2)

# contr2Mean = contrTaxaWrap(resList2Rows, resList2Cols, IDlist2, groupFactor2, centerFun = mean)
# contrPlot(contr2Mean)
```

The RC(M) and CA methods are best at identifying taxa that contribute to the separation of the clusters. This is not surprising since they are the only ones that have genuine taxon scores, for CoDA and PCoA we have to rely on weighted sample scores for the taxa. Even though this is often used, our results reveal their poor performance. Hellinger does well because of outliers, it does not really paint a good picture.

##### Simulate from related distributions, confounded by library sizes

In this setting we will again simulate data from 4 related negative binomial distributions, all obtained from the same mother distributions but 10\% of the taxa have modified abundances. In this case we will make the library sizes different in each group. Th initial library sizes are all sampled from the same pool of library sizes observed in real data. The first group has unmodified library sizes, the second group library sizes multiplied by 1.5, the third by 2 and the fourth by 3.

```{r NBrelatedDistrConf, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions
TPR = 0.1
FC = 5

groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactor3 = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor
if(!file.exists(file = "RCMsim/resList3.RData")){
  if(!file.exists(file = "RCMsim/datList3.RData")){
datList3 = mclapply(mc.cores = 4, 1:reps, function(i){
## Sample parameters
rhosSampled0 = sample(rhoEst, p)
rhosSampled0 = rhosSampled0/sum(rhosSampled0)
#For each population, modify a fraction TPR of the abundances (compensation is not so badly needed here, we do not test individual species). Record the taxa that have been modified.
rhosIDSampled = lapply(1:nPop, function(x, rhos){
  id = sample(seq_len(p), round(TPR*p))
  rhos[id] = rhos[id]*FC
  cbind(rhos = rhos/sum(rhos), id = id) #renormalize
}, rhos = rhosSampled0)

rhosSampled00 = sapply(rhosIDSampled, function(x){x[, "rhos"]})
rhosSampled = rhosSampled00[,groupInd]
IDsampled = sapply(rhosIDSampled, function(x){x[, "id"]})

thetasSampled = sample(thetaEst, p) 
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes, confounded with group membership
libSizesSampled[groupInd==2] = round(libSizesSampled[groupInd==2]*1.5)
libSizesSampled[groupInd==3] = round(libSizesSampled[groupInd==3]*2)
libSizesSampled[groupInd==4] = round(libSizesSampled[groupInd==4]*3)

## Create dataset
meanMat = libSizesSampled * t(rhosSampled)
dataMat = matrix(rnbinom(prod(dim(meanMat)), mu = meanMat, size = matrix(thetasSampled,n,p,byrow = TRUE)),n,p)
colnames(dataMat) = rownames(rhosSampled) 
rownames(dataMat) = groupFactor3
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, thetasSampled = thetasSampled, rhosSampled = rhosSampled00, IDsampled = IDsampled)
})
save(datList3,groupFactor3,  file = "RCMsim/datList3.RData")
} else {load(file = "RCMsim/datList3.RData")}

# Results
  if(!file.exists("RCMsim/resList3RCM.RData")){
RCMresList3 = mclapply(mc.cores=4, datList3, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresList3, file = "RCMsim/resList3RCM.RData")
} else {load("RCMsim/resList3RCM.RData")}
resList3 = otherAnalyses(RCMresList3)
IDlist3 = lapply(datList3, function(x){
   id = apply(x$IDsampled, 2, unique)
   taxa = apply(id, 2, function(y){rownames(x$rhosSampled)[y]})
   })
resList3Rows = lapply(resList3, extractRows)
resList3Cols = mapply(resList3, resList3Rows,FUN = extractCols, SIMPLIFY = FALSE)
save(resList3,resList3Rows, resList3Cols, IDlist3, groupFactor3, file = "RCMsim/resList3.RData")
} else {load(file = "RCMsim/resList3.RData")}
```

```{r resList3 sampleBiplot, purl = FALSE}
#A sample biplot
resList3Rows = lapply(resList3, extractRows)
randomScatter(resList3Rows, col = groupFactor3)
```

```{r resList3Evalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette3}"}
plotSil(resList3Rows, groupFactor3)
```

The RC(M) method has the highest silhouettes, PCoA is second best but far below.

```{r resList3Evaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:dist3}"}
distRatioPlot(resList3Rows, groupFactor3)
```

RCM comes out as best, CA as worst

```{r libSizeCorr3, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:libCor3}"}
datList3 = lapply(resList3, function(x){x$RCM$X})
plotCor(resList3Rows, datList3, dataMat = TRUE)
```

```{r AbCorr3, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCor3}"}
plotCor(resList3Cols, datList3, dataMat = TRUE, scoreDim = "columns")
```

The CoDa method is clearly affected by the library sizes, as is CA and Bray-Curtis on absolute abundances.

```{r taxRat3, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRat3}"}
contr3 = contrTaxaWrap(resList3Rows, resList3Cols, IDlist3, groupFactor3)
contrPlot(contr3)
```

##### Simulate under H0, but with differences in dispersion

Inspired by Warton _et al._ (2012): "Distance-based multivariate analyses confound location and dispersion effects" we set up a simulation without any signal, but whereby the dispersions in the different groups are different, as a result of the group effect itself or related sampling effects. Note that this also violates the assumption of our RC(M) method of constant dispersions

```{r NBdispsersion, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions

groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactor4 = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor
if(!file.exists(file = "RCMsim/resList4.RData")){
  if(!file.exists(file = "RCMsim/datList4.RData")){
datList4 = mclapply(mc.cores = 4, 1:reps, function(i){
## Sample parameters
rhosSampled = sample(rhoEst, p)
thetasSampled = sample(thetaEst, p) 
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes
#Sampled dispersions differ between the population
thetaMat = matrix(thetasSampled, n, p, byrow = TRUE) * c(rep(c(0.2,0.5,1,5), eac = nPerPop)) #Modify dispersions to 5 times, 2 times and 0.2 times the normal levels

## Create dataset
meanMat = outer(libSizesSampled, rhosSampled)
dataMat = matrix(rnbinom(prod(dim(meanMat)), mu = meanMat, size = thetaMat),n,p)
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, thetaMat = thetaMat, rhosSampled = rhosSampled)
})
save(datList4,groupFactor4,  file = "RCMsim/datList4.RData")
} else {load(file = "RCMsim/datList4.RData")}

# Results
  if(!file.exists("RCMsim/resList4RCM.RData")){
RCMresList4 = mclapply(mc.cores=4, datList4, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresList4, file = "RCMsim/resList4RCM.RData")
} else {load("RCMsim/resList4RCM.RData")}
resList4 = otherAnalyses(RCMresList4)
resList4Rows = lapply(resList4, extractRows)
resList4Cols = mapply(resList4, resList4Rows,FUN = extractCols, SIMPLIFY = FALSE)
save(resList4,resList4Rows, resList4Cols, groupFactor4, file = "RCMsim/resList4.RData")
} else {load(file = "RCMsim/resList4.RData")}
```

```{r resList4 sampleBiplot, purl = FALSE}
#A sample biplot
randomScatter(resList4Rows, col = groupFactor4)
```

RCM comes out as best, CA as worst

```{r libSizeCorr4, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:libCor4}"}
datList4 = lapply(resList4, function(x){x$RCM$X})
plotCor(resList4Rows, datList4, dataMat = TRUE)
```

```{r AbCorr4, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCor4}"}
plotCor(resList4Cols, datList4, dataMat = TRUE, scoreDim = "columns")
```

##### Simulate under H1, with correlated taxa

We know that taxa living in the same niche are dependent on each other, and as a result their abundances are correlated. This phenomenon can affect the differentialn abundance detection, so it could also affect our visualization procedures. I would also expect it to have an effect on the CoDa method, which relies on ratios between taxa of the same sample. We'll use the correlation networks previously estimated from the HMP and AGP datasets.

```{r NBrelatedDistrCorr, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions
TPR = 0.1
FC = 5
load("/home/stijn/PhD/Simulations/data/covList.RData")
groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactor5 = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor
if(!file.exists(file = "RCMsim/resList5.RData")){
  if(!file.exists(file = "RCMsim/datList5.RData")){
datList5 = mclapply(mc.cores = 4, 1:reps, function(i){
## Sample parameters
rhosSampled0 = sample(rhoEst, p)
rhosSampled0 = rhosSampled0/sum(rhosSampled0)
#For each population, modify a fraction TPR of the abundances (compensation is not so badly needed here, we do not test individual species). Record the taxa that have been modified.
rhosIDSampled = lapply(1:nPop, function(x, rhos){
  id = sample(seq_len(p), round(TPR*p))
  rhos[id] = rhos[id]*FC
  cbind(rhos = rhos/sum(rhos), id = id) #renormalize
}, rhos = rhosSampled0)

rhosSampled00 = sapply(rhosIDSampled, function(x){x[, "rhos"]})
rhosSampled = rhosSampled00[,groupInd]
IDsampled = sapply(rhosIDSampled, function(x){x[, "id"]})

thetasSampled = sample(thetaEst, p) 
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes, confounded with group membership
covMatName = sample(names(covList), 1)

## Create dataset
meanMat = libSizesSampled * t(rhosSampled)

dataMat = rmvnegbin(n = n,mu = meanMat, ks = thetasSampled, Sigma = as.matrix(covList[[covMatName]]))
colnames(dataMat) = rownames(rhosSampled) 
rownames(dataMat) = groupFactor5
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, thetasSampled = thetasSampled, rhosSampled = rhosSampled00, IDsampled = IDsampled, covMatName = covMatName)
})
save(datList5,groupFactor5,  file = "RCMsim/datList5.RData")
} else {load(file = "RCMsim/datList5.RData")}

# Results
  if(!file.exists("RCMsim/resList5RCM.RData")){
RCMresList5 = mclapply(mc.cores=4, datList5, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresList5, file = "RCMsim/resList5RCM.RData")
} else {load("RCMsim/resList5RCM.RData")}
resList5 = otherAnalyses(RCMresList5)
IDlist5 = lapply(datList5, function(x){
   id = apply(x$IDsampled, 2, unique)
   taxa = apply(id, 2, function(y){rownames(x$rhosSampled)[y]})
   })
resList5Rows = lapply(resList5, extractRows)
resList5Cols = mapply(resList5, resList5Rows,FUN = extractCols, SIMPLIFY = FALSE)
save(resList5,resList5Rows, resList5Cols, IDlist5, groupFactor5, file = "RCMsim/resList5.RData")
} else {load(file = "RCMsim/resList5.RData")}
```

```{r resList5 sampleBiplot, purl = FALSE}
#A sample biplot
randomScatter(resList5Rows, col = groupFactor5)
```

```{r resList5Evalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette5}"}
plotSil(resList5Rows, groupFactor5)
```

```{r resList5Evaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:dist5}"}
distRatioPlot(resList5Rows, groupFactor5)
```

RCM comes out as best, CA as worst

```{r libSizeCorr5, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:libCor5}"}
datList5 = lapply(resList5, function(x){x$RCM$X})
plotCor(resList5Rows, datList5, dataMat = TRUE)
```

```{r AbCorr5, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCor5}"}
plotCor(resList5Cols, datList5, dataMat = TRUE, scoreDim = "columns")
```

```{r taxRat5, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRat5}"}
contr5 = contrTaxaWrap(resList5Rows, resList5Cols, IDlist5, groupFactor5)
contrPlot(contr5)
```

##### Negative binomial with phylogenetic information

In order to also include the very popular distance measures UniFrac and weighted UniFrac, we also generate data with the negative binomial and Dirichlet multinomial taking into account phylogenetic structure. We base our strategy on the second strategy described by Chen _et al._ ("Associating microbiome composition with environmental
covariates using generalized UniFrac distances", 2012). We generate random trees for the template vector of relative abundances, and modify abundances within one of the 20 clusters

```{r UniFrac, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions
TPR = 0.1
FC = 5
k = round(1/TPR) #Number of clusters

groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactorPhy = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor

clusters = seq_len(k)

if(!file.exists(file = "RCMsim/resListPhy.RData")){
  if(!file.exists(file = "RCMsim/datListPhy.RData")){
datListPhy = mclapply(mc.cores = 4, 1:reps, function(i){
## Sample parameters
cluSampled = sample(clusters, nPop) #Cluster to be made DA

rhosSampled0 = sample(rhoEst, p)
rhosSampled0 = rhosSampled0/sum(rhosSampled0)

randomTree = rtree(n = p, tip.label = names(rhosSampled0)) # The random tree
dd = as.dist(cophenetic.phylo(randomTree)) #Get branch lenghts
psclust = cutree(as.hclust(hclust(dd)), k = k) #Split into clusters
  
#For each population, modify the abundances of a cluster of related taxa. Record the clusters that have been modified.
rhosIDSampled = lapply(cluSampled, function(x, rhos){
  id = x == psclust
  rhos[id] = rhos[id]*FC
  cbind(rhos = rhos/sum(rhos), id = id) #renormalize
}, rhos = rhosSampled0)

rhosSampled00 = sapply(rhosIDSampled, function(x){x[, "rhos"]})
rhosSampled = rhosSampled00[,groupInd]
IDsampled = sapply(rhosIDSampled, function(x){x[, "id"]})

thetasSampled = sample(thetaEst, p) 
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes

## Create dataset
meanMat = libSizesSampled * t(rhosSampled)
dataMat = matrix(rnbinom(prod(dim(meanMat)), mu = meanMat, size = matrix(thetasSampled,n,p,byrow = TRUE)),n,p)
colnames(dataMat) = rownames(rhosSampled) 
rownames(dataMat) = groupFactorPhy
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, thetasSampled = thetasSampled, rhosSampled = rhosSampled00, IDsampled = IDsampled, cluSampled = cluSampled, randomTree = randomTree)
})
save(datListPhy, groupFactorPhy, file = "RCMsim/datListPhy.RData")
} else {load(file = "RCMsim/datListPhy.RData")}

# Results
 if(!file.exists("RCMsim/resListPhyRCM.RData")){
RCMresListPhy = mclapply(mc.cores=4, datListPhy, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresListPhy, file = "RCMsim/resListPhyRCM.RData")
  } else {load("RCMsim/resListPhyRCM.RData")}
  listID = sapply(RCMresListPhy, is.list)
resListPhy = otherAnalyses(RCMresListPhy[listID], unifrac = TRUE, treeList = lapply(datListPhy[listID], function(x){x$randomTree}) )
IDlistPhy = lapply(datListPhy[listID], function(x){
   taxa = apply(x$IDsampled, 2, function(y){rownames(x$rhosSampled)[as.logical(y)]})
   })
resListPhyRows = lapply(resListPhy, extractRows)
resListPhyCols = mapply(resListPhy, resListPhyRows,FUN = extractCols, SIMPLIFY = FALSE)
save(resListPhy, resListPhyRows, resListPhyCols, groupFactorPhy, IDlistPhy, file = "RCMsim/resListPhy.RData")
} else {load(file = "RCMsim/resListPhy.RData")}
```

```{r resListPhy sampleBiplot, purl = FALSE}
#A sample biplot
randomScatter(resListPhyRows, col = groupFactorPhy)
```

```{r resListPhyEvalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouettePhy}"}
plotSil(resListPhyRows, groupFactorPhy, las = 2)
```

The Unifrac distances do not seem to get any advantage from their phylogenetic distances

```{r resListPhyEvaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:distPhy}"}
distRatioPlot(resListPhyRows, groupFactorPhy)
```

```{r libSizeCorrPhy, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:libCorPhy}"}
datListPhy = lapply(resListPhy, function(x){x$RCM$X})
plotCor(resListPhyRows, datListPhy, dataMat = TRUE)
```

Regular unifrac suffers from correlations with the library sizes

```{r AbCorrPhy, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCorPhy}"}
plotCor(resListPhyCols, datListPhy, dataMat = TRUE, scoreDim = "columns")
```

```{r taxRatPhy, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRatPhy}"}
contrPhy = contrTaxaWrap(resListPhyRows, resListPhyCols, IDlistPhy, groupFactorPhy)
contrPlot(contrPhy, ylim = c(-4,4)) #Excluding outliers
```

#### Other distributions

We also simulate datasets from other distributions than the negative binomial to see how the method behaves when its distributional assumptions are violated.

##### Zero-inflated negative binomial distribution

The zero-inflated negative binomial (ZINB) distribution was used by Zhang _et al._ (2016) for differential abundance testing. Our own goodness-of-fit tests indicate that for 14\% of the taxa, the ZINB provides a better fit than the regular NB

```{r simZINB, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions
TPR = 0.1
FC = 5

groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactorZINB = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor

if(!file.exists(file = "RCMsim/resListZINB.RData")){
load(file="/home/stijn/PhD/Biplots/Zicoef.RData")

if(!file.exists(file = "RCMsim/datListZINB.RData")){
datListZINB = mclapply(mc.cores=1, 1:reps, function(i){
## Sample parameters
rhosSampled0 = sample(rhoEst, p)
rhosSampled0 = rhosSampled0/sum(rhosSampled0)
#For each population, modify a fraction TPR of the abundances (compensation is not so badly needed here, we do not test individual species). Record the taxa that have been modified.
rhosIDSampled = lapply(1:nPop, function(x, rhos){
  id = sample(seq_len(p), round(TPR*p))
  rhos[id] = rhos[id]*FC
  cbind(rhos = rhos/sum(rhos), id = id) #renormalize
}, rhos = rhosSampled0)

rhosSampled00 = sapply(rhosIDSampled, function(x){x[, "rhos"]})
rhosSampled = rhosSampled00[,groupInd]
IDsampled = sapply(rhosIDSampled, function(x){x[, "id"]})

thetasSampled = sample(thetaEst, p) 
propSampled = expit(sample(ZIcoef, p))
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes

## Create dataset
meanMat = libSizesSampled * t(rhosSampled)
dataMat = matrix(rzinegbin(prod(dim(meanMat)), munb = meanMat, size = matrix(thetasSampled,n,p, byrow = TRUE), pstr0 = matrix(propSampled,n,p,byrow = TRUE)),n,p)
colnames(dataMat) = rownames(rhosSampled) 
rownames(dataMat) = groupFactorZINB
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, thetasSampled = thetasSampled, rhosSampled = rhosSampled00, IDsampled = IDsampled, propSampled = propSampled)
})
save(datListZINB,groupFactorZINB,  file = "RCMsim/datListZINB.RData")
} else {load(file = "RCMsim/datListZINB.RData")}

# Results

 if(!file.exists("RCMsim/resListZINBRCM.RData")){
RCMresListZINB = mclapply(mc.cores=4, datListZINB, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresListZINB, file = "RCMsim/resListZINBRCM.RData")
} else {load("RCMsim/resListZINBRCM.RData")}
resListZINB = otherAnalyses(RCMresListZINB)
IDlistZINB = lapply(datListZINB, function(x){
  id = apply(x$IDsampled, 2, unique)
  taxa = apply(id, 2, function(y){rownames(x$rhosSampled)[y]})
  })
save(resListZINB,groupFactorZINB, IDlistZINB, file = "RCMsim/resListZINB.RData")
} else {load(file = "RCMsim/resListZINB.RData")}
```

```{r resListZINB sampleBiplot, purl = FALSE}
#A sample biplot
resListZINBRows = lapply(resListZINB, extractRows)
randomScatter(resListZINBRows, col = groupFactorZINB)
# In this setting, the RCM really appears to outperform the other methods

# par(mfrow = c(ZINB,4), mar = c(3,3,3,3))
# ID = sample(size=1, seq_along(resListZINB))
# obj = resListZINBRows[[ID]]
# lapply(names(obj),function(x){plot(obj[[x]][,c(1,3)], col = groupFactorZINB, main=x, xlab = "Dim 1",ylab = "Dim 3", pty = "s")})
# rm(ID, obj)
# par(mfrow = c(1,1))
```

```{r resListZINBEvalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette1}"}
plotSil(resListZINBRows, groupFactorZINB)
```

The RC(M) method has the highest silhouettes, PCoA is second best but far below.

```{r resListZINBEvaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:dist1}"}
distRatioPlot(resListZINBRows, groupFactorZINB)
```

RCM comes out as best, CA as worst

```{r libSizeCorrZINB, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:libCor1}"}
datListZINB = lapply(resListZINB, function(x){x$RCM$X})
plotCor(resListZINBRows, datListZINB)
```

```{r AbCorrZINB, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCorZINB}"}
resListZINBCols = mapply(resListZINB, resListZINBRows,FUN = extractCols, SIMPLIFY = FALSE)
plotCor(resListZINBCols, datListZINB, dataMat = TRUE, scoreDim = "columns")
```

CA suffers from correlation with relative abundances, but not on average.

```{r taxRatZINB, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRatZINB}"}
contrZINB = contrTaxaWrap(resListZINBRows, resListZINBCols, IDlistZINB, groupFactorZINB)
contrPlot(contrZINB)
```

Also here RCM and CA perform best in identifying taxa thanks to their explicit taxon scores.

\clearpage

##### Dirichlet-multinomial

The use of the Dirichlet multinomial for microbiome data was proposed by La Rosa _et al._ (2012). Although it is very restrictive in terms of variance and covariance structures, we include it here because of its relative popularity. Also Chen _et al._ (2012) use it to simulate data for a visualization.

```{r simDM, purl = FALSE}
p = 1000 #Number of taxa
n = 60 #Number of samples
nPop = 4 #Number of underlying populations
nPerPop = n/nPop #Number of samples per population
reps = 100 #replicates
nDim  = 3 #Number of dimensions
TPR = 0.1
FC = 5

groupInd = rep(1:nPop, each = nPerPop) #Group indicator, as integers
groupFactorDM = factor(groupInd, levels = 1:nPop, labels =  paste("Population", 1:nPop), ordered = TRUE) #as factor
load("/home/stijn/PhD/American Gut/AGdm.RData")

if(!file.exists(file = "RCMsim/datListDM.RData")){
datListDM = mclapply(mc.cores = 1, 1:reps, function(i){
## Sample parameters
rhosSampled0 = sample(rhoEst, p)
rhosSampled0 = rhosSampled0/sum(rhosSampled0)
#For each population, modify a fraction TPR of the abundances (compensation is not so badly needed here, we do not test individual species). Record the taxa that have been modified.
rhosIDSampled = lapply(1:nPop, function(x, rhos){
  id = sample(seq_len(p), round(TPR*p))
  rhos[id] = rhos[id]*FC
  cbind(rhos = rhos/sum(rhos), id = id) #renormalize
}, rhos = rhosSampled0)

rhosSampled00 = sapply(rhosIDSampled, function(x){x[, "rhos"]})
rhosSampled = rhosSampled00[,groupInd]
IDsampled = sapply(rhosIDSampled, function(x){x[, "id"]})

thetaDM = AGdm$theta
libSizesSampled = sample(libSizesEst, n) #Sampled libSizes

## Create dataset
dataMat = t(sapply(seq_along(libSizesSampled),function(i){Dirichlet.multinomial(libSizesSampled[i], shape = rhosSampled[,i]*(1-thetaDM)/thetaDM)}))
colnames(dataMat) = rownames(rhosSampled) 
rownames(dataMat) = groupFactorDM
dataMat = dataMat[rowSums(dataMat)>0,colSums(dataMat)>0]
list(dataMat = dataMat, libSizesSampled = libSizesSampled, rhosSampled = rhosSampled00, IDsampled = IDsampled)
})
save(datListDM,groupFactorDM, file = "RCMsim/datListDM.RData")
} else {load(file = "RCMsim/datListDM.RData")}

# Results
if(!file.exists(file = "RCMsim/resListDM.RData")){
  if(!file.exists("RCMsim/resListDMRCM.RData")){
RCMresListDM = mclapply(mc.cores=4, datListDM, function(y){
  RCM(y$dataMat, k = nDim, distribution = "NB")
  })
save(RCMresListDM, file = "RCMsim/resListDMRCM.RData")
  } else {load("RCMsim/resListDMRCM.RData")}
  listID = sapply(RCMresListDM, class)=="RCM"
RCMresListDM = Filter(RCMresListDM, f = is.list)
resListDM = otherAnalyses(RCMresListDM)
IDlistDM = lapply( datListDM[listID], function(x){
  id = apply(x$IDsampled, 2, unique)
  taxa = apply(id, 2, function(y){rownames(x$rhosSampled)[y]})
  })
save(resListDM, groupFactorDM,IDlistDM, file = "RCMsim/resListDM.RData")
} else {load(file = "RCMsim/resListDM.RData")}
```

Only 68% fits

```{r resListDM sampleBiplot, purl = FALSE}
#A sample biplot
resListDMRows = lapply(resListDM, extractRows)
randomScatter(resListDMRows, col = groupFactorDM)
# Our method and CoDa show only white noise, CA and PCoA show artificial clusters. Confounding mean and dispersion effects?
#randomScatter(resListDMRows, Dim = c(1,3))
```

```{r libSizeCorrDM, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 1DMDM simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.\\label{fig:libCor1}"}
datListDM = lapply(resListDM, function(x){x$RCM$X})
plotCor(resListDMRows, datListDM, scoreDim = "rows")
```

Hellinger distance, CoDa ana Bray-Curtis on absolut abundances show correlations with library sizes.

The taxa

```{r resListDMcols sampleBiplot, purl = FALSE, include = FALSE}
#A sample biplot
resListDMCols = mapply(resListDM, resListDMRows, FUN=extractCols, SIMPLIFY=FALSE)
randomScatter(resListDMCols)
# CA has outliers, CoDa and Hellinger reveal weird patterns
#randomScatter(resListDMCols, Dim = c(1,3))
```

CoDa and Hellinger distances have unacceptable correlatios with the library sizes

```{r AbsCorrDM, purl = FALSE, fig.cap = "Correlations of column scores with average relative abundances, results of 1DMDM simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.\\label{fig:libCor1}"}
plotCor(resListDMCols, lapply(resListDM, function(x){x$RCM$X}), scoreDim = "columns", dataMat = TRUE)
```

```{r resListDMEvalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette1}"}
plotSil(resListDMRows, groupFactorDM)
```

JSD and Bray-Curtis have the best performance in this setting in terms of silhouette

```{r resListDMEvaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.  \\label{fig:dist1}"}
distRatioPlot(resListDMRows, groupFactorDM)
```

RCM comes out as best, together with JSD and Bray-Curtis

```{r AbCorrDM, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCorDM}"}
resListDMCols = mapply(resListDM, resListDMRows,FUN = extractCols, SIMPLIFY = FALSE)
plotCor(resListDMCols, datListDM, dataMat = TRUE, scoreDim = "columns")
```

CA suffers from correlation with relative abundances, but not on average.

```{r taxRatDM, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRatDM}"}
contrDM = contrTaxaWrap(resListDMRows, resListDMCols, IDlistDM, groupFactorDM)
contrPlot(contrDM)
```

In this setting none of the methods does really well at identifying causal taxa

\clearpage

##### Nonparametric simulation

An objective simulation approach would be to use non-parametric resampling from a true dataset, as in _SimSeq_. For this we need microbiome datasets with covariates known to be related to bacterial abundance, preferably with more than two groups. The Zeller data is one such dataset, with the cancer variable expected to be related to relative abundance and having three levels (Normal, small adenoma and cancer). For the American Gut data we can use IBD. For the Turnbaugh lean and obese twins dataset we can use health status (lean, overweight or obese), even though we have only 18 samples. For the Turnbaugh humanized mouse dataset we could use "Diet", with levels "BK" and "Western".

We generate data as follows

1. Select a covariate and test for differential abundance using WMW/Kruskal-Wallis
2. Calculate local false discovery rates (lfdr)
3. Sample non DA taxa with equal weights from all taxa
4. Sample DA taxa from all taxa with weights equal to 1-lfdr
5. Sample counts from non DA taxa from samples with the most frequent covariate level
6. Sample counts from DA taxa from the samples with other covariate levels, and correct for differences in library sizes. This maintains the same distribution of covariate levels and overal daat matrix size as the original dataset. 

```{r simSeqFunctions}
# A wrapper function for phyloseq objects for the testing, yields the lfdr's of the taxa. We extend the original code to allow for Kruskal-Wallis test
testSimSeq = function(physeq, variable, filterTaxa = FALSE){
  countMat = if(taxa_are_rows(physeq)){t(otu_table(physeq)@.Data)} else {otu_table(physeq)@.Data}
  treatment = get_variable(physeq, variable)
  if(filterTaxa){
  # Filter physeq to no all-zeroes in either of the groups
  physeq = prune_taxa(x = physeq, taxa = apply(countMat, 2, function(y){
    !any(tapply(y, treatment, function(z){
      all(z==0)
    }))
  }))
  }
  #Filter physeq for NA's in variables
  physeq = prune_samples(x = physeq, !is.na(treatment))
  treatment = factor(get_variable(physeq, variable))
  countMat = if(taxa_are_rows(physeq)){t(otu_table(physeq)@.Data)} else {otu_table(physeq)@.Data}
  pvals = apply(countMat, 2, function(x){kruskal.test(x, g=treatment)$p.value})
  result = apply(countMat, 2, function(x){
    avRank = length(x)/2 + 0.5
    rankX = rank(x)
    sapply(levels(treatment),avRank = avRank, function(y, avRank){
    ifelse(mean(rankX[treatment==y])>avRank, "up","down")
        })}) #Was taxon up or down regulated in group 1 (first level of treatment?
  lfdr = fdrtool(pvals, statistic = "pvalue", plot=FALSE, verbose=FALSE)$lfdr
  list(lfdr = lfdr, physeq = physeq, variable = variable, result = result)
}
genSimSeq = function(testres, Ntaxa = ntaxa(testres$physeq), fracTaxa = 0.2){
  physeq = testres$physeq
  weights = 1-testres$lfdr
  weights = weights/sum(weights) #Normalize sampling weights
  normFactors = sample_sums(physeq)
  treatment = get_variable(physeq, testres$variable)
  DEtaxa = sample(taxa_names(physeq), round(Ntaxa*fracTaxa), prob = weights) # The DE taxa
  whichTrt = names(sort(table(treatment),decreasing=TRUE)[1]) # The most frequent treatment group, frow which we will sample the EE taxa
  trtLeft = unique(treatment)[unique(treatment) != whichTrt]
  countMat = if(taxa_are_rows(physeq)){t(otu_table(physeq)@.Data)} else {otu_table(physeq)@.Data}
  samIDee = sample(which(treatment == whichTrt),nrow(countMat), replace = TRUE)
  dataSimSeq = countMat[samIDee,] #EE taxa
  for (i in trtLeft){
  rowInd = which(treatment==i)
  dataSimSeq[rowInd,DEtaxa] = round(countMat[rowInd, DEtaxa] * normFactors[samIDee][rowInd]/normFactors[rowInd])
  }
  keepID = colSums(dataSimSeq)>0
  list(data = dataSimSeq[, keepID], treatment = treatment, DEtaxa = DEtaxa[DEtaxa %in% colnames(dataSimSeq[, keepID])])
}
```

```{r SimSeqTests, purl = FALSE}
if(!file.exists(file = "SimSeqTests.RData")){
load(file = "/home/stijn/PhD/Datasets/Humanized mouse/HumMicePhy.RData")
testHum = testSimSeq(HumMicePhy, "Diet")
load(file = "/home/stijn/PhD/Datasets/TurnbaughObeseTwins/TurnTwins.RData")
testTwins = testSimSeq(TurnTwins, "Status_p") # No significance
testAG = testSimSeq(AGphylo, "IBD")
testZeller = testSimSeq(zellerSphy, "Diagnosis")
save(testHum, testTwins, testAG, testZeller, file = "SimSeqTests.RData")
} else {load(file = "SimSeqTests.RData")}
```

No DA for the lean and obese twins. Then again it is a small dataset.

```{r SimSeqGenerateData, purl = FALSE}
if(!file.exists(file = "SimSeqData.RData")){
NsimSimSeq = 100
HumSimSeqList = mclapply(mc.cores = 1, 1:NsimSimSeq, function(i){
  genSimSeq(testHum)
})
AGSimSeqList = mclapply(mc.cores = 1, 1:NsimSimSeq, function(i){
  genSimSeq(testAG)
})
ZellerSimSeqList = mclapply(mc.cores = 1, 1:NsimSimSeq, function(i){
  genSimSeq(testZeller)
})
save(HumSimSeqList, ZellerSimSeqList, AGSimSeqList, file = "SimSeqData")
} else {load(file = "SimSeqData.RData")}
```

```{r simSeqAnalysis, purl = FALSE}
if(!file.exists("RCMsim/resListSimSeq.RData")){
  if(!file.exists("RCMsim/resListSimSeqRCM.RData")){
RCMresListSimSeqHum = mclapply(mc.cores=4, HumSimSeqList, function(y){
  RCM(y$data, k = nDim, distribution = "NB")
  })
RCMresListSimSeqAG = mclapply(mc.cores=4, AGSimSeqList, function(y){
  RCM(y$data, k = nDim, distribution = "NB")
  })
RCMresListSimSeqZeller = mclapply(mc.cores=4, ZellerSimSeqList, function(y){
  RCM(y$data, k = nDim, distribution = "NB", round = TRUE)
  })
save(RCMresListSimSeqHum, RCMresListSimSeqAG, RCMresListSimSeqZeller, file = "RCMsim/resListSimSeqRCM.RData")
  } else {load("RCMsim/resListSimSeqRCM.RData")}
  resListSimSeqHum = otherAnalyses(RCMresListSimSeqHum)
  IDlistHum = mapply(HumSimSeqList,resListSimSeqHum, FUN =  function(y,x, test){
    taxa = y$DEtaxa[y$DEtaxa %in% colnames(x$RCM$X)]
    test[taxa,]
    }, SIMPLIFY = FALSE, MoreArgs = list(test = t(testHum$result)))
  groupFactorHum = HumSimSeqList[[1]]$treatment
  resListSimSeqHumRows = lapply(resListSimSeqHum, extractRows)
  resListSimSeqHumCols = mapply(resListSimSeqHum, resListSimSeqHumRows, FUN=extractCols, SIMPLIFY=FALSE)
  datListHum = lapply(resListSimSeqHum, function(x){x$RCM$X})
  # Zeller
  zellerID = sapply(RCMresListSimSeqZeller, class) == "RCM"
  resListSimSeqZeller = otherAnalyses(RCMresListSimSeqZeller[zellerID])
  IDlistZeller = mapply(ZellerSimSeqList[zellerID],resListSimSeqZeller, FUN =  function(y,x, test){
    taxa = y$DEtaxa[y$DEtaxa %in% colnames(x$RCM$X)]
    test[taxa,]
    }, SIMPLIFY = FALSE, MoreArgs = list(test = t(testZeller$result)))
  groupFactorZeller = ZellerSimSeqList[[1]]$treatment
  resListSimSeqZellerRows = lapply(resListSimSeqZeller, extractRows)
  resListSimSeqZellerCols = mapply(resListSimSeqZeller, resListSimSeqZellerRows, FUN=extractCols, SIMPLIFY=FALSE)
  datListZeller = lapply(resListSimSeqZeller, function(x){x$RCM$X})
save(datListHum, IDlistHum, groupFactorHum, resListSimSeqHumRows, resListSimSeqHumCols, IDlistZeller, resListSimSeqZellerRows, resListSimSeqZellerCols, groupFactorZeller, datListZeller, file = "RCMsim/resListSimSeq.RData")
} else {load("RCMsim/resListSimSeq.RData")}
```

```{r resListSimSeqHum sampleBiplot, purl = FALSE}
#A sample biplot

randomScatter(resListSimSeqHumRows, col = groupFactorHum)
# Our method and CoDa show only white noise, CA and PCoA show artificial clusters. Confounding mean and dispersion effects?
#randomScatter(resListSimSeqHumRows, Dim = c(1,3))
```

Except for PCoA with Bray-Curtis distances and CA, all methods seem to separate the clusters rather well

```{r libSizeCorrHum, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 1HumHum simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.\\label{fig:libCor1}"}
plotCor(resListSimSeqHumRows, datListHum, scoreDim = "rows")
```

CoDa struggles with the library sizes, but also RC(M) and Hellinger seems to suffer from it slightly!

The taxa

```{r resListSimSeqHumcols sampleBiplot, purl = FALSE, include = FALSE}
#A sample biplot
randomScatter(resListSimSeqHumCols)
```

```{r AbsCorrHum, purl = FALSE, fig.cap = "Correlations of column scores with average relative abundances, results of 1HumHum simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.\\label{fig:libCor1}"}
plotCor(resListSimSeqHumCols, datListHum, scoreDim = "columns", dataMat = TRUE)
```

Taxa scores of correspondence analysis tend to correlate with taxon abundances in some cases

```{r resListSimSeqHumEvalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette1}"}
plotSil(resListSimSeqHumRows, groupFactorHum)
```

In the non-parametric setting RCM achieves by far the highest silhouette!

```{r resListSimSeqHumEvaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.  \\label{fig:dist1}"}
distRatioPlot(resListSimSeqHumRows, groupFactorHum)
```

In the non-parametric setting, RCM achieves the lowest distance ratios, CA the highest

```{r AbCorrHum, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCorHum}"}
plotCor(resListSimSeqHumCols, datListHum, dataMat = TRUE, scoreDim = "columns")
```

CA suffers from correlation with relative abundances, but not on average.

```{r taxRatHum, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRatHum}"}
contrHum = contrTaxaWrap(resListSimSeqHumRows, resListSimSeqHumCols, IDlistHum, groupFactorHum, upDown = FALSE)
contrPlot(contrHum)
```

In the SimSeq setting neither method does really well, surprisingly here Hellinger distance and CoDa are the best performers

```{r resListSimSeqZeller sampleBiplot, purl = FALSE}
#A sample biplot
randomScatter(resListSimSeqZellerRows, col = groupFactorZeller)
# Clearly a tough job
```

Except for PCoA with Bray-Curtis distances and CA, all methods seem to separate the clusters rather well

```{r libSizeCorrZeller, purl = FALSE, fig.cap = "Correlations of row scores with library sizes, results of 1ZellerZeller simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.\\label{fig:libCor1}"}
plotCor(resListSimSeqZellerRows, datListZeller, scoreDim = "rows")
```

CoDa struggles with the library sizes, but also RC(M) and Hellinger seems to suffer from it slightly!

The taxa

```{r resListSimSeqZellercols sampleBiplot, purl = FALSE, include = FALSE}
#A sample biplot
randomScatter(resListSimSeqZellerCols)
```

```{r AbsCorrZeller, purl = FALSE, fig.cap = "Correlations of column scores with average relative abundances, results of 1ZellerZeller simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.\\label{fig:libCor1}"}
plotCor(resListSimSeqZellerCols, datListZeller, scoreDim = "columns", dataMat = TRUE)
```

Taxa scores of correspondence analysis tend to correlate with taxon abundances in some cases

```{r resListSimSeqZellerEvalsil, purl = FALSE, fig.cap = "Mean silhouettes by method, results of 100 simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means. A silhouette of 1 point to perfect classification, 0 to borderline cases, -1 to perfect wrong classification \\label{fig:silhouette1}"}
plotSil(resListSimSeqZellerRows, groupFactorZeller)
```

In the non-parametric setting RCM achieves by far the highest silhouette!

```{r resListSimSeqZellerEvaldist, purl = FALSE, fig.cap = "Mean ratios of within to overal distance by method, results of 100 simulations shown for first two dimensions for simulation with the Dirichlet-multinomial distribution with 10% DA taxa. Red diamonds indicate means.  \\label{fig:dist1}"}
distRatioPlot(resListSimSeqZellerRows, groupFactorZeller)
```

In the non-parametric setting, RCM achieves the lowest distance ratios, CA the highest

```{r AbCorrZeller, purl = FALSE, fig.cap = "Correlations of column scores with abundances, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. 'Control' is a negative control setting of an independent standard normal variable. \\label{fig:abCorZeller}"}
plotCor(resListSimSeqZellerCols, datListZeller, dataMat = TRUE, scoreDim = "columns")
```

CA suffers from correlation with relative abundances, but not on average.

```{r taxRatZeller, purl = FALSE, fig.cap = "Mean ratio of signal taxa to top 50% non-signal taxa in terms of contribution to cluster separation, results of 100 simulations shown for first two dimensions for simulation with the zero-inflated negative binomial distribution with 10% DA taxa. Red diamonds indicate means. \\label{fig:taxRatZeller}"}
contrZeller = contrTaxaWrap(resListSimSeqZellerRows, resListSimSeqZellerCols, IDlistZeller, groupFactorZeller, upDown = TRUE)
contrPlot(contrZeller)
```

## Constrained RC(M)

\newpage

# Homogeneous linear equation and the singular value decomposition

A homogeneous linear equations has zero as its constant term. As a result, a system of homogeneous linear equations can be written as:

$$Ax = 0$$

with matrix __A__ known and vector __x__ unknown.

# Some comments on analysing microbiome count data

##### Do not log-transform count data

- log-transforms come with arbitrary pseudocounts for the many zeroes
- models based on transformed data are hard to interpret in funcion of the untransformed data
- log-transformation does not stabilize the variance
- Accept non-normality and heteroscedasticity instead of trying to transform them away!
- Log-transformation is more robust though and provides better type I error control (see papers O'Hara 2010, Ives 2015, Warton 2016 and this [post](http://stats.stackexchange.com/questions/114848/negative-binomial-glm-vs-log-transforming-for-count-data-increased-type-i-erro?rq=1) )

##### Do not rely on residuals-based approaches

- Most residual based approaches rely on normally behaved residuals
- Microbiome residuals are skewed to the right 
- The expectations should be larger than or equal to 5 for this approximation to hold, which is not the case for the majority of microbiome entries

##### Use a GLM

- Normalizing to relative or rarefied abundances throws away information on the variance and valuable counts
- Leave the data untransformed and use a proper GLM for counts! This will properly model mean and variance and allows incorporation of covariate information

