---
title: "RCMrealDataExamples"
author: "Stijn"
date: "November 9, 2017"
output: pdf_document
---

```{r setup, warning=FALSE, message=FALSE, echo=FALSE, purl=TRUE}
WD = "/home/stijn/PhD/Biplots"
knitr::opts_knit$set(root.dir = WD)
knitr::opts_chunk$set(cache = TRUE, autodep = TRUE,  warning = FALSE, message=FALSE, echo=FALSE, eval=TRUE, tidy = TRUE, fig.width = 9, fig.height=6, purl=TRUE, fig.show = "hold", cache.lazy = FALSE)
```

```{r check-install-load-packages, warning=FALSE, message=FALSE, echo=FALSE, purl=TRUE}
# The required package list:
reqpkg <- c("phyloseq", "ggplot2", "reshape2")
# Load all required packages and show version
for(i in reqpkg)
{
#   print(i) 
#   print(packageVersion(i))
  library(i, quietly=TRUE, verbose=FALSE, warn.conflicts=FALSE, character.only=TRUE)
} 
if(detectCores()==4) setwd(WD)
par(pty="s", mar = c(8,2,4,2), cex.main = 0.9) #Make sure the biplots are square!
palStore = palette()
funFiles = dir("R")
funFilesPub = dir("pubFun")
for (i in funFiles) {source(file.path("R",i))}
for (i in funFilesPub) {source(file.path("pubFun",i))};rm(i)
load("/home/stijn/PhD/American Gut/AGphylo.RData")
load("/home/stijn/PhD/Simulations/data/physeqListV13.RData")
```

We apply the RC(M) method to a number of real datasets.

# HMP

The Human Microbiome Project (HMP) aimed to characterize the healthy human microbiome.

```{r RC(M)_NB real datasets, purl = FALSE}
realNames = c("AGP", "Hard.palate", "Buccal.mucosa", "Anterior.nares", "Left.Antecubital.fossa", "Posterior.fornix",  "Mid.vagina", "Left.Retroauricular.crease", "Vaginal.introitus", "Stool", "Hard.palate")
if(!file.exists(file = "resListRCMagphmp.RData")){
resListRCM = lapply(c(AGphylo, physeqListV13[realNames[-1]]),RCM, k = 3)
names(resListRCM) = realNames
save(resListRCM, file = "resListRCMagphmp.RData")
} else {load(file = "resListRCMagphmp.RData")}

SClevels  = unique(c(unlist(sapply(physeqListV13, function(x){
  levels(get_variable("RUNCENTER",physeq=x))
}))))
#Give similar colours for the same sequencing center
cols = c("cadetblue","cyan", "darkblue","blue","magenta","orange","yellow","brown","black",  "grey80","grey70","pink") 
palette(cols)
```

```{r HMPplot, purl = FALSE, include = FALSE, fig.cap="Biplot of the RC(M) ordination of the Anterior nares dataset of the HMP. Colours indicate sequencing center. Sequencing center clearly affects the obtained microbiome compositions. \\label{fig:samHMP}"}
plot(resListRCM[["Anterior.nares"]], samColour ="RUNCENTER", colLegend = "Sequencing center")
```

## Conditioning

Since we know that sequencing facility affects the outcome of the sequencing assay and one is not interested in visualizing this technical variability we can condition out this variable. We condition only on the primary sequencing center.

```{r HMP filter out variable effect, purl=FALSE, fig.cap="Biplot of the RC(M) ordination of the HMP Anterior nares dataset after conditioning. Colours indicate sequencing center. The effect of sequencing center has been largely filtered out. \\label{fig:samHMP}"}
if(!file.exists(file="/home/stijn/PhD/Biplots/filtResHMP.RData")){
  K=2
nleqslv.control = list()
#Group the runcenters together

RCMhmpFilt = lapply(physeqListV13, function(x){
  runcenters = data.frame(sample_data(x))[ "RUNCENTER"]
runcenters[["RUNCENTER"]][runcenters[["RUNCENTER"]] %in% c("BCM","BI,BCM","BCM,BI","BCM,WUGC", "BCM,JCVI")] = "BCM"
runcenters[["RUNCENTER"]][runcenters[["RUNCENTER"]] %in% c("JCVI","JCVI,BI","JCVI,WUGC")] = "JCVI"
runcenters[["RUNCENTER"]][runcenters[["RUNCENTER"]] %in% c("WUGC,JCVI","WUGC","WUGC,BCM")] = "WUGC"
sample_data(x) = data.frame(sample_data(x) ,RUNCENTER2 = runcenters[[1]])
x})
filtRes = lapply(RCMhmpFilt, function(x){
  RCM(x, distribution="NB", k = K, nleqslv.control= nleqslv.control, maxItOut=5e3, prevCutOff=0.01, colWeights="marginal", rowWeights = "uniform", marginEst = "MLE", round=TRUE, confounders = "RUNCENTER2")
  })
save(filtRes, file="/home/stijn/PhD/Biplots/filtResHMP.RData")
} else {load(file="/home/stijn/PhD/Biplots/filtResHMP.RData")}

plot(filtRes[["Anterior.nares"]], samColour ="RUNCENTER", colLegend = "Sequencing center")
```

# The American gut project

The American gru project consits of stool samples sampled by volunteers at home, together with their answers to a questionnaire. Since they are sampled at home the variability is expected to be large.

```{r AGPplots, purl = FALSE}
plot(resListRCM[["AGP"]], samColour = "IBD")
```

The AGP dataset is presumably too noisy to find much of a signal

# The Zeller data

The Zeller data are obtained from a study on colorectal cancer in cancer patients and healthy controls \cite{Zeller2014}. Patient covariates recorded were age, gender, BMI, cancer diagnosis(healthy, small adenoma or cancer) and country(France or Germany).

```{r load zeller RC(M) NB results, purl = FALSE, include = FALSE}
load("/home/stijn/PhD/Simulations/data/zellerData.RData")
zellers = c("zellerMeta","zeller16S")
if(!file.exists("zellerSphyRCM.RData")){
zellerSphyRCM = RCM(zellerSphy, k=3, round = TRUE)
save(zellerSphyRCM, file="zellerSphyRCM.RData")
} else {load("zellerSphyRCM.RData")}

if(!file.exists("zellerMphyRCM.RData")){
zellerMphyRCM = RCM(zellerMphy, k=3, round = TRUE, prevCutOff = 0.05, minFraction = 0.15) # Less than one minute! More trimming needed here
save(zellerMphyRCM, file="zellerMphyRCM.RData")
} else {load("zellerMphyRCM.RData")}

zellerRCM  =list(zellerMphyRCM, zellerSphyRCM)
names(zellerRCM) = zellers

# sapply(zellerRCM,function(x){x$converged}) # All converged
# sapply(zellerRCM,function(x){x$runtime})
# sapply(zellerRCM,function(x){x$iter})
```

```{r plotZellerSphyDiagnosis, purl = FALSE, eval = FALSE}
plot.RCM(zellerSphyRCM, samColour = "Diagnosis", samShape = "Country")
```

```{r plotZellerMphy, purl = FALSE, eval = FALSE}
plot.RCM(zellerMphyRCM, samColour = "Diagnosis")
```

For the 16S data we see signals for Country and Cancer diagnosis. For the metagenomics data we do not see a trend. Notice we are doing indirect gradient analysis again, but this is just to look for confirmation that the unconstrained analysis is detecting true groups.

\clearpage

## The Zeller data: Constrained RC(M)

In the constrained analysis we use all the available covariates to construct the environmental gradient.

```{r zellerConstrained, purl =FALSE}
if(!file.exists("zellerSphyRCMconstr.RData")){
zellerSphyRCMconstrLin = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "linear") # Use smallest possible cndtol!
zellerSphyRCMconstrLinML = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "linear", envGradEst = "ML")
zellerSphyRCMconstrQuad = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "quadratic")
zellerSphyRCMconstrNonParam = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "nonparametric")
zellerSphyRCMconstrNonParamML = RCM(zellerSphy, k=2, round = TRUE, covariates = c("Age","Gender","BMI","Country", "Diagnosis"), responseFun = "nonparametric", envGradEst = "ML")
save(zellerSphyRCMconstrLin, zellerSphyRCMconstrLinML, zellerSphyRCMconstrQuad, zellerSphyRCMconstrNonParam,file="zellerSphyRCMconstr.RData")
} else {load("zellerSphyRCMconstr.RData")}
```

```{r ZellerConstrainedBiplot, purl=FALSE, eval = FALSE, fig.cap="Triplot of constrained RC(M) analysis on zeller data with linear response functions. Cancer diagnosis and country appear to be the most important variables. \\label{fig:triZellerDiagnosis}"}
plot.RCM(zellerSphyRCMconstrLin, samColour = "Diagnosis", samShape = "Country")
```

```{r ZellerConstrainedBiplot, purl=FALSE, eval = FALSE, fig.cap="Triplot of constrained RC(M) analysis on Zeller data with non-parametric response functions. The environmental gradient is similar as to the linear case, nly age is more important according to this model. \\label{fig:triZellerDiagnosisNP}"}
plot.RCM(zellerSphyRCMconstrNonParam, samColour = "Diagnosis", samShape = "Country")
```

# Turnbaugh et al. (2009)

In this study the fecal microbiome of monozygotic and dizygotic twins was characterized \cite{Turnbaugh2009}. On of the aims was to find a link with obesity.

```{r turnTwins, eval = FALSE, purl = FALSE}
if(!file.exists("turntwinsRCm.RData")){
  load(file = "/home/stijn/PhD/Datasets/TurnbaughObeseTwins/TurnTwins.RData")
turntwinsRCm = RCM(TurnTwins, k=3)
save(turntwinsRCm, file = "turntwinsRCm.RData")
} else {load("turntwinsRCm.RData")}
plot.RCM(turntwinsRCm, samColour = "Status_p", colLegend = "Weight status")
rm(turntwinsRCm)
```

With such small datasets it is hard to draw conclusions.

# Turnbaugh et al. (2009b)

In this group, 15 gnotobiotic mice were inoculated with human feces, and for one group the diet was switched to a Western diet after one month \cite{Turnbaugh200b9}. Then a second generation of mice was inoculated with cecal samples from the previous groups, and here also diet was varied. The variables recorded were:

- Diet: current diet
- Generation_p: "Recipient1" if second generation, otherwise diet of cecal samples with which they were inoculated
- Time: age of mouse at sampling
- DGS: Diet and previous diet
- DTG: combines sampling diet, time, and previous generation (if any) 

```{r turnmice, purl = FALSE}
if(!file.exists("turnMiceRCM.RData")){
  load(file = "/home/stijn/PhD/Datasets/Humanized mouse/HumMicePhy.RData")
HumMiceRCM = RCM(HumMicePhy, k=3)
HumMiceRCMlin = RCM(HumMicePhy, k=3, covariates = c("Diet","Generation_p","Time"))
save(HumMiceRCM, file = "turnMiceRCM.RData")
} else {load("turnMiceRCM.RData")}
```

```{r humMicePlot, purl = FALSE}
plot.RCM(HumMiceRCM) # Two or three distinct groups
```

This dataset appears to have a very strong signal, we can distinguish three or four clusters.

```{r humMicePlotSecondLook, purl = FALSE}
plot.RCM(HumMiceRCM, samColour = "Diet", samShape = "Generation_p")
```

It is clear that current diet is the dominant variable in the first dimension.

```{r humMicePlotSecondLook, purl = FALSE}
plot.RCM(HumMiceRCM, samColour = "Time", samShape = "Generation_p")
```

Also the older group of generation 1 mice is different, it separates in the second dimension.

##### Kostic colorectal cancer

This is a study on the microbiome of colorectal cancer in humans. Nine tumor patients were matched with 9 healthy patients, samples were taken repeatedly. The researchers find a enrichment of Fusobacteria in the tumor and a depletion of Bacteroidetes and Firmicutes.

```{r load Kostic, purl=FALSE}
if(!file.exists("Kostic.RData")){
filepath = system.file("extdata", "study_1457_split_library_seqs_and_mapping.zip", package="phyloseq")
kostic = microbio_me_qiime(filepath)
save(kostic, file = "Kostic.RData")
} else {load("Kostic.RData")}
```

```{r KosticRCM}, purl = FALSE, eval = FALSE}
if(!file.exists("kosticRCM.RData")){
kosticRCM = RCM(kostic, k = 3)
save(kosticRCM, file = "kosticRCM.RData")
} else {load("kosticRCM.RData")}
plot(kosticRCM)
plot(kosticRCM, samColour = "DIAGNOSIS")
plot(kosticRCM, samColour = "OSH_DIAGNOSIS")
```

##### Dr. armpit data

The dataset contains 16S sequencing data from armpit swabs of volunteers.

```{r ArmPit, purl =  FALSE}
if(!file.exists("armpit.RData")){
load("Armpit/okseldata.rda") 
load("Armpit/div.rda") 
load("Armpit/odor.rda")
db<-data.frame(odor,info,food,washing)
db$subject = db$subject.1
db$subject.2 = db$subject.1 = NULL
load("Armpit/okseldata_20160229.rda") # data in object counts
Counts<-counts[,-903]
rownames(Counts) = rownames(db) = db$subject
OTUa = Counts[,-903]
armpit = phyloseq(otu_table(apply(as.matrix(OTUa), 1:2, as.numeric), taxa_are_rows =FALSE), sample_data(data.frame(db)))
save(armpit, file = "armpit.RData")
} else {load(file = "armpit.RData")}

if(!file.exists(file = "armpitRCM.RData")){
  armpitRCM = RCM(armpit, k=3)
  armpitRCMconstrLin = RCM(armpit, covariates = sample_variables(armpit)[!sample_variables(armpit) %in% c("subject")], k=3, responseFun = "linear")
  armpitRCMconstrQuad = RCM(armpit, covariates = sample_variables(armpit)[!sample_variables(armpit) %in% c("subject")], k=3, responseFun = "quadratic")
  armpitRCMconstrNonParam = RCM(armpit, covariates = sample_variables(armpit)[!sample_variables(armpit) %in% c("subject")], k=3, responseFun = "nonparametric")
  lessVariablesArmpit = c("BMI", "Fastfood", "Meat", "Vegetables", "gender")
  armpitRCMconstrLinLess = RCM(armpit, covariates = lessVariablesArmpit, k=3, responseFun = "linear")
  armpitRCMconstrNonParamLess = RCM(armpit, covariates = lessVariablesArmpit, k=3, responseFun = "nonparametric")
    save(armpitRCM, armpitRCMconstrLin, armpitRCMconstrQuad, armpitRCMconstrNonParam, armpitRCMconstrLinLess, armpitRCMconstrNonParamLess, file = "armpitRCM.RData")
} else {load(file = "armpitRCM.RData")}
```
##### CMET data

This is a longitudinal dataset on growth in a water cooling system. It has the unique property of providing cell concentrations from flow cytometry from the same samples as used for the 16S sequencing. We could use this information to alleviate the compositional effect. This is relatively new in the literature, and no standard way exists of incorporating this information. In general, the cell concentratios and library sizes correlate poorly.

See separate document for extended discussion.

```{r readInCMET, purl = FALSE, eval = FALSE}
if(!file.exists(file = "CMETdata.RData")){
  library(openxlsx)
  samData = read.xlsx("/home/stijn/PhD/Datasets/CMET/Metadata.xlsx")
  rownames(samData) = samData$sample_title
  samData$Reactor.cycle = as.character(samData$Reactor.cycle)
  samData$cellDensity = samData$`Cell.density.(cells/mL)`
  samData$`Cell.density.(cells/mL)` = NULL
  otuTab = read.csv("/home/stijn/PhD/Datasets/CMET/OTU_table_raw.csv", sep = ";")
  rownames(otuTab) = otuTab$Sample_ID
  otuTab$Sample_ID = NULL
  #Select only samples present in samData
  otuTab = otuTab[as.character(samData$sample_title),]
  #Multiply relative abundances by cell counts in samData
  otuTab = round(as.matrix(otuTab)/rowSums(as.matrix(otuTab))*samData$cellDensity)
  taxTab = read.csv("/home/stijn/PhD/Datasets/CMET/tax_table_raw.csv", sep = ";")
  rownames(taxTab) = taxTab$X
  taxTab$X = NULL
  samData = samData[rowSums(otuTab)>0,]
  taxTab = taxTab[ colSums(otuTab)>0,]
  otuTab = otuTab[rowSums(otuTab)>0, colSums(otuTab)>0] #Number of zeroes is not related to rounding
  CMETwater = phyloseq(otu_table(otuTab, taxa_are_rows = FALSE), sample_data(samData), tax_table(as.matrix(taxTab)))
  save(CMETwater, file = "CMETdata.RData")
}else {load(file = "CMETdata.RData")}
```

```{r analysCMET, purl = FALSE}
if(!file.exists(file = "CMETRCM.RData")){
  CMETRCM = RCM(CMETwater, k = 3, prevCutOff = 0.05)
  envVarsCMET = c("ph","temp","Conductivity","cellDensity", "Reactor.phase", "Reactor.cycle")
  CMETRCMlin = RCM(CMETwater, k = 3, responseFun = "linear", covariates = envVarsCMET, prevCutOff = 0.05)
  CMETRCMlinML = RCM(CMETwater, k = 3, responseFun = "linear", covariates = envVarsCMET, prevCutOff = 0.05, envGradEst = "ML")
  CMETRCMquad = RCM(CMETwater, k = 3, responseFun = "quadratic", covariates = envVarsCMET, prevCutOff = 0.05)
  CMETRCMnonParam= RCM(CMETwater, k = 3, responseFun = "nonparametric", covariates = envVarsCMET, prevCutOff = 0.05)
  save(CMETRCM, CMETRCMlin, CMETRCMquad,CMETRCMnonParam, file = "CMETRCM.RData")
} else {load(file = "CMETRCM.RData")}
```

No convergence for quadratic case. That one seems hard to fit.

```{r CMETliks, purl = FALSE, eval = FALSE}
liksCMET = sapply(list("unconstrained" = CMETRCM, "linear" = CMETRCMlin,"quadratic" = CMETRCMquad, "nonParametric" = CMETRCMnonParam), liks)
#The comparison is complicated here because we trimmed the linear case.
liksCMERmelt = melt(liksCMET, varnames = c("Dimension","responseFun"),value.name = "logLik")
ggplot(aes(x = Dimension, y = logLik, colour =responseFun), data = liksCMERmelt) + geom_point() +geom_line(aes(group = responseFun))
```

```{r plotCMET, purl = FALSE}
plot(CMETRCM)
plot(CMETRCM, samColour = "ph")
plot(CMETRCM, samColour = "temp")
plot(CMETRCM, samColour = "Conductivity")
plot(CMETRCM, samColour = "cellDensity")
plot(CMETRCM, samColour = "Reactor.phase")
plot(CMETRCM, samColour = "Timepoint")
plot(CMETRCM, samColour = "collection_date")
plot(CMETRCM, samColour = "Reactor.cycle")
```

We clearly have two groups in the second dimension, partly explained by the _Reactor.phase_ and _collection.date_ variables. Still feels as if we're missing a confounder here. In the first dimension we have a strong time signal.

```{r plotCMETlin, purl = FALSE}
plot(CMETRCMlin, plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMlin, plotTaxa = FALSE, square = FALSE, Dim = c(2,1))
plot(CMETRCMlin, samColour = "ph", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMlin, samColour = "temp", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMlin, samColour = "Conductivity", plotTaxa = FALSE,  Dim = c(2,1))
plot(CMETRCMlin, samColour = "cellDensity", plotTaxa = FALSE,  Dim = c(2,1))
plot(CMETRCMlin, samColour = "Reactor.phase", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMlin, samColour = "Timepoint", plotTaxa = FALSE,  Dim = c(2,1))
plot(CMETRCMlin, samColour = "collection_date", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMlin, samColour = "Reactor.cycle", plotTaxa = FALSE, Dim = c(2,1))
```

Reactor phase clearly explains most of the variation in the data

```{r plotCMETquad, purl = FALSE}
plot(CMETRCMquad, plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMquad, plotTaxa = FALSE, square = FALSE, Dim = c(2,1))
plot(CMETRCMquad, samColour = "ph", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMquad, samColour = "temp", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMquad, samColour = "Conductivity", plotTaxa = FALSE,  Dim = c(2,1))
plot(CMETRCMquad, samColour = "cellDensity", plotTaxa = FALSE,  Dim = c(2,1))
plot(CMETRCMquad, samColour = "Reactor.phase", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMquad, samColour = "Timepoint", plotTaxa = FALSE,  Dim = c(2,1))
plot(CMETRCMquad, samColour = "collection_date", plotTaxa = FALSE, Dim = c(2,1))
plot(CMETRCMquad, samColour = "Reactor.cycle", plotTaxa = FALSE, Dim = c(2,1))
```

The control phase seems to be very different from the others, also conductivity plays a role. Also the second dimension is the most important. Should we reorder the dimensions? Also the quadratic graph differs quite a lot from the linear one.
